---
title: "Auscad update 2022"
author: "Jen Li"
output:
  html_document: default
---

Set up for analysis
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
setwd("~/Documents/Project data - R files/auscad/")
set.seed(42)

  knitr::opts_chunk$set(echo = FALSE, cache = FALSE, 
                        message = FALSE, warning = FALSE)
  
  source("data/setup.auscad.R")
  source("data/auscad.extra.functions.R")
  # if new run/comp, run: source("data/clean.data.R")

  # record of pkgs
  writeLines(capture.output(sessionInfo()),"data/current.packages.txt")
  write.table(pkg.versions, "data/package.versions.txt", append = FALSE,   
              sep = " ", dec = ".",row.names = TRUE, col.names = TRUE)
  
  
# My data organised in the following way: metadata = clinical metadata, samples = sequencing metadata,  counts = raw counts matrix,  conversions = ensembl, entrez, symbols. Keep original so not overwritten
  auscad.sorted = new("auscad.sorted",
                      metadata.all = readRDS("data/database/annotatedDB.RDS"),
                      samples.all = readRDS("data/database/samples.all.RDS"),
                      counts.all = readRDS("data/database/counts.all.RDS"),
                      conversions = readRDS("data/database/conversion.RDS"),
                      samples.biopsy.0m = readRDS("data/database/biopsy0m.RDS"),
                      counts.biopsy.0m = readRDS("data/database/counts0m.RDS"),
                      samples.biopsy.1m = readRDS("data/database/biopsy1m.RDS"),
                      counts.biopsy.1m = readRDS("data/database/counts1m.RDS"),
                      samples.biopsy.3m = readRDS("data/database/biopsy3m.RDS"),
                      counts.biopsy.3m = readRDS("data/database/counts3m.RDS"),
                      samples.blood.0m = readRDS("data/database/blood0m.RDS"),
                      counts.blood.0m = readRDS("data/database/bloodcounts0m.RDS"),
                      samples.blood.3m = readRDS("data/database/blood3m.RDS"),
                      counts.blood.3m = readRDS("data/database/bloodcounts3m.RDS"))


### Extract the clinical dataset (time in months unless specified)
  dgf.filter = auscad.sorted@metadata.all
  
  # Keep if at least 12months in study, or event before 12m
  dgf.filter = dgf.filter %>% filter (time.in.study >=12 | event.main == 1)
  
  dgf.filter = dgf.filter %>% 
  mutate(mgfr_12m1 = 
           case_when(mgfr_12m < 30 ~ "<30", 
                     mgfr_12m >=30 & mgfr_12m < 45 ~ "30-45",
                     mgfr_12m >=45 & mgfr_12m < 60 ~ "45-60",
                     mgfr_12m >= 60 ~ ">60"))

  dgf.filter = dgf.filter %>% 
  mutate(mgfr_12m2 = 
           case_when(mgfr_12m < 45 ~ "<45", 
                     mgfr_12m >= 45 ~ ">=45"))
  
  dgf.filter = dgf.filter %>% 
  mutate(donor_criteria.main = 
           case_when(donor_criteria == "LKD" ~ "LKD", 
                     donor_criteria == "DBD" |
                       donor_criteria == "DBD + ECD" ~ "DBD",
                     donor_criteria == "DCD" |
                       donor_criteria == "DCD + ECD" ~ "DCD"))
    
  # saveRDS(dgf.filter, "data/auscad.filtered.jen2022.RDS")

gc()
```

At a glance - clinical and RNA sets
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
# Basic summaries
# dgf.filter %>% ggplot(aes(x = time.in.study)) + geom_histogram(color="gray", fill="white") + facet_grid(.~dgf_combined)

# Clinical event summaries
print("Auscad clinical (>=12m or event)")
paste0("n participants: ", nrow(dgf.filter))
paste0("n pre-filter: ", nrow(dgf.filter[dgf.filter$time.in.study >=12,]))
paste0("n >= 36m: ", nrow(dgf.filter[dgf.filter$time.in.study >=36,]))
paste0("n >= 60m: ", nrow(dgf.filter[dgf.filter$time.in.study >=60,]))
paste0("composite events: ", sum(dgf.filter$event.main))
paste0("all graft loss: ",sum(dgf.filter$graft_loss == "Kidney Loss"))
paste0("non-death.cs graft loss: ",sum(dgf.filter$graft.loss.noncensored))
summary(dgf.filter$death_functioninggraft)
paste0("death (all):", sum(dgf.filter$death == "Deceased"))
paste0("bpar - any: ",sum(dgf.filter$bpar_any))
paste0("bpar - within 1m: ",sum(dgf.filter$bpar_0_1m == "BPAR (0-1m)"))
paste0("subclin - any: ",sum(dgf.filter$subclin_any))
paste0("subclin - within 1m: ",sum(dgf.filter$Subclin_0_1m == "Subclin (0-1m)"))
paste0("ddKTx: ",sum(dgf.filter$organ_type == "Kidney Transplant"))
paste0("ldKTx: ",sum(dgf.filter$organ_type == "Living Kidney Transplant"))
paste0("SPK: ",sum(dgf.filter$organ_type == "SPK"))

summary(dgf.filter$dgf_define.comb)
summary(dgf.filter$dgf_define)
summary(dgf.filter$dgf.severity)


####### Number of biopsy RNAseq available at 0m
gene.check = auscad.sorted@samples.biopsy.0m
gene.check = gene.check %>% filter (event.main == 1 |  time.in.study >=12)

print("Kidney Bulk RNAseq (0m samples)")
summary(as.factor(gene.check$dgf_define))
summary(as.factor(gene.check$dgf.severity))

#######  Number of biopsy RNAseq available at 1m
gene.check = auscad.sorted@samples.biopsy.1m
gene.check = gene.check %>% filter (event.main == 1 |  time.in.study >=12)

print("Kidney Bulk RNAseq (1m samples)")
summary(as.factor(gene.check$dgf_define))
summary(as.factor(gene.check$dgf.severity))

#######  Number of biopsy RNAseq available at 3m
gene.check = auscad.sorted@samples.biopsy.3m
gene.check = gene.check %>% filter (event.main == 1 |  time.in.study >=12)


print("Kidney Bulk RNAseq (3m samples)")
summary(as.factor(gene.check$dgf_define))
summary(as.factor(gene.check$dgf.severity))

```


Biopsy kinetics
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
# Sankey/Alluvial
dgf.filter1 = dgf.filter %>% filter(dgf_define!= "DGF.Cr") %>%
  select(dgf_define, organ_type,donor_criteria, bpar_any1, IFTA_score_3m_meena1, IFTA_score_12m_meena1)
  dgf.filter1$dgf_define = droplevels(dgf.filter1$dgf_define)
  dgf.filter1 = data.frame(ftable(dgf.filter1))

ggplot(dgf.filter1,
              aes(axis1 = dgf_define, axis2 = bpar_any1, axis3 = IFTA_score_3m_meena1, 
                  axis4 = IFTA_score_12m_meena1,y = Freq)) +
  geom_alluvium(aes(fill = dgf_define),curve_type = "cubic",na.rm = TRUE) +
  geom_stratum() + geom_text(stat = "stratum", aes(label = after_stat(stratum))) +
  scale_x_discrete(limits = c("DGF criteria", "BPAR.any", "IFTA3m", "IFTA12m"), expand = c(.1, .1)) +
  labs(title = "Auscad", subtitle = "stratified by organ_type, and IFTA12m",y = "Proportion") +
  theme_minimal() + scale_colour_brewer() + theme(legend.position = "bottom") 

# Bx kinetics function 
# data, ("all" or "dgf"), ("full", "any"), ("1m", "3m", "12m"), ("facet", "plain")
plots.comb = bx.kinetics(dgf.filter, "all", "full", "12m", "facet") 
plots.comb = bx.kinetics(dgf.filter, "all", "full", "12m", "plain") 
plots.comb = bx.kinetics(dgf.filter, "all", "full", "3m", "plain") 
plots.comb = bx.kinetics(dgf.filter, "all", "full", "1m", "plain") 
plots.comb

plots.comb.dgf = bx.kinetics(dgf.filter, "dgf", "full", "12m", "facet") 
plots.comb.dgf = bx.kinetics(dgf.filter, "dgf", "full", "12m", "facet") 
plots.comb.dgf = bx.kinetics(dgf.filter, "dgf", "full", "3m", "facet") 
plots.comb.dgf = bx.kinetics(dgf.filter, "dgf", "full", "1m", "facet") 

plots.comb.dgf = bx.kinetics(dgf.filter, "dgf", "full", "12m", "facet") 
plots.comb.dgf = ggarrange(plots.comb.dgf[[1]], plots.comb.dgf[[2]], plots.comb.dgf[[3]], plots.comb.dgf[[4]])
plots.comb.dgf


```


mGFR/eGFR slope
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}

dgf.filter1 = dgf.filter 
dgf.filter1 = dgf.filter1 %>% 
  filter(!is.na(mgfr_12m) & !is.na(mgfr_3m)) %>%
  mutate(delta.mgfr = mgfr_12m - mgfr_3m)
dgf.filter1 = dgf.filter1 %>% 
  filter(!is.na(egfr_3m) & !is.na(egfr_12m)) %>%
  mutate(delta.egfr = egfr_12m - egfr_3m)

ggplot(dgf.filter1[dgf.filter1$donor_criteria.main != "LKD"
                   & dgf.filter1$dgf_define != "DGF.Cr",], 
       aes(x = donor_criteria.main, y = delta.mgfr))+
  geom_boxplot(aes(colour = dgf_define)) + 
  geom_jitter(width = 0.3, alpha = 0.5)+
  ggtitle("delta mGFR 3-12m")+theme_bw()+
  stat_compare_means() + facet_grid(bpar_any1~bparupto03m)

ggplot(dgf.filter1[dgf.filter1$donor_criteria.main != "LKD"
                   & dgf.filter1$dgf_define != "DGF.Cr",], 
       aes(x = donor_criteria.main, y = delta.egfr))+
  geom_boxplot(aes(colour = dgf_define)) + 
  geom_jitter(width = 0.3, alpha = 0.5,)+
  ggtitle("delta eGFR 3-12m")+theme_bw()+
  stat_compare_means()+ facet_grid(bpar_any1~bparupto03m)

dgf.filter2 = dgf.filter1 %>%
  pivot_longer(cols = c("mgfr_3m", "mgfr_12m"),
               names_to = "mgfr.time",
               values_to = "mgfr.value")

ggpaired(dgf.filter2[dgf.filter2$dgf_define == "DGF.HD",], 
         x = "mgfr.time", 
         y = "mgfr.value") + 
  facet_grid(bparupto03m~.)+
  stat_compare_means(paired = TRUE)


```


Extra analysis
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
# basics
cv.upto5yrs = table(dgf.filter$ihd_pre, dgf.filter$cvd_0_5y)
cv.upto5yrs = cv.upto5yrs[c(2,1),]
cv.RR.pre = riskratio.wald(cv.upto5yrs)
cv.RR.pre

### CV 0 - 5yrs
fit3 = dgf.filter %>% finalfit("cvd_0_5y", 
                              c("dgf_define", "ihd_pre",
                                "htn_pre", "dm_pre", "nodat",
                                "renal_dx", "bpar_any"), 
                              keep_models = TRUE,
                              metrics = TRUE)

knitr::kable(fit3[[1]], row.names=FALSE)
knitr::kable(fit3, row.names=FALSE, 
             align=c("l", "l", "r", "r", "r", "r"))

### Recurrent UTI
fit3 = dgf.filter %>% finalfit("recurrent_uti", 
                              c("dgf_define", 
                                "gender",
                                "nodat",
                                "tac_d21",
                                "tac_1m",
                                "tac_3m",
                                "tac_12m",
                                "renal_dx", 
                                "Subclin_0_1m",
                                "induct_extra",
                                "cmv_viremia",
                                "cmv_disease",
                                "bpar_any", 
                                "subclin_any", 
                                "bkvan_any"), 
                              keep_models = TRUE,
                              metrics = TRUE)

knitr::kable(fit3[[1]], row.names=FALSE)
knitr::kable(fit3, row.names=FALSE, 
             align=c("l", "l", "r", "r", "r", "r"))

### BPAR_0_1m and 1m
fit3 = dgf.filter[dgf.filter$organ_type != "Living Kidney Transplant",] %>% 
  finalfit("bpar_0_1m", c("dgf_define",  "Subclin_0_1m",
                          "induct_extra", "graft_number",
                          "donor_criteria", "HLA_mm",
                          "dsa1_pre", "dsa2_pre",
                          "donor_age", "tx_age", 
                          "gender", "donor_gender",
                          "tac_d2", "tac_d7", "tac_d21",
                          "organ_type"), 
                              keep_models = TRUE,
                              metrics = TRUE)

knitr::kable(fit3[[1]], row.names=FALSE)
knitr::kable(fit3, row.names=FALSE, 
             align=c("l", "l", "r", "r", "r", "r"))

fit3 = dgf.filter[dgf.filter$organ_type != "Living Kidney Transplant",] %>% 
  finalfit("bpar_1m", c("dgf_define",  "Subclin_0_1m",
                          "induct_extra", "graft_number",
                          "donor_criteria", "HLA_mm",
                          "dsa1_pre", "dsa2_pre",
                          "donor_age", "tx_age", 
                          "gender", "donor_gender",
                          "tac_d2", "tac_d7", "tac_d21",
                          "organ_type"), 
                              keep_models = TRUE,
                              metrics = TRUE)

knitr::kable(fit3[[1]], row.names=FALSE)
knitr::kable(fit3, row.names=FALSE, 
             align=c("l", "l", "r", "r", "r", "r"))

### subclin0_1m
fit3 = dgf.filter[dgf.filter$organ_type != "Living Kidney Transplant",] %>% 
  finalfit("Subclin_0_1m", c("dgf_define",  "bpar_0_1m",
                          "induct_extra", "graft_number",
                          "donor_criteria", "HLA_mm",
                          "dsa1_pre", "dsa2_pre",
                          "donor_age", "tx_age", 
                          "gender", "donor_gender",
                          "tac_d2", "tac_d7", "tac_d21",
                          "organ_type"), 
                              keep_models = TRUE,
                              metrics = TRUE)

knitr::kable(fit3[[1]], row.names=FALSE)
knitr::kable(fit3, row.names=FALSE, 
             align=c("l", "l", "r", "r", "r", "r"))



```



Basic plots
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
new.matrix = dgf.filter %>% select(dgf_define, 
                                   cr_1m, egfr_1m, 
                                   cr_3m, egfr_3m,
                                   cr_12m, egfr_12m,
                                   mgfr_3m, mgfr_12m, cr_60m,
                                   IFTA_score_3m_meena1, iIFTA_score_3m_meena1,
                                   IFTA_score_12m_meena1, iIFTA_score_12m_meena1)
new.matrix$IFTA.12 = new.matrix$IFTA_score_12m_meena1
new.matrix$iIFTA.12 = new.matrix$iIFTA_score_12m_meena1
new.matrix$IFTA.3 = new.matrix$IFTA_score_3m_meena1

# new.matrix = new.matrix[!is.na(new.matrix$IFTA_score_12m_meena1),]
# new.matrix$IFTA_score_12m_meena1 = as.numeric(new.matrix$IFTA_score_12m_meena1)
my.comp = list(c( "Control", "DGF.Cr"), c( "DGF.Cr", "DGF.HD"), c("Control", "DGF.HD"))

p.cr1m = ggplot(new.matrix, aes(x = dgf_define,  y = cr_1m)) +
  geom_boxplot(aes(fill = dgf_define), alpha = 0.5, outlier.shape = NA) +
  geom_jitter(position=position_jitter(width=.3, height=0), alpha = 0.5)+
  theme_bw() + ylab("umol/L") + xlab("")+ggtitle("Cr - 1 month")+
  stat_compare_means( method = "anova",  size = 3, label.y = 600) + 
  stat_compare_means(comparisons = my.comp, size = 3)

p.egfr1m = ggplot(new.matrix, aes(x = dgf_define,  y = egfr_1m)) +
  geom_boxplot(aes(fill = dgf_define), alpha = 0.5, outlier.shape = NA) + 
  geom_jitter(position=position_jitter(width=.3, height=0), alpha = 0.5)+
  theme_bw() + ylab("eGFR (ml/min/1.73m2)") + xlab("")+ggtitle("eGFR - 1 month") +
  stat_compare_means( method = "anova",  size = 3, label.y = 120) + 
  stat_compare_means(comparisons = my.comp, size = 3)

p.cr3m = ggplot(new.matrix, aes(x = dgf_define,  y = cr_3m)) +
  geom_boxplot(aes(fill = dgf_define), alpha = 0.5, outlier.shape = NA) + 
  geom_jitter(position=position_jitter(width=.3, height=0), alpha = 0.5)+
  theme_bw() + ylab("umol/L") + xlab("")+ggtitle("Cr - 3 months") +
  stat_compare_means( method = "anova", label.y = 350, size = 3) + 
  stat_compare_means(comparisons = my.comp, size = 3)

p.egfr3m = ggplot(new.matrix, aes(x = dgf_define,  y = egfr_3m)) +
  geom_boxplot(aes(fill = dgf_define), alpha = 0.5, outlier.shape = NA) +
  geom_jitter(position=position_jitter(width=.3, height=0), alpha = 0.5)+
  theme_bw() + ylab("eGFR (ml/min/1.73m2)") + xlab("")+ggtitle("eGFR - 3 months") + 
  stat_compare_means( method = "anova", label.y = 130, size = 3) + 
  stat_compare_means(comparisons = my.comp, size = 3)


p.mgfr3 = ggplot(new.matrix, aes(x = dgf_define,  y = mgfr_3m)) +
  geom_boxplot(aes(fill = dgf_define), alpha = 0.5, outlier.shape = NA) + 
  geom_jitter(position=position_jitter(width=.3, height=0), alpha = 0.5)+
  theme_bw() + ylab("mGFR (ml/min)") + xlab("")+ggtitle("mGFR - 3 months")+
  stat_compare_means( method = "anova", label.y = 190, size = 3) + 
  stat_compare_means(comparisons = my.comp, size = 3)


p.cr12m = ggplot(new.matrix, aes(x = dgf_define,  y = cr_12m)) +
  geom_boxplot(aes(fill = dgf_define), alpha = 0.5, outlier.shape = NA) + 
  geom_jitter(position=position_jitter(width=.3, height=0), alpha = 0.5)+
  theme_bw() + ylab("umol/L") + xlab("")+ggtitle("Cr - 12 months") +
  stat_compare_means( method = "anova", label.y = 350, size = 3) + 
  stat_compare_means(comparisons = my.comp, size = 3)

p.egfr12m = ggplot(new.matrix, aes(x = dgf_define,  y = egfr_12m)) +
  geom_boxplot(aes(fill = dgf_define), alpha = 0.5, outlier.shape = NA) +
  geom_jitter(position=position_jitter(width=.3, height=0), alpha = 0.5)+
  theme_bw() + ylab("eGFR (ml/min/1.73m2)") + xlab("")+ggtitle("eGFR - 12 months") + 
  stat_compare_means( method = "anova", label.y = 130, size = 3) + 
  stat_compare_means(comparisons = my.comp, size = 3)


p.mgfr12= ggplot(new.matrix, aes(x = dgf_define,  y = mgfr_12m)) +
  geom_boxplot(aes(fill = dgf_define), alpha = 0.5, outlier.shape = NA) + 
  geom_jitter(position=position_jitter(width=.3, height=0), alpha = 0.5)+
  theme_bw() + ylab("GFR (ml/min)") + xlab("") + ggtitle("mGFR - 12 months")+
  stat_compare_means( method = "anova", label.y = 190, size = 3) + 
  stat_compare_means(comparisons = my.comp, size = 3)

p.cr60m = ggplot(new.matrix, aes(x = dgf_define,  y = cr_60m)) +
  geom_boxplot(aes(fill = dgf_define), alpha = 0.5, outlier.shape = NA) + 
  geom_jitter(position=position_jitter(width=.3, height=0), alpha = 0.5)+
  theme_bw() + ylab("umol/L") + xlab("")+ggtitle("Cr - 60 months") +
  stat_compare_means( method = "anova", label.y = 350, size = 3) + 
  stat_compare_means(comparisons = my.comp, size = 3)

p.mgfr3.ifta = ggplot(new.matrix[!is.na(new.matrix$IFTA_score_3m_meena1),], 
                       aes(x = dgf_define,  y = mgfr_3m)) +
  geom_boxplot(aes(fill = IFTA.3), alpha = 0.5, outlier.shape = NA) + 
  geom_jitter(position=position_jitter(width=.3, height=0), alpha = 0.5)+
  theme_bw() + ylab("GFR (ml/min)") + xlab("") + ggtitle("mGFR/IFTA - 3 months")+
  stat_compare_means( method = "anova", label.y = 190, size = 3) + 
  stat_compare_means(comparisons = my.comp, size = 3)

p.mgfr12.ifta = ggplot(new.matrix[!is.na(new.matrix$IFTA_score_12m_meena1),], 
                       aes(x = dgf_define,  y = mgfr_12m)) +
  geom_boxplot(aes(fill = IFTA.12), alpha = 0.5, outlier.shape = NA) + 
  geom_jitter(position=position_jitter(width=.3, height=0), alpha = 0.5)+
  theme_bw() + ylab("GFR (ml/min)") + xlab("") + ggtitle("mGFR/IFTA - 12 months")+
  stat_compare_means( method = "anova", label.y = 190, size = 3) + 
  stat_compare_means(comparisons = my.comp, size = 3)

p.mgfr12.iifta = ggplot(new.matrix[!is.na(new.matrix$iIFTA_score_12m_meena1),], 
                       aes(x = iIFTA.12,  y = mgfr_12m)) +
  geom_boxplot(aes(fill = dgf_define ), alpha = 0.5, outlier.shape = NA) + 
  geom_jitter(position=position_jitter(width=.3, height=0), alpha = 0.5)+
  theme_bw() + ylab("GFR (ml/min)") + xlab("") + ggtitle("mGFR/iIFTA - 12 months")+
  stat_compare_means( method = "anova", label.y = 190, size = 3) + 
  stat_compare_means(comparisons = my.comp, size = 3)

g1 = ggarrange(p.cr1m,  p.egfr1m, p.cr3m,  p.egfr3m, 
               p.cr12m,  p.egfr12m,  p.mgfr3, p.mgfr12, 
               common.legend = TRUE, legend = "right", ncol = 4, nrow = 2)

g2 = p.mgfr3.ifta + p.mgfr12.ifta + p.cr60m + plot_layout(width = c(1.5, 1.5, 1))

ggarrange(g1, g2, nrow = 2)
 
tiff("figures/mgfr.plots.1.tiff", units = "cm", res = 300, height = 18, width = 36)
g1
dev.off()

tiff("figures/mgfr.plots.2.tiff", units = "cm", res = 300, height = 9, width = 36)
g2
dev.off()

```


Basic plots of renal function
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
# Remove ECD and LKD and  use DGF.HD criteria
dgf.filter1 = dgf.filter[dgf.filter$donor_criteria != "DBD + ECD",]
dgf.filter1 = dgf.filter1[dgf.filter1$donor_criteria != "DCD + ECD",]
dgf.filter1 = dgf.filter1[dgf.filter1$donor_criteria != "LKD",]
dgf.filter1 = dgf.filter1[dgf.filter1$dgf_define != "DGF.Cr",]
dgf.filter1$dgf_define = droplevels(dgf.filter1$dgf_define)

# Function split by donor criteria
    p.cr1m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_1m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) + ylab("SCr, 1m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(donor_criteria)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.cr3m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_3m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) + ylab("SCr, 3m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(donor_criteria)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.m3 = ggplot(dgf.filter1, aes(x=dgf_define, y=mgfr_3m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) + ylab("mGFR, 3m (ml/min)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(donor_criteria)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.cr12m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_12m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) + ylab("SCr, 12m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(donor_criteria)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)

    p.m12 = ggplot(dgf.filter1, aes(x=dgf_define, y=mgfr_12m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) + ylab("mGFR, 12m (ml/min)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(donor_criteria)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.cr60m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_60m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) + ylab("SCr, 60m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(donor_criteria)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.sum.function1 = ggarrange(p.cr1m, p.cr3m, p.m3, p.cr12m, p.m12, p.cr60m, nrow = 1)
    
# Function, split by early rejection 0-1
    p.cr1m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_1m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 1m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(bpar_0_1m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.cr3m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_3m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 3m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(bpar_0_1m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.m3 = ggplot(dgf.filter1, aes(x=dgf_define, y=mgfr_3m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("mGFR, 3m (ml/min)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(bpar_0_1m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
        
    p.cr12m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_12m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 12m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(bpar_0_1m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.m12 = ggplot(dgf.filter1, aes(x=dgf_define, y=mgfr_12m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("mGFR, 12m (ml/min)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(bpar_0_1m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.cr60m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_60m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 60m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(bpar_0_1m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.sum.function2 = ggarrange(p.cr1m, p.cr3m, p.m3, p.cr12m, p.m12, p.cr60m, nrow = 1)

# Function, split by early rejection up to 3mm
    p.cr1m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_1m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 1m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(bparupto03m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.cr3m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_3m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 3m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(bparupto03m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.m3 = ggplot(dgf.filter1, aes(x=dgf_define, y=mgfr_3m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("mGFR, 3m (ml/min)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(bparupto03m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
        
    p.cr12m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_12m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 12m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(bparupto03m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.m12 = ggplot(dgf.filter1, aes(x=dgf_define, y=mgfr_12m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("mGFR, 12m (ml/min)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(bparupto03m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.cr60m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_60m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 60m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(bparupto03m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.sum.function3 = ggarrange(p.cr1m, p.cr3m, p.m3, p.cr12m, p.m12, p.cr60m, nrow = 1)
    
# Function, split by early subclinical rejection
    p.cr1m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_1m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 1m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(Subclin_0_1m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.cr3m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_3m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 3m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(Subclin_0_1m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.cr12m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_12m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 12m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(Subclin_0_1m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.cr60m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_60m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 60m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(Subclin_0_1m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.m3 = ggplot(dgf.filter1, aes(x=dgf_define, y=mgfr_3m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) + ylab("mGFR, 3m (ml/min)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(Subclin_0_1m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.m12 = ggplot(dgf.filter1, aes(x=dgf_define, y=mgfr_12m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("mGFR, 12m (ml/min)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(Subclin_0_1m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.sum.function4 = ggarrange(p.cr1m, p.cr3m, p.m3, p.cr12m, p.m12, p.cr60m, nrow = 1)
    
# Function, split by early subclinical rejection
    dgf.filter1$subclin.0to3m =ifelse(dgf.filter1$Subclin_0_1m == "No Subclin" & 
                                      dgf.filter1$subclin_1m == "No Subclin" &
                                      dgf.filter1$subclin_1_3m == "No Subclin" & 
                                      dgf.filter1$subclin_3m == "No Subclin", 
                                      "No Sublcin", "Subclin 0 - 3m")
        
    p.cr1m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_1m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 1m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(subclin.0to3m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.cr3m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_3m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 3m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(subclin.0to3m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.cr12m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_12m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 12m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(subclin.0to3m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.cr60m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_60m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 60m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(subclin.0to3m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.m3 = ggplot(dgf.filter1, aes(x=dgf_define, y=mgfr_3m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("mGFR, 3m (ml/min)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(subclin.0to3m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.m12 = ggplot(dgf.filter1, aes(x=dgf_define, y=mgfr_12m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("mGFR, 12m (ml/min)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(subclin.0to3m)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.sum.function5 = ggarrange(p.cr1m, p.cr3m, p.m3, p.cr12m, p.m12, p.cr60m, nrow = 1)

# Function split by organ type
    p.cr1m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_1m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 1m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(organ_type)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.cr3m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_3m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 3m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(organ_type)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.m3 = ggplot(dgf.filter1, aes(x=dgf_define, y=mgfr_3m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("mGFR, 3m (ml/min)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(organ_type)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.cr12m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_12m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) + ylab("SCr, 12m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(organ_type)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)

    p.m12 = ggplot(dgf.filter1, aes(x=dgf_define, y=mgfr_12m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("mGFR, 12m (ml/min)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(organ_type)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.cr60m = ggplot(dgf.filter1, aes(x=dgf_define, y=cr_60m, fill = dgf_define)) +
      geom_boxplot(alpha=0.3, outlier.shape = NA) +ylab("SCr, 60m (umol/L)") + xlab("")+
      theme_bw() + facet_grid(cols = vars(organ_type)) +
      geom_jitter(color="black", size=0.3, alpha = 0.4, position=position_jitter(0.25))+
      theme(legend.position="none") + stat_compare_means(label = "p.format", vjust = 1.2)
    
    p.sum.function6 = ggarrange(p.cr1m, p.cr3m, p.m3, p.cr12m, p.m12, p.cr60m, nrow = 1)
    
# Combine plots
    tiff("figures/function.dgf.splits.tiff", res = 300, units = "cm", height = 35, width = 50)
    p.sum.function6/p.sum.function1/p.sum.function2/p.sum.function3/p.sum.function4/p.sum.function5
    dev.off()
    
```


Sankey
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
colnames(dgf.filter)

auscad.sankey(dgf.filter, FALSE, c("donor_criteria", "dgf_define"))

auscad.sankey(dgf.filter, FALSE,  c("dgf_define", "bpar_0_1m", 
                                    "bpar_1m", "bpar_3m"))

auscad.sankey(dgf.filter, FALSE,  c("dgf_define", "Subclin_0_1m", 
                                    "subclin_1m", "subclin_3m"))
  
dgf.filter1 = dgf.filter %>% 
  filter(dgf_define != "DGF.Cr") %>%
  filter(!is.na(IFTA_score_12m_meena1)) %>%
  filter(!is.na(IFTA_score_3m_meena1))

  auscad.sankey(dgf.filter1, FALSE, c("dgf_define", "IFTA_score_3m_meena1"))
  
  tiff("figures/ifta.simple.sankey.tiff", units = "cm", res = 300, height = 10, width = 14)
  auscad.sankey(dgf.filter1, FALSE, c("dgf_define",  "IFTA_score_12m_meena1"))
  dev.off()
    
  tiff("figures/ifta.simple.sankey.tiff", units = "cm", res = 300, height = 10, width = 14)
  auscad.sankey(dgf.filter1, FALSE, c("dgf_define", "subclin_any1" , "bpar_any1","IFTA_score_12m_meena1"))
  dev.off()
  
  tiff("figures/ifta.simple.sankey.1.tiff", units = "cm", res = 300, height = 10, width = 14)
  auscad.sankey(dgf.filter1, FALSE, c("dgf_define", "bpar_any1",  "IFTA_score_12m_meena1"))
  dev.off()
  
  tiff("figures/ifta.simple.sankey.2.tiff", units = "cm", res = 300, height = 10, width = 14)
  auscad.sankey(dgf.filter1, FALSE, c("dgf_define", "bkvan_any1",  "IFTA_score_12m_meena1"))
  dev.off()

  dgf.filter1$mgfr_12m1 = as.factor(dgf.filter1$mgfr_12m1)
  dgf.filter1$mgfr_12m1 = factor(dgf.filter1$mgfr_12m1, levels = c("<30", "30-45", "45-60", ">60"))
  levels(dgf.filter1$mgfr_12m1)
  
  
  auscad.sankey(dgf.filter1, FALSE, c("dgf_define", "IFTA_score_3m_meena1", 
                                    "IFTA_score_12m_meena1"))
  
  auscad.sankey(dgf.filter1, FALSE, c("dgf_define", "iIFTA_score_3m_meena1", 
                                    "iIFTA_score_12m_meena1"))
  
dgf.filter1 = dgf.filter %>% 
  filter(!is.na(IFTA_score_12m_meena1)) %>%
  filter(!is.na(IFTA_score_3m_meena1)) %>%
  filter(dgf_define != "DGF.Cr") %>%
  filter(organ_type == "Kidney Transplant")  %>% # deceased donor kidney only tx
  select(dgf_define, organ_type, IFTA_score_3m_meena1, IFTA_score_12m_meena1)

  auscad.sankey(dgf.filter1, FALSE, c("dgf_define", "IFTA_score_3m_meena1"))
  
  auscad.sankey(dgf.filter1, FALSE, c("dgf_define", 
                                    "IFTA_score_12m_meena1"))
    
  auscad.sankey(dgf.filter1, FALSE, c("dgf_define", "IFTA_score_3m_meena1", 
                                    "IFTA_score_12m_meena1"))

  
  
dgf.filter1 = dgf.filter1%>%make_long(c("dgf_define", "IFTA_score_3m_meena1", 
                                    "IFTA_score_12m_meena1"))
  
```


Alluvial plots
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
  ###########
  # ALLUVIAL
  dgf.filter1 = dgf.filter %>% 
  filter(dgf_define!= "DGF.Cr") %>%
  select(dgf_define, organ_type, 
         donor_criteria, bkvan_any1,
         bpar_any1, Subclin_0_1m,
         iIFTA_score_3m_meena1,
         IFTA_score_12m_meena1)

  dgf.filter1$dgf_define = droplevels(dgf.filter1$dgf_define)
  
  # dgf.filter1 = dgf.filter1[dgf.filter1$dgf_define != "DGF.Cr", ]
  # dgf.filter1$dgf_define = droplevels(dgf.filter1$dgf_define)

  # dgf.filter1 = dgf.filter1[dgf.filter1$donor_criteria != "LKD", ]
  # dgf.filter1$donor_criteria = droplevels(dgf.filter1$donor_criteria)
  
  dgf.filter1 = data.frame(ftable(dgf.filter1))

  # Plots
  p1 = ggplot(dgf.filter1,
       aes(axis1 = dgf_define,
           axis2 = Subclin_0_1m,,
           axis3 = bpar_any1,
           axis4 = IFTA_score_12m_meena1,
           y = Freq)) +
  geom_alluvium(aes(fill = dgf_define),
                curve_type = "cubic",
                na.rm = TRUE) +
  geom_stratum() + 
    geom_text(stat = "stratum", 
              aes(label = after_stat(stratum))) +
  scale_x_discrete(limits = c("DGF criteria", 
                              "Subclin.early",
                              "BPAR.any",
                              "IFTA12m"), expand = c(.1, .1)) +
  labs(title = "Auscad",
       subtitle = "stratified by organ_type, and IFTA12m",
       y = "Proportion") +
  theme_minimal() + scale_colour_brewer() +
  theme(legend.position = "bottom") 
  
tiff("figures/ifta.alluvial.1.tiff", units = "cm", res = 300, height = 16, width = 36)
  p1
dev.off()






dgf.filter1 = dgf.filter %>% 
  filter(dgf_define!= "DGF.Cr") %>%
  select(dgf_define, organ_type, 
         donor_criteria, bkvan_any1,
         bpar_any1, Subclin_0_1m,
         iIFTA_score_3m_meena1,
         IFTA_score_12m_meena1)
  
dgf.filter1$dgf_define = droplevels(dgf.filter1$dgf_define)

dgf.filter1 = dgf.filter1%>%
  mutate(dgf_define = 
           case_when(dgf_define == "Control" & Subclin_0_1m == "No Subclin" ~ "Control",
                     dgf_define == "Control" & Subclin_0_1m != "No Subclin" ~ "Control.S",
                     dgf_define == "DGF.HD" & Subclin_0_1m == "No Subclin" ~ "DGF",
                     dgf_define == "DGF.HD" & Subclin_0_1m != "No Subclin" ~ "DGF.S"))

  
  # dgf.filter1 = dgf.filter1[dgf.filter1$dgf_define != "DGF.Cr", ]
  # dgf.filter1$dgf_define = droplevels(dgf.filter1$dgf_define)

  # dgf.filter1 = dgf.filter1[dgf.filter1$donor_criteria != "LKD", ]
  # dgf.filter1$donor_criteria = droplevels(dgf.filter1$donor_criteria)
  
  dgf.filter1 = data.frame(ftable(dgf.filter1))

  # Plots
  p1 = ggplot(dgf.filter1,
       aes(axis1 = dgf_define,
           axis2 = bpar_any1,
           axis4 = IFTA_score_12m_meena1,
           y = Freq)) +
  geom_alluvium(aes(fill = dgf_define),
                curve_type = "cubic",
                na.rm = TRUE) +
  geom_stratum() + 
    geom_text(stat = "stratum", 
              aes(label = after_stat(stratum))) +
  scale_x_discrete(limits = c("DGF(S=subclin_0_1m)", 
                               "BPAR",
                              "IFTA.12m"), expand = c(.1, .1)) +
  labs(title = "Auscad",
       subtitle = "stratified by organ_type, and IFTA12m",
       y = "Proportion") +
  theme_minimal() + scale_colour_brewer() +
  theme(legend.position = "right") 
  
tiff("figures/ifta.alluvial.1.tiff", units = "cm", res = 300, height = 20, width = 40)
  p1
dev.off()


```


### Baseline characteristics 
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
base.char = c()
base.char = dgf.filter %>% 
  select(dgf_define, dgf_define.comb, dgf.severity,
         organ_type, HLA_mm, dsa1_pre, dsa2_pre,dcd_wit_min, swit, 
         donor_criteria, donor_gender,  donor_age, donor_inotropes,
         cit_min, donor_terminal_cr, gender, tx_age, age, chol_pre,
         htn_pre, dm_pre, ihd_pre, induct_extra, ondialysis_pretx, renal_dx,
         graft_age_months, graft_number)

base.char = base.char[base.char$dgf_define != "Control", ]

### Choose one
baseline.table = tbl_summary(base.char,  missing = "no", by = dgf_define)
baseline.table = tbl_summary(base.char,  missing = "no", by = dgf_define.comb)
baseline.table = tbl_summary(base.char,  missing = "no", by = dgf.severity)

baseline.table1 = baseline.table %>% add_n() %>% 
  add_p() %>% modify_header(label = "Baseline") %>% bold_labels()

baseline.table1

# SGF
base.char = c()
base.char = dgf.filter %>% 
  select(dgf_define, dgf_define.comb, dgf.severity,
         organ_type, HLA_mm, dsa1_pre, dsa2_pre,
         donor_criteria, donor_gender,  donor_age, donor_inotropes,
         cit_min, donor_terminal_cr, gender, tx_age, age, chol_pre,
         htn_pre, dm_pre, ihd_pre, induct_extra, ondialysis_pretx, renal_dx,
         graft_age_months, graft_number)

levels(base.char$dgf_define)
base.char = base.char[base.char$dgf_define != "DGF.HD",]
base.char$dgf_define = droplevels(base.char$dgf_define)
levels(base.char$dgf_define)

baseline.table = tbl_summary(base.char,  missing = "no", by = dgf_define)

baseline.table1 = baseline.table %>% add_n() %>% 
  add_p() %>% modify_header(label = "Baseline") %>% bold_labels()

baseline.table1


# excluding living donor
base.char = c()
base.char = dgf.filter %>% 
  filter(organ_type != "Living Kidney Transplant") %>%
  select(dgf_define, dgf_define.comb, dgf.severity,
         organ_type, HLA_mm, dsa1_pre, dsa2_pre,
         donor_criteria, donor_gender,  donor_age, donor_inotropes,
         cit_min, donor_terminal_cr, gender, tx_age, age, chol_pre,
         htn_pre, dm_pre, ihd_pre, induct_extra, ondialysis_pretx, renal_dx,
         graft_age_months, graft_number)
levels(base.char$organ_type)
base.char = base.char[base.char$organ_type != "Living Kidney Transplant",]
base.char$organ_type = droplevels(base.char$organ_type)
levels(base.char$organ_type)

levels(base.char$donor_criteria)
base.char = base.char[base.char$donor_criteria != "LKD",]

levels(base.char$dgf_define)
base.char = base.char[base.char$dgf_define != "DGF.Cr",]
base.char$dgf_define = droplevels(base.char$dgf_define)
levels(base.char$dgf_define)

baseline.table = tbl_summary(base.char,  missing = "no", by = dgf_define)
baseline.table1 = baseline.table %>% add_n() %>% 
  add_p() %>% modify_header(label = "Baseline") %>% bold_labels()

baseline.table1
```



### EDA clinical variables. 

FAMD for continuous and categorical variables (avoid dependent variables)
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
dgf.filter.mca = c()
dgf.filter.mca = dgf.filter[dgf.filter$time.in.study>12,]
dgf.filter.mca$tx_yr.2016 = ifelse(dgf.filter.mca$tx_yr <= 2016, "pre-2016", "post-2016")
dgf.filter.mca$tx_yr.2016 = as.factor(dgf.filter.mca$tx_yr.2016)
dgf.filter.mca = dgf.filter.mca %>% 
  mutate(tx_age = case_when(tx_age <=40 ~ "<40 yo",
                            tx_age > 40 & tx_age <= 60 ~ "40-60 yo",
                            tx_age > 60 ~ ">60yo"))
dgf.filter.mca$tx_age = as.factor(dgf.filter.mca$tx_age)

dgf.filter.mca = dgf.filter.mca %>% 
  mutate(cit_min = case_when(cit_min <=360 ~ "<6hr",
                            cit_min > 360 & cit_min <= 600 ~ "6-12hr",
                            cit_min > 600 ~ ">12hr"))
dgf.filter.mca$cit_min = as.factor(dgf.filter.mca$cit_min)

dgf.filter.mca = dgf.filter.mca %>% 
  mutate(bpar_any = case_when(bpar_any == 0 ~ "No BPAR",
                            bpar_any == 1 ~ "BPAR any"))
dgf.filter.mca$bpar_any = as.factor(dgf.filter.mca$bpar_any)

dgf.filter.mca = dgf.filter.mca %>% 
  mutate(subclin_any = case_when(subclin_any == 0 ~ "No BPAR",
                            subclin_any == 1 ~ "BPAR any"))
dgf.filter.mca$subclin_any = as.factor(dgf.filter.mca$subclin_any)


# Pick and fix variables
cat.var = c("dgf_define","dgf.severity", "gender", 
            "tx_age", "induct_extra", 
            "renal_dx", "graft_loss", "organ_type", 
            "death", "graft_number",
            "donor_ionotropes", 
            "tx_yr.2016", "donor_criteria", 
            "cit_min", "bkvan_any1", 
            "dsa1_pre", "dsa2_pre","donor_gender",  
            "htn_pre", "ihd_pre", 
            "bpar_any", "subclin_any",
            "bpar_0_1m", "Subclin_0_1m", 
            "bpar_1m", "subclin_1m" , 
            "bpar_1_3m", "subclin_1_3m", 
            "bpar_3m", "subclin_3m",
            "bpar_3_12m", "subclin_3_12m", 
            "bpar_12m", "subclin_12m",
            "bpar_post12m", "subclin_post12m",
            "resist_cmv", "ebv_infection", "bk_viremia", 
            "bkvan", "nodat", "cvd_0_5y", "skin_ca", 
            "fungal_invasive_orblood","recurrent_uti",  
            "gasteroenteritis", "chest_infection", "PTLD", 
            "transfusion_01m", "wound_infection_01m")

cont.var = c("donor_terminal_cr", "HLA_mm", 
             "egfr_1m", "egfr_3m", "egfr_12m", 
              "tac_d2", "tac_d7",  
             "tac_d14", "tac_d21", "tac_1m", "tac_3m",
             "tac_12m",   "ci_0m", "ci_1m", "ci_3m", "ci_12m", 
             "c4d_0m", "c4d_1m", "c4d_3m", "c4d_12m", 
             "ct_0m", "ct_1m", "ct_3m", "ct_12m", 
             "cg_0m", "cg_1m", "cg_3m", "cg_12m",
             "iIFTA_1m", "iIFTA_3m", "iIFTA_12m")

mca.var = c(cat.var, cont.var)
dgf.filter.mca = dgf.filter.mca[,mca.var]

# Analysis - FAMD
res.famd = FAMD(dgf.filter.mca, graph = TRUE)
  
  fviz_eig(res.famd)
  fviz_contrib(res.famd, "var", axes = 1)
  fviz_contrib(res.famd, "var", axes = 2)
  fviz_contrib(res.famd, "var", axes = 3)
  fviz_contrib(res.famd, "var", axes = 4)
  fviz_contrib(res.famd, "var", axes = 5)
  fviz_contrib(res.famd, "var", axes = 6)
  
  fviz_famd_var(res.famd, repel = TRUE)
  fviz_famd_var(res.famd, repel = TRUE, col.var = "contrib", 
                gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))
  fviz_famd_var(res.famd, repel = TRUE, col.var = "cos2", 
                gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))

  # Graph of individuals
  fviz_pca_ind(res.famd, addEllipses = TRUE, habillage = "dgf_define", 
               ellipse.level = 0.8, ellipse.type = "t")
  
  fviz_pca_var(res.famd, col.var = "black")

  # Graph individuals and groups
  fviz_mfa_ind(res.famd, repel = TRUE, ellipse.level = 0.85,
               habillage = "dgf_define", # color by groups 
               label.rectangle = TRUE, geom = "point",
               palette = c("#00AFBB", "#E7B800", "#FC4E07"),
               addEllipses = TRUE,  ellipse.type = "t") 
  
  fviz_mfa_ind(res.famd, repel = TRUE ,ellipse.level = 0.85,
               habillage = "organ_type", # color by groups 
               palette = c("#00AFBB", "#E7B800", "#FC4E07"),
               addEllipses = TRUE, ellipse.type = "t") 
  
  fviz_mfa_ind(res.famd, repel = TRUE ,ellipse.level = 0.85,
               habillage = "donor_criteria", # color by groups 
               addEllipses = TRUE, ellipse.type = "t") 

  tiff("figures/famd.clinical.1.tiff", res = 300, units = "cm", height = 20, width = 30)
   fviz_ellipses(res.famd, geom = "point",
                ellipse.type = "t", ellipse.level = 0.7, 
                c("dgf_define", "dgf.severity", 
                "donor_criteria", "organ_type", "donor_ionotropes", 
                "death"))
  dev.off()

  tiff("figures/famd.clinical.2.tiff", res = 300, units = "cm", height = 20, width = 30)
    fviz_ellipses(res.famd, ellipse.type = "t", geom = "point", ellipse.level = 0.7,
              c("tx_age", "renal_dx", "induct_extra", "HLA_mm",
                "tx_yr.2016", "cit_min", "graft_number", 
                "graft_loss", "bkvan_any1"))
  dev.off()
  
  tiff("figures/famd.clinical.3.tiff", res = 300, units = "cm", height = 20, width = 30)
    fviz_ellipses(res.famd, ellipse.type = "t", geom = "point", ellipse.level = 0.7,
              c("bpar_0_1m", "Subclin_0_1m", 
                "bpar_1m", "subclin_1m" ,  "bpar_1_3m", "subclin_1_3m", 
                "bpar_3m", "subclin_3m",   "bpar_3_12m", "subclin_3_12m",
                "bpar_12m", "subclin_12m"))
  dev.off()
 

p1 =fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "dgf_define",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("DGF criteria")


p2 = fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "donor_criteria",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("Donor criteria")

p3 = fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "organ_type",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("Transplant type")

p4 = fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "dgf.severity",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("DGF severity")

p5 =fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "donor_ionotropes",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("Donor Inotrope use")

p6 =fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "cit_min",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("Cold Ischemic Time")


tiff("figures/FAMD.clinical.summary.tiff", res = 300, units = "cm", height = 20, width = 36)
ggarrange(p1, p2, p3, p4, p5,p6, nrow = 2, ncol = 3)
dev.off()

  
# Analysis: unsupervised clustering
# res.hcpc = HCPC(res.famd, nb.clust=-1, graph = TRUE)
# 
# res.hcpc$desc.var$quanti
# 
# fviz_cluster(res.hcpc, geom = "point")
# 
# fviz_dend(res.hcpc, 
#           cex = 0.7,                     # Label size
#           palette = "jco",               # Color palette see ?ggpubr::ggpar
#           rect = TRUE, rect_fill = TRUE, # Add rectangle around groups
#           rect_border = "jco",           # Rectangle color
#           labels_track_height = 0.8      # Augment the room for labels
#           )
# 
# fviz_cluster(res.hcpc,
#              repel = TRUE,            # Avoid label overlapping
#              show.clust.cent = TRUE, # Show cluster centers
#              palette = "jco",         # Color palette see ?ggpubr::ggpar
#              ggtheme = theme_minimal(),
#              main = "Factor map"
#              )
  
  tiff("figures/FAMD.clinical.4.tiff", res = 300, units = "cm", height = 18, width = 18)
  plot.FAMD(res.famd, choix="quali",new.plot=TRUE)
  dev.off()
  
  tiff("figures/FAMD.clinical.5.tiff", res = 300, units = "cm", height = 15, width = 15)
  fviz_famd_var(res.famd, repel = TRUE, axes = c(1, 2), 
                col.var = "contrib", gradient.cols = "nejm") + ggtitle("axes 1-2")
  dev.off()
  
  tiff("figures/FAMD.clinical.6.tiff", res = 300, units = "cm", height = 15, width = 15)
  fviz_famd_var(res.famd, repel = TRUE, axes = c(3, 4), 
                col.var = "contrib", gradient.cols = "nejm") + ggtitle("axes 3-4")
  dev.off()
  
  tiff("figures/FAMD.clinical.7.tiff", res = 300, units = "cm", height = 15, width = 15)
  fviz_famd_var(res.famd, repel = TRUE, axes = c(4, 5), 
                col.var = "contrib", gradient.cols = "nejm") + ggtitle("axes 4-5")
  dev.off()
  

```


FAMD for continuous and categorical variables (avoid dependent variables) - exclude living donors
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
dgf.filter.mca = c()
dgf.filter.mca = dgf.filter[dgf.filter$time.in.study>12,]
dgf.filter.mca = dgf.filter.mca[dgf.filter.mca$organ_type != "Living Kidney Transplant",]

levels(dgf.filter.mca$donor_criteria)
dgf.filter.mca = dgf.filter.mca[dgf.filter.mca$donor_criteria != "LKD",]
dgf.filter.mca$donor_criteria = droplevels(dgf.filter.mca$donor_criteria)
levels(dgf.filter.mca$donor_criteria)

dgf.filter.mca$tx_yr.2016 = ifelse(dgf.filter.mca$tx_yr <= 2016, "pre-2016", "post-2016")
dgf.filter.mca$tx_yr.2016 = as.factor(dgf.filter.mca$tx_yr.2016)

dgf.filter.mca = dgf.filter.mca %>% 
  mutate(tx_age = case_when(tx_age <=40 ~ "<40 yo",
                            tx_age > 40 & tx_age <= 60 ~ "40-60 yo",
                            tx_age > 60 ~ ">60yo"))
dgf.filter.mca$tx_age = as.factor(dgf.filter.mca$tx_age)

dgf.filter.mca = dgf.filter.mca %>% 
  mutate(cit_min = case_when(cit_min <=360 ~ "<6hr",
                            cit_min > 360 & cit_min <= 600 ~ "6-12hr",
                            cit_min > 600 ~ ">12hr"))
dgf.filter.mca$cit_min = as.factor(dgf.filter.mca$cit_min)

dgf.filter.mca = dgf.filter.mca %>% 
  mutate(bpar_any = case_when(bpar_any == 0 ~ "No BPAR",
                            bpar_any == 1 ~ "BPAR any"))
dgf.filter.mca$bpar_any = as.factor(dgf.filter.mca$bpar_any)

dgf.filter.mca = dgf.filter.mca %>% 
  mutate(subclin_any = case_when(subclin_any == 0 ~ "No BPAR",
                            subclin_any == 1 ~ "BPAR any"))
dgf.filter.mca$subclin_any = as.factor(dgf.filter.mca$subclin_any)


# Pick and fix variables
cat.var = c("dgf_define","dgf.severity", "gender", 
            "tx_age", "induct_extra", 
            "renal_dx", "graft_loss", "organ_type", 
            "death", "graft_number",
            "donor_ionotropes", 
            "tx_yr.2016", "donor_criteria", 
            "cit_min", "bkvan_any1", 
            "dsa1_pre", "dsa2_pre","donor_gender",  
            "htn_pre", "ihd_pre", 
            "bpar_any", "subclin_any",
            "bpar_0_1m", "Subclin_0_1m", 
            "bpar_1m", "subclin_1m" , 
            "bpar_1_3m", "subclin_1_3m", 
            "bpar_3m", "subclin_3m",
            "bpar_3_12m", "subclin_3_12m", 
            "bpar_12m", "subclin_12m",
            "bpar_post12m", "subclin_post12m",
            "resist_cmv", "ebv_infection", "bk_viremia", 
            "bkvan", "nodat", "cvd_0_5y", "skin_ca", 
            "fungal_invasive_orblood","recurrent_uti",  
            "gasteroenteritis", "chest_infection", "PTLD", 
            "transfusion_01m", "wound_infection_01m")

cont.var = c("donor_terminal_cr", "HLA_mm", 
             "egfr_1m", "egfr_3m", "egfr_12m", 
              "tac_d2", "tac_d7",  
             "tac_d14", "tac_d21", "tac_1m", "tac_3m",
             "tac_12m",   "ci_0m", "ci_1m", "ci_3m", "ci_12m", 
             "c4d_0m", "c4d_1m", "c4d_3m", "c4d_12m", 
             "ct_0m", "ct_1m", "ct_3m", "ct_12m", 
             "cg_0m", "cg_1m", "cg_3m", "cg_12m",
             "iIFTA_1m", "iIFTA_3m", "iIFTA_12m")

mca.var = c(cat.var, cont.var)
dgf.filter.mca = dgf.filter.mca[,mca.var]

# Analysis - FAMD
res.famd = FAMD(dgf.filter.mca, graph = TRUE)
  
  fviz_eig(res.famd)
  fviz_contrib(res.famd, "var", axes = 1)
  fviz_contrib(res.famd, "var", axes = 2)
  
  fviz_famd_var(res.famd, repel = TRUE)
  fviz_famd_var(res.famd, repel = TRUE, col.var = "contrib", 
                gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))
  fviz_famd_var(res.famd, repel = TRUE, col.var = "cos2", 
                gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))

  # Graph of individuals
  fviz_pca_ind(res.famd, addEllipses = TRUE, habillage = "dgf_define", 
               ellipse.level = 0.8, ellipse.type = "t")
  
  fviz_pca_var(res.famd, col.var = "black")

  # Graph individuals and groups
  fviz_mfa_ind(res.famd, repel = TRUE, ellipse.level = 0.85,
               habillage = "dgf_define", # color by groups 
               label.rectangle = TRUE, geom = "point",
               palette = c("#00AFBB", "#E7B800", "#FC4E07"),
               addEllipses = TRUE,  ellipse.type = "t") 
  
  fviz_mfa_ind(res.famd, repel = TRUE ,ellipse.level = 0.85,
               habillage = "organ_type", # color by groups 
               palette = c("#00AFBB", "#E7B800", "#FC4E07"),
               addEllipses = TRUE, ellipse.type = "t") 
  
  fviz_mfa_ind(res.famd, repel = TRUE ,ellipse.level = 0.85,
               habillage = "donor_criteria", # color by groups 
               addEllipses = TRUE, ellipse.type = "t") 

  tiff("figures/famd.clinical.1.dktx.tiff", res = 300, units = "cm", height = 20, width = 30)
   fviz_ellipses(res.famd, geom = "point",
                ellipse.type = "t", ellipse.level = 0.7, 
                c("dgf_define", "dgf.severity", 
                "donor_criteria", "organ_type", "donor_ionotropes", 
                "death"))
  dev.off()

  tiff("figures/famd.clinical.2.dktx.tiff", res = 300, units = "cm", height = 24, width = 30)
    fviz_ellipses(res.famd, ellipse.type = "t", geom = "point", ellipse.level = 0.7,
              c("tx_age", "renal_dx", "induct_extra", "HLA_mm",
                "tx_yr.2016", "cit_min", "graft_number", 
                "graft_loss", "bkvan_any1"), ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=3, byrow=TRUE)) + ggtitle("other variables")
  dev.off()
  
  tiff("figures/famd.clinical.3.dktx.tiff", res = 300, units = "cm", height = 20, width = 24)
    fviz_ellipses(res.famd, ellipse.type = "t", geom = "point", ellipse.level = 0.7,
              c("bpar_0_1m", "Subclin_0_1m", 
                "bpar_1m", "subclin_1m" ,  "bpar_1_3m", "subclin_1_3m", 
                "bpar_3m", "subclin_3m",   "bpar_3_12m", "subclin_3_12m",
                "bpar_12m", "subclin_12m"), ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("other variables")
  dev.off()
  
  tiff("figures/famd.clinical.4.dktx.tiff", res = 300, units = "cm", height = 30, width = 38)
    fviz_ellipses(res.famd, ellipse.type = "t", geom = "point", ellipse.level = 0.7,
              c("tx_age", "renal_dx", "induct_extra", "HLA_mm","transfusion_01m",
                "tx_yr.2016", "cit_min", "graft_number", "dsa1_pre", "dsa2_pre", 
                "graft_loss", "bkvan_any1", "bpar_0_1m", "Subclin_0_1m", "nodat",
                "bpar_1m", "subclin_1m" ,  "bpar_1_3m", "subclin_1_3m", 
                "bpar_3m", "subclin_3m",   "bpar_3_12m", "subclin_3_12m",
                "bpar_12m", "subclin_12m"), ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=4, byrow=TRUE)) + ggtitle("other variables")
  dev.off()
 
 

p1 =fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "dgf_define",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("DGF criteria")


p2 = fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "donor_criteria",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("Donor criteria")

p3 = fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "organ_type",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("Transplant type")

p4 = fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "dgf.severity",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("DGF severity")

p5 =fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "donor_ionotropes",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("Donor Inotrope use")

p6 =fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "cit_min",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("Cold Ischemic Time")


tiff("figures/FAMD.clinical.summary.dktx.tiff", res = 300, units = "cm", height = 20, width = 36)
ggarrange(p1, p2, p3, p4, p5,p6, nrow = 2, ncol = 3)
dev.off()

  
  fviz_famd_var(res.famd, repel = TRUE, col.var = "contrib", 
                 select.var = list(contrib = 15),
                gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))

  
# Analysis: unsupervised clustering
# res.hcpc = HCPC(res.famd, nb.clust=-1, graph = TRUE)
# 
# res.hcpc$desc.var$quanti
# 
# fviz_cluster(res.hcpc, geom = "point")
# 
# fviz_dend(res.hcpc, 
#           cex = 0.7,                     # Label size
#           palette = "jco",               # Color palette see ?ggpubr::ggpar
#           rect = TRUE, rect_fill = TRUE, # Add rectangle around groups
#           rect_border = "jco",           # Rectangle color
#           labels_track_height = 0.8      # Augment the room for labels
#           )
# 
# fviz_cluster(res.hcpc,
#              repel = TRUE,            # Avoid label overlapping
#              show.clust.cent = TRUE, # Show cluster centers
#              palette = "jco",         # Color palette see ?ggpubr::ggpar
#              ggtheme = theme_minimal(),
#              main = "Factor map"
#              )

```


FAMD for continuous and categorical variables (avoid dependent variables) - exclude living donors/SPK
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
dgf.filter.mca = c()
dgf.filter.mca = dgf.filter[dgf.filter$time.in.study>12,]
dgf.filter.mca = dgf.filter.mca[dgf.filter.mca$organ_type == "Kidney Transplant",]

levels(dgf.filter.mca$donor_criteria)
dgf.filter.mca = dgf.filter.mca[dgf.filter.mca$donor_criteria != "LKD",]
dgf.filter.mca$donor_criteria = droplevels(dgf.filter.mca$donor_criteria)
levels(dgf.filter.mca$donor_criteria)

dgf.filter.mca$tx_yr.2016 = ifelse(dgf.filter.mca$tx_yr <= 2016, "pre-2016", "post-2016")
dgf.filter.mca$tx_yr.2016 = as.factor(dgf.filter.mca$tx_yr.2016)

dgf.filter.mca = dgf.filter.mca %>% 
  mutate(tx_age = case_when(tx_age <=40 ~ "<40 yo",
                            tx_age > 40 & tx_age <= 60 ~ "40-60 yo",
                            tx_age > 60 ~ ">60yo"))
dgf.filter.mca$tx_age = as.factor(dgf.filter.mca$tx_age)

dgf.filter.mca = dgf.filter.mca %>% 
  mutate(cit_min = case_when(cit_min <=360 ~ "<6hr",
                            cit_min > 360 & cit_min <= 600 ~ "6-12hr",
                            cit_min > 600 ~ ">12hr"))
dgf.filter.mca$cit_min = as.factor(dgf.filter.mca$cit_min)

dgf.filter.mca = dgf.filter.mca %>% 
  mutate(bpar_any = case_when(bpar_any == 0 ~ "No BPAR",
                            bpar_any == 1 ~ "BPAR any"))
dgf.filter.mca$bpar_any = as.factor(dgf.filter.mca$bpar_any)

dgf.filter.mca = dgf.filter.mca %>% 
  mutate(subclin_any = case_when(subclin_any == 0 ~ "No BPAR",
                            subclin_any == 1 ~ "BPAR any"))
dgf.filter.mca$subclin_any = as.factor(dgf.filter.mca$subclin_any)


# Pick and fix variables
cat.var = c("dgf_define","dgf.severity", "gender", 
            "tx_age", "induct_extra", 
            "renal_dx", "graft_loss", "organ_type", 
            "death", "graft_number",
            "donor_ionotropes", 
            "tx_yr.2016", "donor_criteria", 
            "cit_min", "bkvan_any1", 
            "dsa1_pre", "dsa2_pre","donor_gender",  
            "htn_pre", "ihd_pre", 
            "bpar_any", "subclin_any",
            "bpar_0_1m", "Subclin_0_1m", 
            "bpar_1m", "subclin_1m" , 
            "bpar_1_3m", "subclin_1_3m", 
            "bpar_3m", "subclin_3m",
            "bpar_3_12m", "subclin_3_12m", 
            "bpar_12m", "subclin_12m",
            "bpar_post12m", "subclin_post12m",
            "resist_cmv", "ebv_infection", "bk_viremia", 
            "bkvan", "nodat", "cvd_0_5y", "skin_ca", 
            "fungal_invasive_orblood","recurrent_uti",  
            "gasteroenteritis", "chest_infection", "PTLD", 
            "transfusion_01m", "wound_infection_01m")

cont.var = c("donor_terminal_cr", "HLA_mm", 
             "egfr_1m", "egfr_3m", "egfr_12m", 
              "tac_d2", "tac_d7",  
             "tac_d14", "tac_d21", "tac_1m", "tac_3m",
             "tac_12m",   "ci_0m", "ci_1m", "ci_3m", "ci_12m", 
             "c4d_0m", "c4d_1m", "c4d_3m", "c4d_12m", 
             "ct_0m", "ct_1m", "ct_3m", "ct_12m", 
             "cg_0m", "cg_1m", "cg_3m", "cg_12m",
             "iIFTA_1m", "iIFTA_3m", "iIFTA_12m")

mca.var = c(cat.var, cont.var)
dgf.filter.mca = dgf.filter.mca[,mca.var]

# Analysis - FAMD
res.famd = FAMD(dgf.filter.mca, graph = TRUE)
  
  fviz_eig(res.famd)
  fviz_contrib(res.famd, "var", axes = 1)
  fviz_contrib(res.famd, "var", axes = 2)
  
  fviz_famd_var(res.famd, repel = TRUE)
  fviz_famd_var(res.famd, repel = TRUE, col.var = "contrib", 
                gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))
  fviz_famd_var(res.famd, repel = TRUE, col.var = "cos2", 
                gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))

  # Graph of individuals
  fviz_pca_ind(res.famd, addEllipses = TRUE, habillage = "dgf_define", 
               ellipse.level = 0.8, ellipse.type = "t")
  
  fviz_pca_var(res.famd, col.var = "black")

  # Graph individuals and groups
  fviz_mfa_ind(res.famd, repel = TRUE, ellipse.level = 0.85,
               habillage = "dgf_define", # color by groups 
               label.rectangle = TRUE, geom = "point",
               palette = c("#00AFBB", "#E7B800", "#FC4E07"),
               addEllipses = TRUE,  ellipse.type = "t") 
  
  fviz_mfa_ind(res.famd, repel = TRUE ,ellipse.level = 0.85,
               habillage = "organ_type", # color by groups 
               palette = c("#00AFBB", "#E7B800", "#FC4E07"),
               addEllipses = TRUE, ellipse.type = "t") 
  
  fviz_mfa_ind(res.famd, repel = TRUE ,ellipse.level = 0.85,
               habillage = "donor_criteria", # color by groups 
               addEllipses = TRUE, ellipse.type = "t") 

  tiff("figures/famd.clinical.1.dktx.k.tiff", res = 300, units = "cm", height = 20, width = 30)
   fviz_ellipses(res.famd, geom = "point",
                ellipse.type = "t", ellipse.level = 0.7, 
                c("dgf_define", "dgf.severity", 
                "donor_criteria", "donor_ionotropes", 
                "death"))
  dev.off()

  tiff("figures/famd.clinical.2.dktx.k.tiff", res = 300, units = "cm", height = 20, width = 30)
    fviz_ellipses(res.famd, ellipse.type = "t", geom = "point", ellipse.level = 0.7,
              c("tx_age", "renal_dx", "induct_extra", "HLA_mm",
                "tx_yr.2016", "cit_min", "graft_number", 
                "graft_loss", "bkvan_any1"))
  dev.off()
  
  tiff("figures/famd.clinical.3.dktx.k.tiff", res = 300, units = "cm", height = 20, width = 30)
    fviz_ellipses(res.famd, ellipse.type = "t", geom = "point", ellipse.level = 0.7,
              c("bpar_0_1m", "Subclin_0_1m", 
                "bpar_1m", "subclin_1m" ,  "bpar_1_3m", "subclin_1_3m", 
                "bpar_3m", "subclin_3m",   "bpar_3_12m", "subclin_3_12m",
                "bpar_12m", "subclin_12m"))
  dev.off()
 

p1 =fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "dgf_define",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("DGF criteria")


p2 = fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "donor_criteria",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("Donor criteria")

p3 = fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "tx_yr.2016",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("tx_yr.2016")

p4 = fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "dgf.severity",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("DGF severity")

p5 =fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "donor_ionotropes",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("Donor Inotrope use")

p6 =fviz_mfa_ind(res.famd, repel = TRUE, 
                 addEllipses = TRUE, ellipse.level = 0.8, ellipse.type = "t",
                 habillage = "cit_min",geom = "point", ggtheme = theme_minimal()) +
  theme(legend.position = "bottom", legend.text = element_text(size=8)) +   
  guides(color=guide_legend(nrow=2, byrow=TRUE)) + ggtitle("Cold Ischemic Time")


tiff("figures/FAMD.clinical.summary.dktx.k.tiff", res = 300, units = "cm", height = 20, width = 36)
ggarrange(p1, p2, p3, p4, p5,p6, nrow = 2, ncol = 3)
dev.off()

  
  fviz_famd_var(res.famd, repel = TRUE, col.var = "contrib", 
                 select.var = list(contrib = 20),
                gradient.cols = c("#00AFBB", "#E7B800", "#FC4E07"))


  
# Analysis: unsupervised clustering
# res.hcpc = HCPC(res.famd, nb.clust=-1, graph = TRUE)
# 
# res.hcpc$desc.var$quanti
# 
# fviz_cluster(res.hcpc, geom = "point")
# 
# fviz_dend(res.hcpc, 
#           cex = 0.7,                     # Label size
#           palette = "jco",               # Color palette see ?ggpubr::ggpar
#           rect = TRUE, rect_fill = TRUE, # Add rectangle around groups
#           rect_border = "jco",           # Rectangle color
#           labels_track_height = 0.8      # Augment the room for labels
#           )
# 
# fviz_cluster(res.hcpc,
#              repel = TRUE,            # Avoid label overlapping
#              show.clust.cent = TRUE, # Show cluster centers
#              palette = "jco",         # Color palette see ?ggpubr::ggpar
#              ggtheme = theme_minimal(),
#              main = "Factor map"
#              )

```



### Outcome summaries

All transplants
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
outcomes = c()
outcomes = dgf.filter %>% select(dgf_define, dgf_define.comb,  dgf.severity, 
                                 event_any, event_all,  death, primary_nonfunc,
                                 death.cs.graftloss, death_functioninggraft, 
                                 graft_loss,  censor_followup,return_OT_01m,  transfusion_01m, 
                                 wound_infection_01m, cr_pre, cr_0hr, cr_d1, 
                                 cr_d2, cr_1m, cr_3m, cr_12m, cr_24m,cr_36m, 
                                 cr_48m, cr_60m, cr_72m, cr_84m,cr_96m, egfr_1m,  
                                 egfr_3m, egfr_12m, egfr_24m,  egfr_36m, 
                                 egfr_48m, egfr_60m, egfr_72m, egfr_84m, egfr_96m, 
                                 uacr_1m, uacr_3m, uacr_12m, mgfr_3m, mgfr_12m, tac_d2,
                                 tac_d7,  tac_d14, tac_d21, tac_1m, tac_3m,  tac_12m, 
                                 cmv_viremia, cmv_disease, resist_cmv, ebv_infection, 
                                 bk_viremia, bkvan,
                                 nodat,fungal_invasive_orblood,recurrent_uti, 
                                 gasteroenteritis, chest_infection, PTLD, skin_ca, 
                                 cvd_0_5y, induct_cni, induct_apa, induct_bas, induct_extra, 
                                 bpar_0_1m, bpar_1m, bpar_1_3m,  bpar_3m, bpar_3_12m, 
                                 bpar_12m, 
                                 bpar_post12m, Subclin_0_1m,  subclin_1m, 
                                 subclin_1_3m, subclin_3m, 
                                 subclin_3_12m, subclin_12m,  bkvan_0_1m, bkvan_1m,bk_van_3m, 
                                 bkvan_3_12m,  bk_van_12m, bkvan_post12m,  i_0m, i_1m, i_3m, 
                                 i_12m, t_0m, t_1m, t_3m, t_12m, g_0m, g_1m, g_3m, g_12m, 
                                 v_0m, v_1m, v_3m, v_12m,ci_0m, ci_1m, ci_3m, ci_12m, 
                                 c4d_0m, c4d_1m, c4d_3m, c4d_12m, ct_0m, ct_1m, ct_3m, ct_12m, 
                                 cg_0m, cg_1m, cg_3m, cg_12m,iIFTA_1m, iIFTA_3m, iIFTA_12m,
                                 event_any, event_all,  death, 
                                 IFTA_score_3m_meena1, IFTA_score_3m_meena1, 
                                 iIFTA_score_3m_meena1, iIFTA_score_12m_meena1,
                                 nodat, induct_extra)

outcomes = c()
outcomes = dgf.filter %>% select(dgf_define, dgf_define.comb,  dgf.severity, 
                                 i_0m, i_1m, i_3m, organ_type,
                                 i_12m, t_0m, t_1m, t_3m, t_12m, g_0m, g_1m, g_3m, g_12m, 
                                 v_0m, v_1m, v_3m, v_12m,ci_0m, ci_1m, ci_3m, ci_12m, 
                                 c4d_0m, c4d_1m, c4d_3m, c4d_12m, ct_0m, ct_1m, ct_3m, ct_12m, 
                                 cg_0m, cg_1m, cg_3m, cg_12m,iIFTA_1m, iIFTA_3m, iIFTA_12m,
                                 event_any, event_all,  death, 
                                 IFTA_score_3m_meena1, IFTA_score_12m_meena1, 
                                 iIFTA_score_3m_meena1, iIFTA_score_12m_meena1,
                                 nodat, induct_extra)

# outcomes = outcomes[outcomes$dgf_define != "DGF.HD",]
# outcomes$dgf_define = droplevels(outcomes$dgf_define)
# 
# outcomes = outcomes[outcomes$organ_type != "Living Kidney Transplant",]
# outcomes$organ_type = droplevels(outcomes$organ_type)

outcomes = outcomes %>% mutate(
  ci.0m = ifelse(ci_0m <2, "low", "high"),
  ci.1m = ifelse(ci_1m <2, "low", "high"),
  ci.3m = ifelse(ci_3m <2, "low", "high"),
  ci.12m = ifelse(ci_12m <2, "low", "high"),
  ct.0m = ifelse(ct_0m <2, "low", "high"),
  ct.1m = ifelse(ct_1m <2, "low", "high"),
  ct.3m = ifelse(ct_3m <2, "low", "high"),
  ct.12m = ifelse(ct_12m <2, "low", "high"),
  cg.0m = ifelse(cg_0m <2, "low", "high"),
  cg.1m = ifelse(cg_1m <2, "low", "high"),
  cg.3m = ifelse(cg_3m <2, "low", "high"),
  cg.12m = ifelse(cg_12m <2, "low", "high"), 
  iifta.1m = ifelse(iIFTA_1m <2, "low", "high"),
  iifta.3m = ifelse(iIFTA_3m <2, "low", "high"),
  iifta.3m.meena0 = ifelse(iIFTA_score_3m_meena1 == 0, "zero", "not"), 
  iifta.3m.meena1 = ifelse(iIFTA_score_3m_meena1 == 1, "one", "not"), 
  iifta.3m.meena2 = ifelse(iIFTA_score_3m_meena1 == 2, "two", "not"), 
  # iifta.12m.meena0 = ifelse(iIFTA_score_12m_meena1 == 0, "zero", "not"), 
  # iifta.12m.meena1 = ifelse(iIFTA_score_12m_meena1 == 1, "one", "not"), 
  # iifta.12m.meena2 = ifelse(iIFTA_score_12m_meena1 == 2, "two", "not"),
  ifta.3m.meena0 = ifelse(IFTA_score_3m_meena1 == 0, "zero", "not"), 
  ifta.3m.meena1 = ifelse(IFTA_score_3m_meena1 == 1, "one", "not"), 
  ifta.3m.meena2 = ifelse(IFTA_score_3m_meena1 == 2, "two", "not"), 
  ifta.3m.meena3 = ifelse(IFTA_score_3m_meena1 == 3, "three", "not"))
  # , 
  # ifta.12m.meena0 = ifelse(IFTA_score_12m_meena1 == 0, "zero", "not"), 
  # ifta.12m.meena1 = ifelse(IFTA_score_12m_meena1 == 1, "one", "not"), 
  # ifta.12m.meena2 = ifelse(IFTA_score_12m_meena1 == 2, "two", "not"),
  # ifta.12m.meena3 = ifelse(IFTA_score_3m_meena1 == 3, "three", "not"))

outcomes.table = tbl_summary(outcomes, missing = "no",
                             by = dgf_define)

outcomes.table = outcomes.table %>% add_n() %>% 
  add_p() %>% modify_header(label = "Outcomes") %>% bold_labels()

outcomes.table


```


HD criteria
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
outcomes = c()
outcomes = dgf.filter %>% select(dgf_define, dgf_define.comb,  dgf.severity, 
                                 event_any, event_all,  death, primary_nonfunc,
                                 death.cs.graftloss, death_functioninggraft, 
                                 graft_loss,  censor_followup, dgf_hdsessions,  
                                 dgf_hd_days,return_OT_01m,  transfusion_01m, 
                                 wound_infection_01m, cr_pre, cr_0hr, cr_d1, 
                                 cr_d2, cr_1m, cr_3m, cr_12m, cr_24m,cr_36m, 
                                 cr_48m, cr_60m, cr_72m, cr_84m,cr_96m, egfr_1m,  
                                 egfr_3m, egfr_12m, egfr_24m,  egfr_36m, 
                                 egfr_48m, egfr_60m, egfr_72m, egfr_84m, egfr_96m, 
                                 uacr_1m, uacr_3m, uacr_12m, mgfr_3m, mgfr_12m, tac_d2,
                                 tac_d7,  tac_d14, tac_d21, tac_1m, tac_3m,  tac_12m, 
                                 cmv_viremia, cmv_disease, resist_cmv, ebv_infection, 
                                 bk_viremia, bkvan, nodat,fungal_invasive_orblood,recurrent_uti, 
                                 gasteroenteritis, chest_infection, PTLD, skin_ca, 
                                 cvd_0_5y, induct_cni, induct_apa, induct_bas, induct_extra, 
                                 bpar_0_1m, bpar_1m, bpar_1_3m,  bpar_3m, bpar_3_12m, bpar_12m, 
                                 bpar_post12m, Subclin_0_1m,  subclin_1m, subclin_1_3m, subclin_3m, 
                                 subclin_3_12m, subclin_12m,  bkvan_0_1m, bkvan_1m,bk_van_3m, 
                                 bkvan_3_12m,  bk_van_12m, bkvan_post12m,  i_0m, i_1m, i_3m, 
                                 i_12m, t_0m, t_1m, t_3m, t_12m, g_0m, g_1m, g_3m, g_12m, 
                                 v_0m, v_1m, v_3m, v_12m,ci_0m, ci_1m, ci_3m, ci_12m, 
                                 c4d_0m, c4d_1m, c4d_3m, c4d_12m, ct_0m, ct_1m, ct_3m, ct_12m, 
                                 cg_0m, cg_1m, cg_3m, cg_12m,iIFTA_1m, iIFTA_3m, iIFTA_12m,
                                 event_any, event_all,  death, 
                                 nodat, induct_extra1, dgf_hdsessions, dgf_hd_days)

levels(outcomes$dgf_define)
outcomes = outcomes[outcomes$dgf_define != "DGF.Cr",]
outcomes$dgf_define = droplevels(outcomes$dgf_define)
levels(outcomes$dgf_define)

outcomes = outcomes %>% mutate(
  ci.0m = ifelse(ci_0m <2, "low", "high"),
  ci.1m = ifelse(ci_1m <2, "low", "high"),
  ci.3m = ifelse(ci_3m <2, "low", "high"),
  ci.12m = ifelse(ci_12m <2, "low", "high"),
  ct.0m = ifelse(ct_0m <2, "low", "high"),
  ct.1m = ifelse(ct_1m <2, "low", "high"),
  ct.3m = ifelse(ct_3m <2, "low", "high"),
  ct.12m = ifelse(ct_12m <2, "low", "high"),
  cg.0m = ifelse(cg_0m <2, "low", "high"),
  cg.1m = ifelse(cg_1m <2, "low", "high"),
  cg.3m = ifelse(cg_3m <2, "low", "high"),
  cg.12m = ifelse(cg_12m <2, "low", "high"), 
  iifta.1m = ifelse(iIFTA_1m <2, "low", "high"),
  iifta.3m = ifelse(iIFTA_3m <2, "low", "high"),
  iifta.12m = ifelse(iIFTA_12m <2, "low", "high"))

outcomes.table = tbl_summary(outcomes, missing = "no",
                             by = dgf_define)
outcomes.table = outcomes.table %>% add_n() %>% 
  add_p() %>% modify_header(label = "Outcomes") %>% bold_labels()

outcomes.table

```


HD criteria, deceased donor only
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
outcomes = c()
outcomes = dktx.only %>% select(dgf_define, dgf_define.comb,  dgf.severity, 
                                 event_any, event_all,  death, primary_nonfunc,
                                 death.cs.graftloss, death_functioninggraft, 
                                 graft_loss,  censor_followup, dgf_hdsessions,  
                                 dgf_hd_days,return_OT_01m,  transfusion_01m, 
                                 wound_infection_01m, cr_pre, cr_0hr, cr_d1, 
                                 cr_d2, cr_1m, cr_3m, cr_12m, cr_24m,cr_36m, 
                                 cr_48m, cr_60m, cr_72m, cr_84m,cr_96m, egfr_1m,  
                                 egfr_3m, egfr_12m, egfr_24m,  egfr_36m, 
                                 egfr_48m, egfr_60m, egfr_72m, egfr_84m, egfr_96m, 
                                 uacr_1m, uacr_3m, uacr_12m, mgfr_3m, mgfr_12m, tac_d2,
                                 tac_d7,  tac_d14, tac_d21, tac_1m, tac_3m,  tac_12m, 
                                 cmv_viremia, cmv_disease, resist_cmv, ebv_infection, 
                                 bk_viremia, bkvan, nodat,fungal_invasive_orblood,recurrent_uti, 
                                 gasteroenteritis, chest_infection, PTLD, skin_ca, 
                                 cvd_0_5y, induct_cni, induct_apa, induct_bas, induct_extra, 
                                 bpar_0_1m, bpar_1m, bpar_1_3m,  bpar_3m, bpar_3_12m, bpar_12m, 
                                 bpar_post12m, Subclin_0_1m,  subclin_1m, subclin_1_3m, subclin_3m, 
                                 subclin_3_12m, subclin_12m,  bkvan_0_1m, bkvan_1m,bk_van_3m, 
                                 bkvan_3_12m,  bk_van_12m, bkvan_post12m,  i_0m, i_1m, i_3m, 
                                 i_12m, t_0m, t_1m, t_3m, t_12m, g_0m, g_1m, g_3m, g_12m, 
                                 v_0m, v_1m, v_3m, v_12m,ci_0m, ci_1m, ci_3m, ci_12m, 
                                 c4d_0m, c4d_1m, c4d_3m, c4d_12m, ct_0m, ct_1m, ct_3m, ct_12m, 
                                 cg_0m, cg_1m, cg_3m, cg_12m,iIFTA_1m, iIFTA_3m, iIFTA_12m,
                                 event_any, event_all,  death, organ_type,
                                 nodat, induct_extra1, dgf_hdsessions, dgf_hd_days)

levels(outcomes$dgf_define)
outcomes = outcomes[outcomes$dgf_define != "DGF.Cr",]
outcomes$dgf_define = droplevels(outcomes$dgf_define)
levels(outcomes$dgf_define)

levels(outcomes$organ_type)
outcomes = outcomes[outcomes$organ_type != "Living Kidney Transplant",]
outcomes$organ_type = droplevels(outcomes$organ_type)
levels(outcomes$organ_type)

outcomes = outcomes %>% mutate(
  ci.0m = ifelse(ci_0m <2, "low", "high"),
  ci.1m = ifelse(ci_1m <2, "low", "high"),
  ci.3m = ifelse(ci_3m <2, "low", "high"),
  ci.12m = ifelse(ci_12m <2, "low", "high"),
  ct.0m = ifelse(ct_0m <2, "low", "high"),
  ct.1m = ifelse(ct_1m <2, "low", "high"),
  ct.3m = ifelse(ct_3m <2, "low", "high"),
  ct.12m = ifelse(ct_12m <2, "low", "high"),
  cg.0m = ifelse(cg_0m <2, "low", "high"),
  cg.1m = ifelse(cg_1m <2, "low", "high"),
  cg.3m = ifelse(cg_3m <2, "low", "high"),
  cg.12m = ifelse(cg_12m <2, "low", "high"), 
  iifta.1m = ifelse(iIFTA_1m <2, "low", "high"),
  iifta.3m = ifelse(iIFTA_3m <2, "low", "high"),
  iifta.12m = ifelse(iIFTA_12m <2, "low", "high"))

outcomes.table = tbl_summary(outcomes, missing = "no",
                             by = dgf_define)
outcomes.table = outcomes.table %>% add_n() %>% 
  add_p() %>% modify_header(label = "Outcomes") %>% bold_labels()

outcomes.table

```



### Risk of developing DGF 

Simple regression to check for presence and strength of relationship of variables (donor and recipient) to developing DGF (logistic rather than linear model given outcome is categorical). Check variables ok
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
# For combined DGF criteria - dgf_combined1
dependent = "dgf_define"

explantory.recip = c("htn_pre", "dm_pre", "renal_dx","ihd_pre",  "chol_pre" , "ondialysis_pretx",  "organ_type")
explantory.recip.cont = c("tx_yr", "tx_age")
explantory.donor = c("donor_criteria", "donor_criteria_base",  "donor_ionotropes")
explantory.donor.cont = c("cit_min", "donor_age",  "donor_terminal_cr")
explantory.immune = c("dsa1_pre" , "dsa2_pre" , "graft_number", "induct_extra")
explantory.immune.cont = c("HLA_mm")
explantory.immune1 = explantory.immune[explantory.immune != "aboi"]
explantory.recip1 = explantory.recip[explantory.recip != "organ_type"]

# EDA - Check via chi-sq or fischer tests
dgf.filter %>% summary_factorlist(dependent,
                                  c(explantory.recip, explantory.recip.cont),
                                  add_dependent_label = TRUE, p = TRUE) 

dgf.filter %>% summary_factorlist(dependent,
                                  c(explantory.donor,explantory.donor.cont),
                                  add_dependent_label = TRUE, p = TRUE) 

dgf.filter %>% summary_factorlist(dependent,
                                  c(explantory.immune,explantory.immune.cont), 
                                  add_dependent_label = TRUE,  p = TRUE) 

dktx.only %>% summary_factorlist(dependent,
                                  c(explantory.recip1, explantory.recip.cont),
                                  add_dependent_label = TRUE, p = TRUE) 

dktx.only %>% summary_factorlist(dependent,
                                  c(explantory.donor,explantory.donor.cont),
                                  add_dependent_label = TRUE, p = TRUE) 

dktx.only %>% summary_factorlist(dependent,
                                  c(explantory.immune1,explantory.immune.cont), 
                                  add_dependent_label = TRUE,  p = TRUE) 

# Check linearity for DGF predictors
dgf.filter %>% 
  select(dgf_define, tx_yr, tx_age, donor_age, donor_terminal_cr) %>% 
  pivot_longer(all_of(c("tx_yr", "tx_age", "donor_age", "donor_terminal_cr")), 
               names_to = "predictors") %>% 
  ggplot(aes(x = value, y = dgf_define)) + geom_point(size = 0.5, alpha = 0.5) +
  geom_smooth(method = "loess") + facet_wrap(~predictors, scales = "free_x")

p.dgf.1 = dgf.filter %>% 
  select(one_of(c("dgf_define","gender", "renal_dx", 
                  "donor_criteria", "graft_number","bpar_any1"))) %>% 
  pivot_longer(-dgf_define) %>% 
  ggplot(aes(value, fill = dgf_define)) + 
  geom_bar(position = "fill") + theme_bw()+
  facet_wrap(~name, scale = "free", ncol = 5) +  coord_flip()

p.dgf.2 = dgf.filter %>% 
  select(one_of(c("dgf.severity","gender", "renal_dx", 
                  "donor_criteria", "graft_number","bpar_any1"))) %>% 
  pivot_longer(-dgf.severity) %>% 
  ggplot(aes(value, fill = dgf.severity)) + 
  geom_bar(position = "fill") + theme_bw()+
  facet_wrap(~name, scale = "free", ncol = 5) +  coord_flip()

tiff("figures/dgf.comb.breakdown.barplots.tiff", res = 300, units = "cm", height = 25, width = 42)
p.dgf.1/p.dgf.2
dev.off()

# Check multi-collinearity 
tiff("figures/dgf.multi.col.all.old1.tiff", res = 300, units = "cm", height = 20, width = 25)
dgf.filter %>% remove_labels() %>%
  ggpairs(columns = c(explantory.recip.cont, 
                      explantory.immune.cont, 
                      explantory.donor.cont), 
          aes(color = dgf_define, alpha = 0.5)) 
dev.off()

tiff("figures/dgf.multi.col.all.old2.tiff", res = 300, units = "cm", height = 45, width = 75)
dgf.filter %>% remove_labels() %>%
  ggpairs(columns = c(explantory.recip, 
                      explantory.immune, 
                      explantory.donor), 
          aes(color = dgf.severity, alpha = 0.5)) 
dev.off()

# categorical/nominal variables
dgf.filter %>% remove_labels() %>%
  ggpairs(columns = c(explantory.recip, explantory.recip.cont,
                      explantory.immune),
          upper = list(continuous = wrap("cor", size = 2.5)),
          lower = list(continuous = wrap("cor", size = 2.5)), 
          aes(color = dgf_define, alpha = 0.5)) 

tiff("figures/dgf.multi.col.ddKTx.tiff", res = 300, units = "cm", height = 20, width = 25)
dktx.only %>% remove_labels() %>%
  ggpairs(columns = c(explantory.recip.cont, 
                      explantory.immune.cont, 
                      explantory.donor.cont)) 
dev.off()

# if variation inflation factor > 10, suspect multi-collinearity
dgf.filter %>%  glmmulti(dependent, c(explantory.donor.cont, explantory.recip.cont)) %>%car::vif()
dgf.filter %>% glmmulti(dependent, c(explantory.recip)) %>%car::vif()

# ggplot(dktx.only, aes(x = factor(tx_yr), y = cit_min, fill = factor(dgf_hd_criteria1)))+ 
#   geom_dotplot(binaxis = "y", stackdir = "center", dotsize = 0.5) + 
#   facet_wrap(.~donor_criteria, ncol = 5)

histogram(dktx.only$donor_age)
dktx.only$donor_age1 = NA
dktx.only$donor_age1 = ifelse(dktx.only$donor_age <= 20, "<20", dktx.only$donor_age1)
dktx.only$donor_age1 = ifelse(dktx.only$donor_age > 20 & 
                                dktx.only$donor_age <= 40, "20-40", dktx.only$donor_age1)
dktx.only$donor_age1 = ifelse(dktx.only$donor_age > 40 & 
                                dktx.only$donor_age <= 50, "40-50", dktx.only$donor_age1)
dktx.only$donor_age1 = ifelse(dktx.only$donor_age > 50 & 
                                dktx.only$donor_age <= 60, "50-60", dktx.only$donor_age1)
dktx.only$donor_age1 = ifelse(dktx.only$donor_age > 60 & 
                                dktx.only$donor_age <= 70, "60-70", dktx.only$donor_age1)
dktx.only$donor_age1 = ifelse(dktx.only$donor_age > 70, "70+", dktx.only$donor_age1)

qplot(donor_terminal_cr, cit_min, data = dgf.filter, geom = c("point", "smooth"))

```


Logistical regression Control vs DGF.combined (all transplants)
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
#Logistic regression using glm() makes it dichotomous outcome if family = binomial, family = gaussian for linear and use multinomial regression for more than 2 levels (control, cr, hd). Use final fit when predictive variables decided on by fwd, backward or collett.

# all the numerical and categorical variables together
dgf.filter.HD = c()
dgf.filter.HD = dgf.filter
dgf.filter.HD = dgf.filter.HD[dgf.filter.HD$dgf_define != "DGF.Cr",]
dgf.filter.HD$dgf_define = droplevels(dgf.filter.HD$dgf_define)
dgf.filter.HD$tx_yr.2016 = ifelse(dgf.filter.HD$tx_yr <= 2016, "pre-2016", "post-2016")

fit3 = glm(dgf_define ~ 
             tx_age + tx_yr.2016 + cit_min +  donor_age + 
             donor_terminal_cr + donor_ionotropes + 
             htn_pre + dm_pre+ renal_dx +organ_type + 
             donor_criteria +graft_number + 
             dsa1_pre + dsa2_pre + induct_extra1,
            data = dgf.filter.HD, 
           family = binomial)
fit3 %>% tidy(conf.int = TRUE, exp = TRUE)
fit3 %>% glance()
step3 = stepAIC(fit3, scope=list(lower=NULL, upper=fit3), 
                data=dgf.filter.HD, direction='backward')

# step3 = step(fit3, direction = "backward") # same result
summary(fit3)
exp(coef(fit3))
step3 %>% tidy(conf.int = TRUE, exp = TRUE)

# putting this into final fit for publication styling
exp.var1 =  c("htn_pre" ,"dm_pre" , 
              "renal_dx" , "organ_type" , 
              "donor_age" , "cit_min" , 
              "donor_intubated_days" ,  
              "donor_terminal_cr" ,
              "donor_criteria", 
              "donor_inotropes", 
              "graft_number", 
              "dsa1_pre","dsa2_pre", 
              "tx_yr", "tx_yr.2016" ,
              "induct_extra")

exp.var2 =  c("tx_yr.2016",
              "donor_inotropes", 
              "renal_dx", 
              "donor_criteria")

fit3 = dgf.filter.HD %>% finalfit("dgf_define.comb", 
                              exp.var1, 
                              keep_models = TRUE,
                              metrics = TRUE)

knitr::kable(fit3[[1]], row.names=FALSE)
knitr::kable(fit3, row.names=FALSE, 
             align=c("l", "l", "r", "r", "r", "r"))

fit4 = glm(dgf_define.comb ~ 
             tx_yr +
             donor_inotropes + 
             renal_dx +
             donor_criteria,
            data = dgf.filter.HD, 
           family = binomial)

fit4 %>% tidy(conf.int = TRUE, exp = TRUE)
fit4 %>% glance()
step4 = stepAIC(fit4, scope=list(lower=NULL, upper=fit4), 
                data=dgf.filter.HD, direction='backward')
summary(fit4)
exp(coef(fit4))
fit4 %>% tidy(conf.int = TRUE, exp = TRUE)

fit4 = dgf.filter.HD %>% finalfit("dgf_define.comb", 
                              exp.var2, 
                              keep_models = TRUE,
                              metrics = TRUE)
fit4
knitr::kable(fit4, row.names=FALSE, 
             align=c("l", "l", "r", "r", "r", "r"))

# # Odds ratio of developing DGF
# tiff("figures/p.dgf.OR.tiff", res = 300, units = "cm", height = 10, width = 20)
# dgf.filter.HD %>% or_plot("dgf_define.comb", exp.var2,
#           breaks = c(0.5, 1, 2, 5, 10, 25),
#           dependent_label = "DGF",
#           table_text_size = 3,
#           title_text_size = 16)
# dev.off()

### SLOW GRAFT FUNCTION ########
# all the numerical and categorical variables together
dgf.filter.HD = c()
dgf.filter.HD = dgf.filter
dgf.filter.HD = dgf.filter.HD[dgf.filter.HD$dgf_define != "DGF.HD",]
dgf.filter.HD$dgf_define = droplevels(dgf.filter.HD$dgf_define)
dgf.filter.HD$tx_yr.2016 = ifelse(dgf.filter.HD$tx_yr <= 2016, "pre-2016", "post-2016")

fit3 = glm(dgf_define ~ 
             tx_age + tx_yr.2016 + cit_min +  donor_age + 
             donor_terminal_cr + donor_ionotropes + 
             htn_pre + dm_pre+ renal_dx +organ_type + 
             donor_criteria +graft_number + 
             dsa1_pre + dsa2_pre + induct_extra1,
            data = dgf.filter.HD, 
           family = binomial)
fit3 %>% tidy(conf.int = TRUE, exp = TRUE)
fit3 %>% glance()
step3 = stepAIC(fit3, scope=list(lower=NULL, upper=fit3), 
                data=dgf.filter.HD, direction='backward')

# step3 = step(fit3, direction = "backward") # same result
summary(fit3)
exp(coef(fit3))
step3 %>% tidy(conf.int = TRUE, exp = TRUE)

# putting this into final fit for publication styling
exp.var1 =  c("htn_pre" ,"dm_pre" , 
              "renal_dx" , "organ_type" , 
              "donor_age" , "cit_min" , 
              "donor_intubated_days" ,  
              "donor_terminal_cr" ,
              "donor_criteria", 
              "donor_inotropes", 
              "graft_number", 
              "dsa1_pre","dsa2_pre", 
              "tx_yr", "tx_yr.2016" ,
              "induct_extra")

exp.var2 =  c("donor_inotropes",
              "donor_criteria")

fit3 = dgf.filter.HD %>% finalfit("dgf_define", 
                              exp.var1, 
                              keep_models = TRUE,
                              metrics = TRUE)

knitr::kable(fit3[[1]], row.names=FALSE)
knitr::kable(fit3, row.names=FALSE, 
             align=c("l", "l", "r", "r", "r", "r"))

fit4 = glm(dgf_define ~ tx_age + tx_yr.2016 + cit_min +  donor_age + 
             donor_terminal_cr + donor_inotropes + 
             htn_pre + dm_pre+ renal_dx +organ_type + 
             donor_criteria +graft_number + 
             dsa1_pre + dsa2_pre + induct_extra,
            data = dgf.filter.HD, 
           family = binomial)

fit4 = glm(dgf_define ~ tx_age,
            data = dgf.filter.HD, 
           family = binomial)

summary(fit4)

fit4 %>% tidy(conf.int = TRUE, exp = TRUE)
fit4 %>% glance()
step4 = stepAIC(fit4, scope=list(lower=NULL, upper=fit4), 
                data=dgf.filter.HD, direction='backward')

exp(coef(fit4))
fit4 %>% tidy(conf.int = TRUE, exp = TRUE)

fit4 = dgf.filter.HD %>% finalfit("dgf_define", 
                              exp.var2, 
                              keep_models = TRUE,
                              metrics = TRUE)
fit4
knitr::kable(fit4, row.names=FALSE, 
             align=c("l", "l", "r", "r", "r", "r"))

```


Simple prediction model - to developing DGF (example for categorical outcomes)
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
#############################################
# Select reasonable variables for the outcome
#############################################
set.seed(42)
# with only clinical variables
variables.pre = c("dgf_define", "tx_yr.2016", "tx_age", 
                   "htn_pre", "ihd_pre", "dm_pre", "renal_dx",
                    "donor_criteria", "HLA_mm",
                   "dsa1_pre", "dsa2_pre", "gender",
                   "organ_type", "graft_number", "donor_age", 
                   "donor_inotropes", "donor_terminal_urea",
                   "donor_terminal_cr", "donor_intubated_days",
                   "cit_min", "induct_bas")
dgf.filter.pred = dgf.filter[,variables.pre]
dgf.filter.pred = dgf.filter.pred[dgf.filter.pred$dgf_define != "DGF.Cr",]
dgf.filter.pred$dgf_define = droplevels(dgf.filter.pred$dgf_define)

# Split data, p = %
split1 = createDataPartition(dgf.filter.pred$dgf_define, p= 0.6, list = FALSE)
dgf.filter.pred.1 = dgf.filter.pred[split1,]
dgf.filter.pred.2 = dgf.filter.pred[-split1,]
X = dgf.filter.pred.1[,2:ncol(dgf.filter.pred.1)]
Y = dgf.filter.pred.1$dgf_define

trainData = X
testData = dgf.filter.pred.2[,2:ncol(dgf.filter.pred.2)]

###################################
########### TRAIN SET #############
###################################
# Impute missing
preProcess_missingdata_model = preProcess(trainData, method='knnImpute')
preProcess_missingdata_model
trainData = predict(preProcess_missingdata_model, newdata = trainData)
anyNA(trainData)

trainData$dgf_define = dgf.filter.pred.1$dgf_define

# ONE-HOT encoding:  Creating dummy variables is converting a categorical variable to as many binary variables as here are categories.
dummies_model = dummyVars(dgf_define ~ ., data=trainData)
trainData_mat = predict(dummies_model, newdata = trainData) # Create the dummy variable
trainData = data.frame(trainData_mat) # Convert to dataframe

# Transform/scale
x = trainData
preprocessparam = preProcess(x, method = c("center", "scale"))

X_train = predict(preprocessparam, x)
X_train = as.data.frame(apply(X_train, 2, function(x) as.numeric(x)))
str(X_train)
Y_train = dgf.filter.pred.1$dgf_define
data.train = cbind(dgf_define = ifelse(Y_train == "Control", 1, 2), X_train)
data.train = as.data.frame(apply(data.train, 2, function(x) as.numeric(x)))

##################################
########### TEST SET #############
##################################
# Impute missing
preProcess_missingdata_model = preProcess(testData, method='knnImpute')
preProcess_missingdata_model
testData = predict(preProcess_missingdata_model, newdata = testData)
anyNA(testData)

testData$dgf_define = dgf.filter.pred.2$dgf_define

# ONE-HOT encoding:  Creating dummy variables is converting a categorical variable to as many binary variables as here are categories.
dummies_model = dummyVars(dgf_define ~ ., data=testData)
trainData_mat = predict(dummies_model, newdata = testData) # Create the dummy variable
testData = data.frame(trainData_mat) # Convert to dataframe

# Transform/scale
x = testData
preprocessparam = preProcess(x, method = c("center", "scale"))

X_test = predict(preprocessparam, x)
X_test = as.data.frame(apply(X_test, 2, function(x) as.numeric(x)))

Y_test= dgf.filter.pred.2$dgf_define
data.test = cbind(dgf_define = ifelse(Y_test == "Control", 1, 2), X_test)
data.test = as.data.frame(apply(data.test, 2, function(x) as.numeric(x)))

####################################################
######### MODEL ####################################
####################################################
# Penalised logistic regression (lasso) with 10-fold CV)
train.1 = train(dgf_define ~.,
                data = data.train,
                preProc = c("center", "scale"),
                method = "lasso",
                trControl = trainControl(method = "repeatedcv", 
                                         number = 10, 
                                         repeats = 10))

# assess.lasso = assess.glmnet(train.1, 
#                              newx = as.matrix(data.train[,2:ncol(data.train)]), 
#                              newy = data.train$dgf_define )

summary(train.1$results)
plot(train.1$results)

tiff("figures/pen.log.regression.var.importance.DGF.roc.tiff", 
     units = "cm", height = 30, width = 12, res = 300)
ggplot(varImp(train.1))
dev.off()

## above + model hyperparameters
# train.2 = train(dgf_define ~.,
#                 data = data.train,
#                 preProc = c("center", "scale"),
#                 method = "lasso",
#                 trControl = trainControl(method = "repeatedcv", 
#                                          number = 10, 
#                                          repeats = 10, 
#                                          search = "random"))

summary(train.2$results)
plot(train.2$results)
ggplot(varImp(train.2))

# Apply to test set
predictions = predict(train.1, X_test)
summary(predictions)
rocs = performance(prediction(predictions, Y_test), 'tpr', 'fpr')
auc(Y_test, predictions)

tiff("figures/pen.log.regression.DGF.roc.tiff", 
     units = "cm", height = 10, width = 12, res = 300)
plot(rocs, curve = TRUE) + abline(a = 0, b = 1)
dev.off()

RMSE(predictions, Y_test)
```




### Long term outcomes

Logistical regression
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
#Logistic regression using glm() makes it dichotomous outcome if family = binomial, family = gaussian for linear and use multinomial regression for more than 2 levels (control, cr, hd). Use final fit when predictive variables decided on by fwd, backward or collett.

# all the numerical and categorical variables together
dgf.filter.HD = c()
dgf.filter.HD = dgf.filter

dgf.filter.HD$bpar.in12m = ifelse(dgf.filter.HD$bpar_0_1m == "No BPAR" &
                                    dgf.filter.HD$bpar_1m  == "No BPAR" &
                                    dgf.filter.HD$bpar_1_3m  == "No BPAR" &
                                    dgf.filter.HD$bpar_3m  == "No BPAR" &
                                    dgf.filter.HD$bpar_3_12m  == "No BPAR",  
                                  "No BPAR",   "BPARupto12m")
dgf.filter.HD$bpar.in12m = ifelse(is.na(dgf.filter.HD$bpar.in12m), "No BPAR", dgf.filter.HD$bpar.in12m)

dgf.filter.HD$bpar.in3m = ifelse(dgf.filter.HD$bpar_0_1m == "No BPAR" &
                                    dgf.filter.HD$bpar_1m  == "No BPAR" &
                                    dgf.filter.HD$bpar_1_3m  == "No BPAR" &
                                    dgf.filter.HD$bpar_3m  == "No BPAR",  
                                  "No BPAR",   "BPARupto3m")
dgf.filter.HD$bpar.in3m = ifelse(is.na(dgf.filter.HD$bpar.in3m), "No BPAR", dgf.filter.HD$bpar.in3m)

dgf.filter.HD$subclin.in12m = ifelse(dgf.filter.HD$Subclin_0_1m == "No Subclin" &
                                    dgf.filter.HD$subclin_1m  == "No Subclin" &
                                    dgf.filter.HD$subclin_1_3m  == "No Subclin" &
                                    dgf.filter.HD$subclin_3m  == "No Subclin" &
                                    dgf.filter.HD$subclin_3_12m  == "No Subclin",  
                                  "No subclin",   "Subclinupto12m")
dgf.filter.HD$subclin.in12m = ifelse(is.na(dgf.filter.HD$subclin.in12m), "No Subclin",
                                     dgf.filter.HD$subclin.in12m)

dgf.filter.HD$subclin.in3m = ifelse(dgf.filter.HD$Subclin_0_1m == "No Subclin" &
                                    dgf.filter.HD$subclin_1m  == "No Subclin" &
                                    dgf.filter.HD$subclin_1_3m  == "No Subclin" &
                                    dgf.filter.HD$subclin_3m  == "No Subclin" ,  
                                  "No subclin",   "Subclinupto3m")
dgf.filter.HD$subclin.in3m = ifelse(is.na(dgf.filter.HD$subclin.in3m), "No Subclin",
                                     dgf.filter.HD$subclin.in3m)

dgf.filter.HD$dgf.long = ifelse(dgf.filter.HD$dgf.severity == "HD => 7 days", "long", "short")

dgf.filter.HD$IFTA_score_12m_meena2 = ifelse(as.numeric(dgf.filter.HD$IFTA_score_12m_meena1)<2, 
                                            0,1)

# predictors of 12 months with 3m IFTA score included
# IFTA3m itself doesn't stay as an independent predictor!!!
fit3 = lm(mgfr_12m ~ 
            mgfr_3m +
            egfr_1m+
            subclin.in12m +
            bpar.in12m +
            bkvan_any + 
            tx_age + tx_yr.2016 +
            dgf_define + 
            graft_number + 
            induct_extra +
            dsa1_pre + dsa2_pre +
            HLA_mm +
            IFTA_score_3m_meena1,
          data = dgf.filter.HD[!is.na(dgf.filter.HD$IFTA_score_3m_meena1), ])
fit3 %>% tidy(conf.int = TRUE, exp = TRUE)
fit3 %>% glance()
step3 = stepAIC(fit3, scope=list(lower=NULL, upper=fit3), 
                data=dgf.filter.HD[!is.na(dgf.filter.HD$IFTA_score_3m_meena1), ], 
                direction='backward')

# So remove IFTA criteria in 12 month prediction
fit3 = lm(mgfr_12m ~ 
            mgfr_3m +
            egfr_1m+
            subclin.in12m +
            bpar.in12m +
            bkvan_any + 
            tx_age + tx_yr.2016 +
            dgf_define + dgf.long +
            graft_number + 
            induct_extra +
            dsa1_pre + dsa2_pre +
            HLA_mm,
          data = dgf.filter.HD)

fit3 = lm(mgfr_12m ~ 
            mgfr_3m +
            egfr_1m+
            bkvan_any + 
            dgf_define +
            dsa1_pre,
          data = dgf.filter.HD)

fit3 %>% tidy(conf.int = TRUE, exp = TRUE)
fit3 %>% glance()
step3 = stepAIC(fit3, scope=list(lower=NULL, upper=fit3), 
                data=dgf.filter.HD, direction='backward')
summary(fit3)

# step3 = step(fit3, direction = "backward") # same result
summary(fit3)
exp(coef(fit3))
step3 %>% tidy(conf.int = TRUE, exp = TRUE)
confint(step3)
sigma(step3)/mean(dgf.filter.HD$mgfr_12m)


fit3 = dgf.filter.HD %>% 
  finalfit("mgfr_12m", 
           c("mgfr_3m", "egfr_1m", "bkvan_any",
             "dgf_define", "dsa1_pre"), 
           keep_models = TRUE,
           metrics = TRUE)
summary(fit3)
knitr::kable(fit3, row.names=FALSE, 
             align=c("l", "l", "r", "r", "r", "r"))

# mGFR 3m
fit4 = lm(mgfr_3m ~ 
          egfr_1m+
            subclin.in3m +
            bpar.in3m +
            bkvan_any + 
            tx_age + tx_yr.2016 +
            dgf_define + dgf.long +
            graft_number + 
            induct_extra +
            dsa1_pre + dsa2_pre +
            HLA_mm,
            data = dgf.filter.HD)

fit4 = lm(mgfr_3m ~ 
          egfr_1m+
            bpar.in3m +
            dgf_define +
            graft_number + 
            dsa1_pre,
            data = dgf.filter.HD)

fit4 %>% tidy(conf.int = TRUE, exp = TRUE)
fit4 %>% glance()
step4 = stepAIC(fit4, scope=list(lower=NULL, upper=fit4), 
                data=dgf.filter.HD, 
                direction='backward')
summary(fit4)

exp(coef(fit4))
fit4 %>% tidy(conf.int = TRUE, exp = TRUE)
confint(fit4)
sigma(fit4)/mean(dgf.filter.HD$mgfr_12m)

fit4 = dgf.filter.HD %>% finalfit("mgfr_12m", 
                              exp.var2, 
                              keep_models = TRUE,
                              metrics = TRUE)
summary(fit4)

knitr::kable(fit4, row.names=FALSE, 
             align=c("l", "l", "r", "r", "r", "r"))


### IFTA 12m
fit4 = glm(IFTA_score_12m_meena2 ~ 
            egfr_1m+
             egfr_3m + 
             mgfr_3m +
             IFTA_score_3m_meena1 +
            subclin.in12m +
            bpar.in12m +
            bkvan_any + 
            tx_age + tx_yr.2016 +
            dgf_define + dgf.long +
            graft_number + 
            induct_extra +
            dsa1_pre + dsa2_pre +
            HLA_mm,
            data = dgf.filter.HD[!is.na(dgf.filter.HD$IFTA_score_12m_meena2),], 
           family = binomial)

fit4 = glm(IFTA_score_12m_meena2 ~ 
             IFTA_score_3m_meena1, 
            data = dgf.filter.HD[!is.na(dgf.filter.HD$IFTA_score_12m_meena2),], 
           family = binomial)

fit4 %>% tidy(conf.int = TRUE, exp = TRUE)
fit4 %>% glance()
step4 = stepAIC(fit4, scope=list(lower=NULL, upper=fit4), 
                data=dgf.filter.HD[!is.na(dgf.filter.HD$IFTA_score_12m_meena2),], 
                direction='backward')
summary(fit4)


# # Odds ratio of developing DGF
# tiff("figures/p.dgf.OR.tiff", res = 300, units = "cm", height = 10, width = 20)
# dgf.filter.HD %>% or_plot("dgf_define.comb", exp.var2,
#           breaks = c(0.5, 1, 2, 5, 10, 25),
#           dependent_label = "DGF",
#           table_text_size = 3,
#           title_text_size = 16)
# dev.off()

```

DGF/IFTA
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
new.matrix = dgf.filter %>% select(dgf_define, IFTA_score_12m_meena1, mgfr_12m)
new.matrix = new.matrix[!is.na(new.matrix$IFTA_score_12m_meena1),]
new.matrix$mgfr_12m = ifelse(new.matrix$mgfr_12m<45, "low", "high")

new.matrix$IFTA_score_12m_meena1 = as.numeric(new.matrix$IFTA_score_12m_meena1)
new.matrix$IFTA_score_12m_meena1 = ifelse(new.matrix$IFTA_score_12m_meena1<2, "low", "high")
counts.dgf = new.matrix %>% group_by(dgf_define) %>% summarise(counts = n())
counts.ifta = new.matrix %>% group_by(IFTA_score_12m_meena1) %>% summarise(counts = n())

ifta.matrix = table(new.matrix$dgf_define, new.matrix$IFTA_score_12m_meena1)
ifta.matrix = ifta.matrix[-2,]
ifta.matrix = ifta.matrix[,c(2,1)]
RR.IFTA.DGF = riskratio.wald(ifta.matrix)
RR.IFTA.DGF

ifta.matrix = table(new.matrix$dgf_define, new.matrix$IFTA_score_12m_meena1)
ifta.matrix = ifta.matrix[-3,]
ifta.matrix = ifta.matrix[,c(2,1)]
RR.IFTA.SGF = riskratio.wald(ifta.matrix)
RR.IFTA.SGF

mgfr.matrix = table(new.matrix$dgf_define, new.matrix$mgfr_12m)
mgfr.matrix = mgfr.matrix[-2,]
RR.mgfr.DGF = riskratio.wald(mgfr.matrix)
RR.mgfr.DGF

mgfr.matrix = table(new.matrix$dgf_define, new.matrix$mgfr_12m)
mgfr.matrix = mgfr.matrix[-3,]
RR.mgfr.SGF = riskratio.wald(mgfr.matrix)
RR.mgfr.SGF

```


IFTA/mGFR/DGF KM curves
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
# Make DGF/IFTA, DGF/mGFR variable
dgf.filter = dgf.filter%>%
  mutate(dgf.IFTA = 
           case_when(dgf_define == "Control" & as.numeric(IFTA_score_12m_meena1)<2 ~ "Control",
                     dgf_define == "Control" & as.numeric(IFTA_score_12m_meena1)>1 ~ "Control.IFTA",
                     dgf_define == "DGF.Cr" & as.numeric(IFTA_score_12m_meena1)<2 ~ "SGF",
                     dgf_define == "DGF.Cr" & as.numeric(IFTA_score_12m_meena1)>1 ~ "SGF.IFTA",
                     dgf_define == "DGF.HD" & as.numeric(IFTA_score_12m_meena1)<2 ~ "DGF",
                     dgf_define == "DGF.HD" & as.numeric(IFTA_score_12m_meena1)>1 ~ "DGF.IFTA"))

dgf.filter = dgf.filter%>%
  mutate(dgf.mgfr = 
           case_when(dgf_define == "Control" & mgfr_12m >=45 ~ "Control",
                     dgf_define == "Control" & mgfr_12m <45  ~ "Control.low",
                     dgf_define == "DGF.Cr" & mgfr_12m >=45 ~ "SGF",
                     dgf_define == "DGF.Cr" & mgfr_12m <45  ~ "SGF.low",
                     dgf_define == "DGF.HD" & mgfr_12m >=45 ~ "DGF",
                     dgf_define == "DGF.HD" & mgfr_12m <45 ~ "DGF.low"))

############################
# Death and non-censored graft loss for IFTA
############################
composite.events = survival::survfit(Surv(time = dgf.filter$time.in.study, 
                                           event = dgf.filter$event.main) 
                                      ~ dgf.IFTA, data = dgf.filter)

coxph(Surv(dgf.filter$time.in.study, dgf.filter$event.main) ~ dgf.IFTA, 
      data = dgf.filter) %>% gtsummary::tbl_regression(exp = TRUE) 

p.comp = p.surv(composite.events, 
                c("Control", "Control.IFTA>2", 
                  "SGF", "SGF.IFTA>2", 
                  "DGF", "DGF.IFTA>2"), "npg", 
                "DGF/IFTA vs composite outcomes")

p.comp = p.comp$plot/p.comp$table + plot_layout(heights = c(3,1))

############################
# Death and non-censored graft loss for mGFR
############################
composite.events1 = survival::survfit(Surv(time = dgf.filter$time.in.study, 
                                           event = dgf.filter$event.main) 
                                      ~ dgf.mgfr, data = dgf.filter)

coxph(Surv(dgf.filter$time.in.study, dgf.filter$event.main) ~ dgf.mgfr, 
      data = dgf.filter) %>% gtsummary::tbl_regression(exp = TRUE) 

p.comp1 = p.surv(composite.events1, 
                c("Control", "Control.mGFR.low", 
                  "SGF", "SGF.mGFR.low", 
                  "DGF", "DGF.mGFR.low"), "npg", 
                "DGF/IFTA vs composite outcomes")
p.comp1 = p.comp1$plot/p.comp1$table + plot_layout(heights = c(3,1))

tiff("figures/km.ifta.mGFR.dgf01.tiff", units = "cm", height = 12, width = 30, res = 300)
p.comp - p.comp1
dev.off()

############################
# Death censored graft loss
############################
sum(dgf.filter$death.cs.graftloss == "yes")
sum(dgf.filter$graft.loss.noncensored == 1)
dgf.filter$death.cs.graftloss1 = ifelse(dgf.filter$death.cs.graftloss == "yes", 1, 0)

# Death censored graft loss for DGF
dgf.filter1 = dgf.filter[dgf.filter$dgf_define != "DGF.Cr",]
composite.events = survival::survfit(Surv(time = dgf.filter1$time.in.study, 
                                           event = dgf.filter1$death.cs.graftloss1) 
                                      ~ dgf_define, data = dgf.filter1)

coxph(Surv(dgf.filter1$time.in.study, dgf.filter1$death.cs.graftloss1) ~ dgf_define, 
      data = dgf.filter1) %>% gtsummary::tbl_regression(exp = TRUE) 

p.comp = p.surv(composite.events, 
                c("Control", "DGF"), "npg", 
                "DGF/IFTA vs composite outcomes")

p.comp = p.comp$plot/p.comp$table + plot_layout(heights = c(3,1))
p.comp

# Death censored graft loss for DGF for IFTA
composite.events = survival::survfit(Surv(time = dgf.filter$time.in.study, 
                                           event = dgf.filter$death.cs.graftloss1) 
                                      ~ dgf.IFTA, data = dgf.filter)

coxph(Surv(dgf.filter$time.in.study, dgf.filter$death.cs.graftloss1) ~ dgf.IFTA, 
      data = dgf.filter) %>% gtsummary::tbl_regression(exp = TRUE) 

p.comp = p.surv(composite.events, 
                c("Control", "Control.IFTA>2", 
                  "SGF", "SGF.IFTA>2", 
                  "DGF", "DGF.IFTA>2"), "npg", 
                "DGF/IFTA vs composite outcomes")

p.comp = p.comp$plot/p.comp$table + plot_layout(heights = c(3,1))
p.comp

# Death censored graft loss for DGF for mGFR
composite.events1 = survival::survfit(Surv(time = dgf.filter$time.in.study, 
                                           event = dgf.filter$death.cs.graftloss1) 
                                      ~ dgf.mgfr, data = dgf.filter)

coxph(Surv(dgf.filter$time.in.study, dgf.filter$death.cs.graftloss1) ~ dgf.mgfr, 
      data = dgf.filter) %>% gtsummary::tbl_regression(exp = TRUE) 

p.comp1 = p.surv(composite.events1, 
                c("Control", "Control.mGFR.low", 
                  "SGF", "SGF.mGFR.low", 
                  "DGF", "DGF.mGFR.low"), "npg", 
                "DGF/IFTA vs composite outcomes")
p.comp1 = p.comp1$plot/p.comp1$table + plot_layout(heights = c(3,1))
p.comp1



#### COMPETITNG RISKS - IFTA
dgf.filter1 = dgf.filter[!is.na(dgf.filter$dgf.IFTA),]

comp.risk = cuminc(ftime = dgf.filter1$time.in.study, 
              fstatus = dgf.filter1$event_all, 
              group = dgf.filter1$dgf.IFTA,
              cencode = "Nil")

ciplotdat = comp.risk %>% 
  list_modify("Tests" = NULL) %>% 
  map_df(`[`, c("time", "est"), .id = "id") %>% 
  mutate(id = recode(
    id, 
    "Control death" = "Death (Control)", 
    "DGF death" = "Death (DGF)", 
    "SGF death" = "Death (SGF)", 
    "Control graft_loss_only" = "Graft loss (Control)", 
    "DGF graft_loss_only" = "Graft loss (DGF)", 
    "SGF graft_loss_only" = "Graft loss (SGF)", 
    "Control Censored" = "Censored (Control)", 
    "DGF Censored" = "Censored (DGF)", 
    "SGF Censored" = "Censored (SGF)",
    "Control.IFTA death" = "Death.IFTA (Control)", 
    "DGF.IFTA death" = "Death.IFTA (DGF)", 
    "SGF.IFTA death" = "Death.IFTA (SGF)", 
    "Control.IFTA graft_loss_only" = "Graft loss.IFTA (Control)", 
    "DGF.IFTA graft_loss_only" = "Graft loss.IFTA (DGF)", 
    "SGF.IFTA graft_loss_only" = "Graft loss.IFTA (SGF)", 
    "Control.IFTA Censored" = "Censored.IFTA (Control)", 
    "DGF.IFTA Censored" = "Censored.IFTA (DGF)", 
    "SGF.IFTA Censored" = "Censored.IFTA (SGF)" )) %>% 
  separate(id, c("Group", "Death", "Graft Loss", "Censored"), ":")

ciplotdat = ciplotdat %>%
  mutate(split = 
           case_when(Group == "Death (Control)" |Group == "Death (DGF)" | 
                       Group == "Death (SGF)" | Group == "Death.IFTA (Control)" |
                       Group == "Death.IFTA (DGF)" | Group == "Death.IFTA (SGF)"  ~ "Death",
                     Group == "Graft loss (Control)" |Group == "Graft loss (DGF)" |
                       Group == "Graft loss (SGF)" |  Group == "Graft loss.IFTA (Control)" |
                       Group == "Graft loss.IFTA (DGF)" |Group == "Graft loss.IFTA (SGF)" ~ "Graft Loss",
                     Group == "Censored (Control)" |Group == "Censored (DGF)" |
                       Group == "Censored (SGF)" | Group == "Censored.IFTA (Control)" |
                       Group == "Censored.IFTA (DGF)" |Group == "Censored.IFTA (SGF)"~ "Censored"))
ciplotdat$split = as.factor(ciplotdat$split)
ciplotdat$split = factor(ciplotdat$split, levels = c("Death", "Graft Loss", "Censored"))
ciplotdat = ciplotdat[!is.na(ciplotdat$split),]

ciplotdat = ciplotdat %>%
  mutate(split.1 = 
           case_when(Group == "Death (Control)" | Group == "Death.IFTA (Control)" |
                     Group == "Graft loss (Control)" | Group == "Graft loss.IFTA (Control)" |
                     Group == "Censored (Control)" | Group == "Censored.IFTA (Control)" ~ "Control", 
                     Group == "Death (SGF)" | Group == "Death.IFTA (SGF)" |
                     Group == "Graft loss (SGF)" | Group == "Graft loss.IFTA (SGF)" |
                     Group == "Censored (SGF)" | Group == "Censored.IFTA (SGF)" ~ "SGF", 
                    Group == "Death (DGF)" | Group == "Death.IFTA (DGF)" |
                     Group == "Graft loss (DGF)" | Group == "Graft loss.IFTA (DGF)" |
                     Group == "Censored (DGF)" | Group == "Censored.IFTA (DGF)" ~ "DGF" ))
ciplotdat$split.1 = as.factor(ciplotdat$split.1)
ciplotdat$split.1 = factor(ciplotdat$split.1, levels = c("Control", "SGF", "DGF"))
ciplotdat = ciplotdat[!is.na(ciplotdat$split.1),]


    "Control death" = "Death (Control)", 
    "DGF death" = "Death (DGF)", 
    "SGF death" = "Death (SGF)", 
    "Control graft_loss_only" = "Graft loss (Control)", 
    "DGF graft_loss_only" = "Graft loss (DGF)", 
    "SGF graft_loss_only" = "Graft loss (SGF)", 
    "Control Censored" = "Censored (Control)", 
    "DGF Censored" = "Censored (DGF)", 
    "SGF Censored" = "Censored (SGF)",
    "Control.IFTA death" = "Death.IFTA (Control)", 
    "DGF.IFTA death" = "Death.IFTA (DGF)", 
    "SGF.IFTA death" = "Death.IFTA (SGF)", 
    "Control.IFTA graft_loss_only" = "Graft loss.IFTA (Control)", 
    "DGF.IFTA graft_loss_only" = "Graft loss.IFTA (DGF)", 
    "SGF.IFTA graft_loss_only" = "Graft loss.IFTA (SGF)", 
    "Control.IFTA Censored" = "Censored.IFTA (Control)", 
    "DGF.IFTA Censored" = "Censored.IFTA (DGF)", 
    "SGF.IFTA Censored" = "Censored.IFTA (SGF)" 
  


tiff("figures/km.ifta.dgf01.tiff", units = "cm", height = 10, width = 18, res = 300)
ggplot(ciplotdat, aes(x = time, y = est, color = Group)) +
  geom_step(lwd = 1.2, aes(fill = Group))  +
  ylim(c(0, 0.3)) + xlim(c(0,120))+
  theme_classic() +
  scale_colour_manual(values = c("deepskyblue1", "deepskyblue1", "deepskyblue1", 
                                 "brown2", "brown2", "brown2", 
                                 "deepskyblue1", "deepskyblue1", "deepskyblue1", 
                                 "brown2", "brown2", "brown2"))+
  theme(plot.title = element_text(size = 14),
        legend.title = element_blank(),
        legend.position = "right") +
  labs(x = "Months", y = "Cumulative incidence",
       title = "Primary outcomes, competing risks")+ 
  facet_grid(split.1~split)
dev.off()

comp.risk$Tests


# Comp risk mGFR
dgf.filter1 = dgf.filter[!is.na(dgf.filter$dgf.mgfr),]

comp.risk = cuminc(ftime = dgf.filter1$time.in.study, 
              fstatus = dgf.filter1$event_all, 
              group = dgf.filter1$dgf.mgfr,
              cencode = "Nil")


ciplotdat = comp.risk %>% 
  list_modify("Tests" = NULL) %>% 
  map_df(`[`, c("time", "est"), .id = "id") %>% 
  mutate(id = recode(
    id, 
    "Control death" = "Death (Control)", 
    "DGF death" = "Death (DGF)", 
    "SGF death" = "Death (SGF)", 
    "Control graft_loss_only" = "Graft loss (Control)", 
    "DGF graft_loss_only" = "Graft loss (DGF)", 
    "SGF graft_loss_only" = "Graft loss (SGF)", 
    "Control Censored" = "Censored (Control)", 
    "DGF Censored" = "Censored (DGF)", 
    "SGF Censored" = "Censored (SGF)",
    "Control.low death" = "Death.low (Control)", 
    "DGF.low death" = "Death.low (DGF)", 
    "SGF.low death" = "Death.low (SGF)", 
    "Control.low graft_loss_only" = "Graft loss.low (Control)", 
    "DGF.low graft_loss_only" = "Graft loss.low (DGF)", 
    "SGF.low graft_loss_only" = "Graft loss.low (SGF)", 
    "Control.low Censored" = "Censored.low (Control)", 
    "DGF.low Censored" = "Censored.low (DGF)", 
    "SGF.low Censored" = "Censored.low (SGF)" )) %>% 
  separate(id, c("Group", "Death", "Graft Loss", "Censored"), ":")

ciplotdat = ciplotdat %>%
  mutate(split = 
           case_when(Group == "Death (Control)" |Group == "Death (DGF)" | 
                       Group == "Death (SGF)" | Group == "Death.low (Control)" |
                       Group == "Death.low (DGF)" | Group == "Death.low (SGF)"  ~ "Death",
                     Group == "Graft loss (Control)" |Group == "Graft loss (DGF)" |
                       Group == "Graft loss (SGF)" |  Group == "Graft loss.low (Control)" |
                       Group == "Graft loss.low (DGF)" |Group == "Graft loss.low (SGF)" ~ "Graft Loss",
                     Group == "Censored (Control)" |Group == "Censored (DGF)" |
                       Group == "Censored (SGF)" | Group == "Censored.low (Control)" |
                       Group == "Censored.low (DGF)" |Group == "Censored.low (SGF)"~ "Censored"))
ciplotdat$split = as.factor(ciplotdat$split)
ciplotdat$split = factor(ciplotdat$split, levels = c("Death", "Graft Loss", "Censored"))
ciplotdat = ciplotdat[!is.na(ciplotdat$split),]

ciplotdat = ciplotdat %>%
  mutate(split.1 = 
           case_when(Group == "Death (Control)" | Group == "Death.low (Control)" |
                     Group == "Graft loss (Control)" | Group == "Graft loss.low (Control)" |
                     Group == "Censored (Control)" | Group == "Censored.low (Control)" ~ "Control", 
                     Group == "Death (SGF)" | Group == "Death.low (SGF)" |
                     Group == "Graft loss (SGF)" | Group == "Graft loss.low (SGF)" |
                     Group == "Censored (SGF)" | Group == "Censored.low (SGF)" ~ "SGF", 
                    Group == "Death (DGF)" | Group == "Death.low (DGF)" |
                     Group == "Graft loss (DGF)" | Group == "Graft loss.low (DGF)" |
                     Group == "Censored (DGF)" | Group == "Censored.low (DGF)" ~ "DGF" ))
ciplotdat$split.1 = as.factor(ciplotdat$split.1)
ciplotdat$split.1 = factor(ciplotdat$split.1, levels = c("Control", "SGF", "DGF"))
ciplotdat = ciplotdat[!is.na(ciplotdat$split.1),]


tiff("figures/km.mgfr.dgf01.tiff", units = "cm", height = 10, width = 18, res = 300)
ggplot(ciplotdat, aes(x = time, y = est, color = Group)) +
  geom_step(lwd = 1.2, aes(fill = Group))  +
  ylim(c(0, 0.3)) + xlim(c(0,120))+
  theme_classic() +
  scale_colour_manual(values = c("deepskyblue1", "deepskyblue1", "deepskyblue1", 
                                 "brown2", "brown2", "brown2", 
                                 "deepskyblue1", "deepskyblue1", "deepskyblue1", 
                                 "brown2", "brown2", "brown2"))+
  theme(plot.title = element_text(size = 14),
        legend.title = element_blank(),
        legend.position = "right") +
  labs(x = "Months", y = "Cumulative incidence",
       title = "Primary outcomes, competing risks")+ 
  facet_grid(split.1~split)
dev.off()

comp.risk$Tests

#########################


##### CAUSE SPECIFIC HR for competiting risk analysis ####

# Cause-specific hazard approach for graft loss vs DGF
dgf.filter2 = dgf.filter1[, c("time.in.study" , "event_all" , "dgf.IFTA")]
chr_fit =coxph(Surv(time.in.study, ifelse(event_all == "death", 1, 0)) ~ dgf.IFTA, 
    data = dgf.filter2)
broom::tidy(chr_fit, exp = TRUE) %>% kable()
gtsummary::tbl_regression(chr_fit, exp = TRUE)


dgf.filter2 = dgf.filter1[, c("time.in.study" , "event_all" , "dgf.mgfr")]
chr_fit =coxph(Surv(time.in.study, ifelse(event_all == "death", 1, 0)) ~ dgf.mgfr, 
    data = dgf.filter2)
broom::tidy(chr_fit, exp = TRUE) %>% kable()
gtsummary::tbl_regression(chr_fit, exp = TRUE)


```

Long term outcomes KM curves
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
# dgf.filter = dgf.filter[dgf.filter$dgf_define != "DGF.Cr",]
# dgf.filter$dgf_define = droplevels(dgf.filter$dgf_define)
p.surv = function(surv.1, legend.labs, palette, title){
  p.surv = ggsurvplot(surv.1, pval = TRUE, conf.int = FALSE, 
                      palette = palette, 
                      legend = "right",
                      legend.labs = legend.labs,
                      title = title,
                      ylab = "Event-free survival",  
                      xlab = "Time (months)",
                      xlim = c(0,60), ylim = c(0, 1), 
                      break.x.by=6,
                      risk.table.col = "strata",  
                      linetype = "strata",   
                      fontsize= 4,
                      risk.table = TRUE, 
                      risk.table.height = 0.2,
                      tables.theme = clean_table_theme(),
                      ggtheme = theme_classic())}



dgf.filter$event_all = ifelse(is.na(dgf.filter$event_all), "Nil",dgf.filter$event_all )
levels(as.factor(dgf.filter$event_all))

dgf.filter = dgf.filter %>%
  mutate(event.all = 
           case_when(event_all == "Nil" ~ 0,
                     event_all == "death" ~ 1,
                     event_all == "graft_loss_only" ~ 2,
                     event_all == "loss_fu" ~3))

###################
# all patients
###################
# Primary outcome: death, graft loss, loss FU
composite.events = survival::survfit(Surv(time = dgf.filter$time.in.study, 
                                          event = dgf.filter$event.all) 
                                      ~ dgf_define, data = dgf.filter)

coxph(Surv(dgf.filter$time.in.study, dgf.filter$event.all) ~ dgf_define, 
      data = dgf.filter) %>% gtsummary::tbl_regression(exp = TRUE) 

p.comp = p.surv(composite.events, c("Control", "SGF", "DGF"), "npg", 
                "Composite outcomes: death, graft loss, loss to follow up")
p.comp = p.comp$plot/p.comp$table + plot_layout(heights = c(3,1))

# Main outcomes: death, graft loss
composite.events1 = survival::survfit(Surv(time = dgf.filter$time.in.study, 
                                           event = dgf.filter$event.main) 
                                      ~ dgf_define, data = dgf.filter)

coxph(Surv(dgf.filter$time.in.study, dgf.filter$event.main) ~ dgf_define, 
      data = dgf.filter) %>%  gtsummary::tbl_regression(exp = TRUE) 

p.comp1 = p.surv(composite.events1, c("Control", "DGF.Cr", "DGF.HD"), "npg", 
                "Composite of death and graft loss only")
p.comp1 = p.comp1$plot/p.comp1$table+ plot_layout(heights = c(3,1))

tiff("figures/km.outcomes.primary01.tiff", units = "cm", height = 12, width = 30, res = 300)
p.comp -p.comp1
dev.off()

# BPAR
dgf.filter$time_to_bpar = ifelse(is.na(dgf.filter$time_to_bpar), 
                                 dgf.filter$time.in.study, dgf.filter$time_to_bpar)
composite.events = survival::survfit(Surv(time = dgf.filter$time_to_bpar, 
                                          event = dgf.filter$bpar_any) 
                                      ~ dgf_define, data = dgf.filter)

coxph(Surv(dgf.filter$time_to_bpar, dgf.filter$bpar_any) ~ dgf_define, 
      data = dgf.filter) %>% gtsummary::tbl_regression(exp = TRUE) 

p.comp = p.surv(composite.events, c("Control", "SGF", "DGF"), "npg", 
                "Biopsy proven acute rejection (BPAR)")
p.comp = p.comp$plot/p.comp$table + plot_layout(heights = c(3,1))

# Subclin rejection
dgf.filter$time_to_subclin = ifelse(is.na(dgf.filter$time_to_subclin), 
                                 dgf.filter$time.in.study, dgf.filter$time_to_subclin)
composite.events1 = survival::survfit(Surv(time = dgf.filter$time_to_subclin, 
                                           event = dgf.filter$subclin_any) 
                                      ~ dgf_define, data = dgf.filter)

coxph(Surv(dgf.filter$time_to_subclin, dgf.filter$subclin_any) ~ dgf_define, 
      data = dgf.filter) %>%  gtsummary::tbl_regression(exp = TRUE) 

p.comp1 = p.surv(composite.events1, c("Control", "DGF.Cr", "DGF.HD"), "npg", 
                "Subclinical Rejection")
p.comp1 = p.comp1$plot/p.comp1$table+ plot_layout(heights = c(3,1))

tiff("figures/km.outcomes.bpar.subclin.01.tiff", units = "cm", height = 12, width = 30, res = 300)
p.comp -p.comp1
dev.off()

####################################################
# Accounting for BPAR/subclin for graft loss/death
###################################################

# BPARc
composite.events = survival::survfit(Surv(time = dgf.filter$time.in.study, 
                                           event = dgf.filter$event.main) 
                                      ~ dgf_define + bpar_any, data = dgf.filter)

coxph(Surv(dgf.filter$time.in.study, dgf.filter$event.main) ~ dgf_define + bpar_any, 
      data = dgf.filter) %>%  gtsummary::tbl_regression(exp = TRUE) 

p.comp = p.surv(composite.events, c("Control/BPAR-", "Control/BPAR+",
                                    "SGF/BPAR-", "SGF/BPAR+",
                                    "DGF/BPAR-", "DGF/BPAR+"), "npg", 
                "Death and graft loss only, considering BPAR")

p.comp = p.comp$plot/p.comp$table+ plot_layout(heights = c(3,1))


# Subclin
composite.events1 = survival::survfit(Surv(time = dgf.filter$time.in.study, 
                                           event = dgf.filter$event.main) 
                                      ~ dgf_define + subclin_any, data = dgf.filter)

coxph(Surv(dgf.filter$time.in.study, dgf.filter$event.main) ~ dgf_define + subclin_any, 
      data = dgf.filter) %>%  gtsummary::tbl_regression(exp = TRUE) 

p.comp1 = p.surv(composite.events1, c("Control/SubClin-", "Control/SubClin+",
                                    "SGF/SubClin-", "SGF/SubClin+",
                                    "DGF/SubClin-", "DGF/SubClin+"), "npg", 
                "Death and graft loss only, considering SubClinical rejection")
p.comp1 = p.comp1$plot/p.comp1$table+ plot_layout(heights = c(3,1))

tiff("figures/km.outcomes.primary02.tiff", units = "cm", height = 12, width = 30, res = 300)
p.comp -p.comp1
dev.off()

# BPARc
dgf.filter1 = dgf.filter[dgf.filter$bpar_any ==0,]
composite.events = survival::survfit(Surv(time = dgf.filter1$time.in.study, 
                                           event = dgf.filter1$event.main) 
                                      ~ dgf_define, data = dgf.filter1)

coxph(Surv(dgf.filter1$time.in.study, dgf.filter1$event.main) ~ dgf_define, 
      data = dgf.filter1) %>%  gtsummary::tbl_regression(exp = TRUE) 

p.comp = p.surv(composite.events, c("Control/BPAR-",
                                    "SGF/BPAR-", 
                                    "DGF/BPAR-"), "npg", 
                "Death and graft loss only, without BPAR")

p.comp = p.comp$plot/p.comp$table+ plot_layout(heights = c(3,1))


dgf.filter2= dgf.filter[dgf.filter$bpar_any ==1,]
composite.events1 = survival::survfit(Surv(time = dgf.filter2$time.in.study, 
                                           event = dgf.filter2$event.main) 
                                      ~ dgf_define + bpar_any, data = dgf.filter2)

coxph(Surv(dgf.filter2$time.in.study, dgf.filter2$event.main) ~ dgf_define, 
      data = dgf.filter2) %>%  gtsummary::tbl_regression(exp = TRUE) 

p.comp1 = p.surv(composite.events1, c("Control/BPAR+",
                                    "SGF/BPAR+",
                                    "DGF/BPAR+"), "npg", 
                "Death and graft loss if BPAR")

p.comp1 = p.comp1$plot/p.comp1$table+ plot_layout(heights = c(3,1))

tiff("figures/km.outcomes.primary02a.tiff", units = "cm", height = 12, width = 30, res = 300)
p.comp -p.comp1
dev.off()

######################
# Just graft loss/BPAR
######################
dgf.filter1 = dgf.filter[dgf.filter$bpar_any == 0,]
dgf.filter1$graft_loss1 = ifelse(dgf.filter1$graft_loss == "Kidney Loss", 1, 0)
dgf.filter1$time_to_graft_loss = ifelse(!is.na(dgf.filter1$time_to_graft_loss),
                                        dgf.filter1$time_to_graft_loss, dgf.filter1$time.in.study)

composite.events = survival::survfit(Surv(time = dgf.filter1$time_to_graft_loss, 
                                           event = dgf.filter1$graft_loss1) 
                                      ~ dgf_define, data = dgf.filter1)

coxph(Surv(dgf.filter1$time.in.study, dgf.filter1$event.main) ~ dgf_define, 
      data = dgf.filter1) %>%  gtsummary::tbl_regression(exp = TRUE) 

p.comp = p.surv(composite.events, c("Control/BPAR-",
                                    "SGF/BPAR-", 
                                    "DGF/BPAR-"), "npg", 
                "Graft loss only, without BPAR")

p.comp = p.comp$plot/p.comp$table+ plot_layout(heights = c(3,1))


dgf.filter2 = dgf.filter[dgf.filter$bpar_any == 1,]
dgf.filter2$graft_loss1 = ifelse(dgf.filter2$graft_loss == "Kidney Loss", 1, 0)
dgf.filter2$time_to_graft_loss = ifelse(!is.na(dgf.filter2$time_to_graft_loss),
                                        dgf.filter2$time_to_graft_loss, dgf.filter2$time.in.study)

composite.events1 = survival::survfit(Surv(time = dgf.filter2$time.in.study, 
                                           event = dgf.filter2$event.main) 
                                      ~ dgf_define, data = dgf.filter2)

coxph(Surv(dgf.filter2$time.in.study, dgf.filter2$event.main) ~ dgf_define, 
      data = dgf.filter2) %>%  gtsummary::tbl_regression(exp = TRUE) 

p.comp1 = p.surv(composite.events1, c("Control/BPAR+",
                                    "SGF/BPAR+",
                                    "DGF/BPAR+"), "npg", 
                "Graft loss, with BPAR episode(s)")

p.comp1 = p.comp1$plot/p.comp1$table+ plot_layout(heights = c(3,1))

tiff("figures/km.outcomes.primary02b.tiff", units = "cm", height = 10, width = 30, res = 300)
p.comp-p.comp1
dev.off()

############################################
# Just graft loss/BPAR - DGF vs control
############################################
dgf.filter1 = dgf.filter[dgf.filter$dgf_define != "DGF.Cr",]
dgf.filter1$dgf_define  = droplevels(dgf.filter1$dgf_define )
dgf.filter1 = dgf.filter1[dgf.filter1$bpar_any == 0,]
dgf.filter1$graft_loss1 = ifelse(dgf.filter1$graft_loss == "Kidney Loss", 1, 0)
dgf.filter1$time_to_graft_loss = ifelse(!is.na(dgf.filter1$time_to_graft_loss),
                                        dgf.filter1$time_to_graft_loss, dgf.filter1$time.in.study)

composite.events = survival::survfit(Surv(time = dgf.filter1$time_to_graft_loss, 
                                           event = dgf.filter1$graft_loss1) 
                                      ~ dgf_define, data = dgf.filter1)

coxph(Surv(dgf.filter1$time.in.study, dgf.filter1$event.main) ~ dgf_define, 
      data = dgf.filter1) %>%  gtsummary::tbl_regression(exp = TRUE) 

p.comp = p.surv(composite.events, c("Control/BPAR-",
                                    "DGF/BPAR-"), "npg", 
                "Graft loss only, without BPAR")

p.comp = p.comp$plot/p.comp$table+ plot_layout(heights = c(3,1))


dgf.filter2 = dgf.filter[dgf.filter$dgf_define != "DGF.Cr",]
dgf.filter2 = dgf.filter2[dgf.filter2$bpar_any == 1,]
dgf.filter2$graft_loss1 = ifelse(dgf.filter2$graft_loss == "Kidney Loss", 1, 0)
dgf.filter2$time_to_graft_loss = ifelse(!is.na(dgf.filter2$time_to_graft_loss),
                                        dgf.filter2$time_to_graft_loss, dgf.filter2$time.in.study)

composite.events1 = survival::survfit(Surv(time = dgf.filter2$time.in.study, 
                                           event = dgf.filter2$event.main) 
                                      ~ dgf_define, data = dgf.filter2)

coxph(Surv(dgf.filter2$time.in.study, dgf.filter2$event.main) ~ dgf_define, 
      data = dgf.filter2) %>%  gtsummary::tbl_regression(exp = TRUE) 

p.comp1 = p.surv(composite.events1, c("Control/BPAR+",
                                      "DGF/BPAR+"), "npg", 
                "Graft loss, with BPAR episode(s)")

p.comp1 = p.comp1$plot/p.comp1$table+ plot_layout(heights = c(3,1))

tiff("figures/km.outcomes.primary02c.tiff", units = "cm", height = 10, width = 30, res = 300)
p.comp-p.comp1
dev.off()



####################################################
# Competing risks - consider all definitions
###################################################
library(cmprsk)

comp.risk = cuminc(ftime = dgf.filter$time.in.study, 
              fstatus = dgf.filter$event_all, 
              group = dgf.filter$dgf_define,
              cencode = "Nil")

glimpse(comp.risk)

plot(comp.risk, xlab = "Months")

ggcompetingrisks(comp.risk, xlab = "Months")

ggcompetingrisks(
  fit = comp.risk, 
  multiple_panels = TRUE,
  conf.int = TRUE,
  xlab = "Months",
  ylab = "Cumulative incidence of event",
  title = "Competing risks: Death, graft loss, loss to follow up",
  ylim = c(0, 0.2), xlim = c(0, 60))

ciplotdat = comp.risk %>% 
  list_modify("Tests" = NULL) %>% 
  map_df(`[`, c("time", "est"), .id = "id") %>% 
  mutate(id = recode(
    id, 
    "Control death" = "Death (control)", 
    "DGF.Cr death" = "Death (SGF)", 
    "DGF.HD death" = "Death (DGF)", 
    "Control graft_loss_only" = "Graft loss (Control)", 
    "DGF.Cr graft_loss_only" = "Graft loss (SGF)", 
    "DGF.HD graft_loss_only" = "Graft loss (DGF)", 
    "Control loss_fu" = "Censored (Control)", 
    "DGF.Cr loss_fu" = "Censored (SGF)", 
    "DGF.HD loss_fu" = "Censored (DGF)")
    ) %>% 
  separate(id, c("Group", "Death", "Graft Loss", "Censored"), ":") 

ciplotdat = ciplotdat %>%
  mutate(split = 
           case_when(Group == "Death (control)" |
                      Group == "Death (SGF)" |
                      Group == "Death (DGF)" ~ "Death",
                     Group == "Graft loss (Control)" |
                      Group == "Graft loss (SGF)" |
                      Group == "Graft loss (DGF)" ~ "Graft Loss",
                     Group == "Censored (Control)" |
                      Group == "Censored (SGF)" |
                      Group == "Censored (DGF)" ~ "Censored"))

ciplotdat = ciplotdat %>%
  mutate(group1 = 
           case_when(Group == "Death (control)" |
                      Group == "Graft loss (Control)" |
                      Group == "Censored (Control)" ~ "Control",
                    Group == "Death (SGF)" |
                      Group == "Graft loss (SGF)" |
                      Group == "Censored (SGF)" ~ "SGF",
                    Group == "Death (DGF)" |
                      Group == "Graft loss (DGF)" |
                      Group == "Censored (DGF)" ~ "DGF"))

ciplotdat$split = as.factor(ciplotdat$split)
ciplotdat$split = factor(ciplotdat$split, levels = c("Death", "Graft Loss", "Censored"))

levels(ciplotdat$split)

tiff("figures/km.outcomes.primary03.tiff", units = "cm", height = 6, width = 20, res = 300)
ggplot(ciplotdat, aes(x = time, y = est, color = group1)) +
  geom_step(lwd = 1.2, aes(fill = group1))  +
  ylim(c(0, 0.2)) + xlim(c(0,120))+
  theme_classic() +
  theme(plot.title = element_text(size = 14),
        legend.title = element_blank(),
        legend.position = "right") +
  labs(x = "Months", 
       y = "Cumulative incidence",
       title = "Primary outcomes, competing risks")+ facet_wrap(~split, ncol = 3)
dev.off()


######################
# comp risk: FOR DGF only
######################
dgf.filter1 = dgf.filter[dgf.filter$dgf_define != "DGF.Cr",]
dgf.filter1$dgf_define = droplevels(dgf.filter1$dgf_define)
dgf.filter1$event_all = ifelse(is.na(dgf.filter1$event_all), "Nil", dgf.filter1$event_all)

comp.risk = cuminc(ftime = dgf.filter1$time.in.study, 
              fstatus = dgf.filter1$event_all, 
              group = dgf.filter1$dgf_define,
              cencode = "Nil")

ciplotdat = comp.risk %>% 
  list_modify("Tests" = NULL) %>% 
  map_df(`[`, c("time", "est"), .id = "id") %>% 
  mutate(id = recode(
    id, 
    "Control death" = "Death (control)", 
    "DGF.Cr death" = "Death (SGF)", 
    "DGF.HD death" = "Death (DGF)", 
    "Control graft_loss_only" = "Graft loss (Control)", 
    "DGF.Cr graft_loss_only" = "Graft loss (SGF)", 
    "DGF.HD graft_loss_only" = "Graft loss (DGF)", 
    "Control loss_fu" = "Censored (Control)", 
    "DGF.Cr loss_fu" = "Censored (SGF)", 
    "DGF.HD loss_fu" = "Censored (DGF)")
    ) %>% 
  separate(id, c("Group", "Death", "Graft Loss", "Censored"), ":") 

ciplotdat = ciplotdat %>%
  mutate(split = 
           case_when(Group == "Death (control)" |
                      Group == "Death (DGF)" ~ "Death",
                     Group == "Graft loss (Control)" |
                      Group == "Graft loss (DGF)" ~ "Graft Loss",
                     Group == "Censored (Control)" |
                      Group == "Censored (DGF)" ~ "Censored"))

ciplotdat = ciplotdat %>%
  mutate(group1 = 
           case_when(Group == "Death (control)" |
                      Group == "Graft loss (Control)" |
                      Group == "Censored (Control)" ~ "Control",
                    Group == "Death (DGF)" |
                      Group == "Graft loss (DGF)" |
                      Group == "Censored (DGF)" ~ "DGF"))

ciplotdat$split = as.factor(ciplotdat$split)
ciplotdat$split = factor(ciplotdat$split, levels = c("Death", "Graft Loss", "Censored"))


tiff("figures/km.outcomes.primary03a.tiff", units = "cm", height = 10, width = 18, res = 300)
ggplot(ciplotdat, aes(x = time, y = est, color = group1)) +
  geom_step(lwd = 1.2, aes(fill = group1))  +
  ylim(c(0, 0.25)) + xlim(c(0,120))+
  theme_classic() +
  theme(plot.title = element_text(size = 14),
        legend.title = element_blank(),
        legend.position = "bottom") +
  labs(x = "Months", 
       y = "Cumulative incidence",
       title = "Primary outcomes, competing risks")+ facet_wrap(~split, ncol = 3)
dev.off()



######################
# comp risk: BPAR
######################

composite.events1 = survival::survfit(Surv(time = dgf.filter$bpar.time, 
                                           event = dgf.filter$event.main) 
                                      ~ bpar_any, data = dgf.filter)

coxph(Surv(dgf.filter$bpar.time, dgf.filter$event.main) ~ bpar_any, 
      data = dgf.filter) %>%  gtsummary::tbl_regression(exp = TRUE) 

p.comp1 = p.surv(composite.events1, 
                c("Control", "BPAR"), "npg", 
                "BPAR vs composite outcomes")

tiff("figures/km.outcomes.primary01.bpar.tiff", units = "cm", height = 10, width = 15, res = 300)
p.comp1$plot/p.comp1$table + plot_layout(heights = c(3,1))
dev.off()

dgf.filter1 = dgf.filter
comp.risk = cuminc(ftime = dgf.filter1$bpar.time, 
              fstatus = dgf.filter1$event_all, 
              group = dgf.filter1$bpar_any,
              cencode = 0)

ciplotdat = comp.risk %>% 
  list_modify("Tests" = NULL) %>% 
  map_df(`[`, c("time", "est"), .id = "id") %>% 
  mutate(id = recode(
    id, 
    "0 death" = "Death (Control)", 
    "1 death" = "Death (BPAR)", 
    "0 graft_loss_only" = "Graft loss (Control)", 
    "1 graft_loss_only" = "Graft loss (BPAR)", 
    "0 loss_fu" = "Censored (Control)", 
    "1 loss_fu" = "Censored (BPAR)")
    ) %>% 
  separate(id, c("Group", "Death", "Graft Loss", "Censored"), ":") 

ciplotdat = ciplotdat %>%
  mutate(split = 
           case_when(Group == "Death (Control)" |
                      Group == "Death (BPAR)" ~ "Death",
                     Group == "Graft loss (Control)" |
                      Group == "Graft loss (BPAR)" ~ "Graft Loss",
                     Group == "Censored (Control)" |
                      Group == "Censored (BPAR)" ~ "Censored"))

ciplotdat = ciplotdat %>%
  mutate(group1 = 
           case_when(Group == "Death (Control)" |
                      Group == "Graft loss (Control)" |
                      Group == "Censored (Control)" ~ "Control",
                    Group == "Death (BPAR)" |
                      Group == "Graft loss (BPAR)" |
                      Group == "Censored (BPAR)" ~ "BPAR"))

ciplotdat$split = as.factor(ciplotdat$split)
ciplotdat$split = factor(ciplotdat$split, levels = c("Death", "Graft Loss", "Censored"))
ciplotdat = ciplotdat[!is.na(ciplotdat$split),]

tiff("figures/km.outcomes.primary03.bpar.tiff", units = "cm", height = 10, width = 18, res = 300)
ggplot(ciplotdat, aes(x = time, y = est, color = group1)) +
  geom_step(lwd = 1.2, aes(fill = group1))  +
  ylim(c(0, 0.25)) + xlim(c(0,120))+
  theme_classic() +
  theme(plot.title = element_text(size = 14),
        legend.title = element_blank(),
        legend.position = "bottom") +
  labs(x = "Months", 
       y = "Cumulative incidence",
       title = "Primary outcomes, competing risks")+ facet_wrap(~split, ncol = 3)
dev.off()

comp.risk$Tests


#########################


##### CAUSE SPECIFIC HR for competiting risk analysis ####

# Cause-specific hazard approach for graft loss vs DGF
dgf.filter2 = dgf.filter1[, c("time.in.study" , "event_all" , "dgf_define")]
chr_fit =coxph(Surv(time.in.study, ifelse(event_all == "graft_loss_only", 1, 0)) ~ dgf_define, 
    data = dgf.filter2)
broom::tidy(chr_fit, exp = TRUE) %>% kable()
gtsummary::tbl_regression(chr_fit, exp = TRUE)

# graft loss vs bpar <1m
dgf.filter2 = dgf.filter1[, c("time.in.study" , "event_all" , "dgf_define", "bpar_0_1m")]
chr_fit =coxph(Surv(time.in.study, ifelse(event_all == "graft_loss_only", 1, 0)) ~ 
                 dgf_define + bpar_0_1m,  data = dgf.filter2)
broom::tidy(chr_fit, exp = TRUE) %>% kable()
gtsummary::tbl_regression(chr_fit, exp = TRUE)

# graft loss vs bpar any
dgf.filter2 = dgf.filter1[, c("time.in.study" , "event_all" , "dgf_define", "bpar_any")]
chr_fit =coxph(Surv(time.in.study, ifelse(event_all == "graft_loss_only", 1, 0)) ~ bpar_any, 
    data = dgf.filter2)
broom::tidy(chr_fit, exp = TRUE) %>% kable()
gtsummary::tbl_regression(chr_fit, exp = TRUE)

# censor vs bpar <1m
dgf.filter2 = dgf.filter1[, c("time.in.study" , "event_all" , "dgf_define", "bpar_any")]
chr_fit =coxph(Surv(time.in.study, ifelse(event_all == "loss_fu", 1, 0)) ~  bpar_any, 
    data = dgf.filter2)
broom::tidy(chr_fit, exp = TRUE) %>% kable()
gtsummary::tbl_regression(chr_fit, exp = TRUE)

# death vs bpar <1m
dgf.filter2 = dgf.filter1[, c("time.in.study" , "event_all" , "dgf_define", "bpar_any")]
chr_fit =coxph(Surv(time.in.study, ifelse(event_all == "death", 1, 0)) ~ bpar_any, 
    data = dgf.filter2)
broom::tidy(chr_fit, exp = TRUE) %>% kable()
gtsummary::tbl_regression(chr_fit, exp = TRUE)

# graft loss vs bpar <3m
dgf.filter2 = dgf.filter[, c("time.in.study" , "event_all" , "dgf_define", "bparupto03m")]
chr_fit =coxph(Surv(time.in.study, ifelse(event_all == "graft_loss_only", 1, 0)) ~  bparupto03m, 
    data = dgf.filter2)
broom::tidy(chr_fit, exp = TRUE) %>% kable()
gtsummary::tbl_regression(chr_fit, exp = TRUE)

# graft loss vs subclin
dgf.filter2 = dgf.filter1[, c("time.in.study" , "event_all" , "dgf_define", "subclin_any")]
chr_fit =coxph(Surv(time.in.study, ifelse(event_all == "graft_loss_only", 1, 0)) ~  subclin_any, 
    data = dgf.filter2)
broom::tidy(chr_fit, exp = TRUE) %>% kable()
gtsummary::tbl_regression(chr_fit, exp = TRUE)

# death vs dgf
dgf.filter2 = dgf.filter1[, c("time.in.study" , "event_all" , "dgf_define")]
chr_fit =coxph(Surv(time.in.study, ifelse(event_all == "death", 1, 0)) ~ dgf_define, 
    data = dgf.filter2)
broom::tidy(chr_fit, exp = TRUE) %>% kable()
gtsummary::tbl_regression(chr_fit, exp = TRUE)

# censor vs dgf
dgf.filter2 = dgf.filter1[, c("time.in.study" , "event_all" , "dgf_define")]
chr_fit =coxph(Surv(time.in.study, ifelse(event_all == "loss_fu", 1, 0)) ~ dgf_define, 
    data = dgf.filter2)
broom::tidy(chr_fit, exp = TRUE) %>% kable()
gtsummary::tbl_regression(chr_fit, exp = TRUE)



```

KM: SEt1
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
# Composite outcomes - combined criteria
dgf.filter = dgf.filter[dgf.filter$dgf_define != "DGF.Cr",]
dgf.filter$dgf_define = droplevels(dgf.filter$dgf_define)

dgf.filter = dgf.filter%>%
  mutate(dgf.IFTA = 
           case_when(dgf_define != "DGF.HD" & as.numeric(IFTA_score_12m_meena1)<2 ~ "Control",
                     dgf_define != "DGF.HD" & as.numeric(IFTA_score_12m_meena1)>1 ~ "Control.IFTA",
                     dgf_define == "DGF.HD" & as.numeric(IFTA_score_12m_meena1)<2 ~ "DGF",
                     dgf_define == "DGF.HD" & as.numeric(IFTA_score_12m_meena1)>1 ~ "DGF.IFTA"))

dgf.filter = dgf.filter%>%
  mutate(dgf.mgfr = 
           case_when(dgf_define != "DGF.HD" & mgfr_12m >=45 ~ "Control",
                     dgf_define != "DGF.HD" & mgfr_12m <45  ~ "Control.low",
                     dgf_define == "DGF.HD" & mgfr_12m >=45 ~ "DGF",
                     dgf_define == "DGF.HD" & mgfr_12m <45 ~ "DGF.low"))

composite.events1 = survival::survfit(Surv(time = dgf.filter$time.in.study, 
                                           event = dgf.filter$event.main) 
                                      ~ dgf.IFTA, data = dgf.filter)

p.comp1 = p.surv(composite.events1, 
                c("Control", "Control.IFTA>2", "DGF", "DGF.IFTA>2"), "npg", 
                "DGF/IFTA vs composite outcomes")

composite.events2 = survival::survfit(Surv(time = dgf.filter$time_to_graft_loss1, 
                                           event = dgf.filter$graft.loss.noncensored) 
                                      ~ dgf.IFTA, data = dgf.filter)

p.comp2 = p.surv(composite.events2,
                c("Control", "Control.IFTA>2", "DGF", "DGF.IFTA>2"), "npg", 
                "DGF/IFTA vs graft loss")

composite.events3 = survival::survfit(Surv(time = dgf.filter$time.in.study, 
                                           event = dgf.filter$event.main) 
                                      ~ dgf.mgfr, data = dgf.filter)

p.comp3 = p.surv(composite.events3, 
                c("Control", "Control.mGFR<45", "DGF", "DGF.mGFR<45"), "npg", 
                "DGF/mGFR vs composite outcomes")

composite.events4 = survival::survfit(Surv(time = dgf.filter$time_to_graft_loss1, 
                                           event = dgf.filter$graft.loss.noncensored) 
                                      ~ dgf.mgfr, data = dgf.filter)

p.comp4 = p.surv(composite.events4, 
                c("Control", "Control.mGFR<45", "DGF", "DGF.mGFR<45"), "npg", 
                "DGF/mGFR vs graft loss")

tiff("figures/km.1.tiff", units = "cm", res = 300, height = 12, width = 18)
p.comp1
dev.off()

tiff("figures/km.2.tiff", units = "cm", res = 300, height = 12, width = 18)
p.comp2
dev.off()

tiff("figures/km.3.tiff", units = "cm", res = 300, height = 12, width = 18)
p.comp3
dev.off()

tiff("figures/km.4.tiff", units = "cm", res = 300, height = 12, width = 18)
p.comp4
dev.off()


```



Simple prediction model - to 12m mGFR (slight diff for numeric outcomes)
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
#############################################
# Select reasonable variables for the outcome
#############################################
set.seed(42)
# with only clinical variables
variables.preTx = c("dgf_define", "tx_yr.2016", "tx_age", 
                   "htn_pre", "ihd_pre", "dm_pre", 
                   "renal_dx", "donor_criteria",  "organ_type",
                   "HLA_mm", "dsa1_pre", "dsa2_pre", 
                   "graft_number", "donor_age", "gender",
                   "donor_inotropes", "donor_terminal_urea",
                   "donor_terminal_cr", "donor_intubated_days",
                   "cit_min", "induct_bas")

variables.postTx = c("IFTA_score_12m_meena1", "iIFTA_score_12m_meena1", 
                     "IFTA_score_3m_meena1", "iIFTA_score_3m_meena1", 
                     "ci_0m", "ct_0m",  "i_1m", "t_1m", "ptc_1m", 
                     "cr_d2", "tac_d2", "cr_1m",  
                     "tac_1m", "tac_3m","cr_3m", "mgfr_3m", "mgfr_12m",
                     "bpar_0_1m", "bpar_1m", "bpar_1_3m", "bpar_3m", 
                     "bpar_3_12m", "bpar_any", "subclin_any", 
                     "Subclin_0_1m", "subclin_1m","subclin_1_3m", 
                     "subclin_3m","subclin_3_12m",
                     "transfusion_01m", "bk_viremia", "bkvan")

keep = c(variables.preTx, variables.postTx)
dgf.filter.pred = dgf.filter[,keep]

# Data cleaning for downstream
dgf.filter.pred = dgf.filter.pred[dgf.filter.pred$dgf_define != "DGF.Cr",]
dgf.filter.pred$dgf_define = droplevels(dgf.filter.pred$dgf_define)
dgf.filter.pred$IFTA_score_12m_meena1 = as.numeric(dgf.filter.pred$IFTA_score_12m_meena1)
dgf.filter.pred$iIFTA_score_12m_meena1 = as.numeric(dgf.filter.pred$iIFTA_score_12m_meena1)
dgf.filter.pred$IFTA_score_3m_meena1 = as.numeric(dgf.filter.pred$IFTA_score_3m_meena1)
dgf.filter.pred$iIFTA_score_3m_meena1 = as.numeric(dgf.filter.pred$iIFTA_score_3m_meena1)

dgf.filter.pred$subclin_3_12m = 
  ifelse(is.na(dgf.filter.pred$subclin_3_12m), 
         levels(as.factor(dgf.filter.pred$subclin_3_12m))[1], 
         dgf.filter.pred$subclin_3_12m)


# Prepare the factor/outcome
dgf.filter.pred = dgf.filter.pred[!is.na(dgf.filter.pred$mgfr_12m),]
dgf.filter.pred$mgfr_12m = ifelse(dgf.filter.pred$mgfr_12m< 45, "low", "high")
sum(dgf.filter.pred$mgfr_12m=="low")

##########################################################
# Split data, p = %, define the outcome factor here in Y
split1 = createDataPartition(dgf.filter.pred$mgfr_12m , p= 0.5, list = FALSE)

train.data = dgf.filter.pred[split1,]
Y_train = train.data$mgfr_12m

test.data = dgf.filter.pred[-split1,]
Y_test = test.data$mgfr_12m

trainData = c()
testData = c()

trainData = train.data[,2:ncol(train.data)]
testData = test.data[,2:ncol(test.data)]

###################################
########### TRAIN SET #############
###################################
# Impute missing
preProcess_missingdata_model = preProcess(trainData, method='knnImpute')
trainData = predict(preProcess_missingdata_model, newdata = trainData)
anyNA(trainData)

trainData$mgfr_12m = Y_train

# ONE-HOT encoding:  Creating dummy variables is converting a categorical variable to as many binary variables as here are categories.
dummies_model = dummyVars(mgfr_12m ~ ., data=trainData)
trainData_mat = predict(dummies_model, newdata = trainData) # Create the dummy variable
trainData = data.frame(trainData_mat) # Convert to dataframe

# Transform/scale
x = trainData
preprocessparam = preProcess(x, method = c("center", "scale"))

X_train = predict(preprocessparam, x)
X_train = as.data.frame(apply(X_train, 2, function(x) as.numeric(x)))
data.train = cbind(mgfr_12m = ifelse(Y_train == levels(as.factor(Y_train))[1], 1, 2), X_train)
data.train = as.data.frame(apply(data.train, 2, function(x) as.numeric(x)))

##################################
########### TEST SET #############
##################################
# Impute missing
preProcess_missingdata_model = preProcess(testData, method='knnImpute')
testData = predict(preProcess_missingdata_model, newdata = testData)
anyNA(testData)

testData$mgfr_12m = as.factor(Y_test)

# ONE-HOT encoding:  Creating dummy variables is converting a categorical variable to as many binary variables as here are categories.
dummies_model = dummyVars(mgfr_12m ~ ., data=testData)
testData_mat = predict(dummies_model, newdata = testData)  # Create the dummy variable
testData = data.frame(testData_mat) # Convert to dataframe

# Transform/scale
x = testData
preprocessparam = preProcess(x, method = c("center", "scale"))

X_test = predict(preprocessparam, x)
X_test = as.data.frame(apply(X_test, 2, function(x) as.numeric(x)))
data.test = cbind(mgfr_12m = ifelse(Y_test == levels(as.factor(Y_test))[1], 1, 2), X_test)
data.test = as.data.frame(apply(data.test, 2, function(x) as.numeric(x)))

####################################################
######### MODEL ####################################
####################################################
# Penalised logistic regression (lasso) with 10-fold CV)
train.1 = train(mgfr_12m ~.,
                data = data.train,
                preProc = c("center", "scale"),
                method = "lasso",
                trControl = trainControl(method = "repeatedcv", 
                                         number = 10, 
                                         repeats = 10))
plot(train.1)
ggplot(train.1)

tiff("figures/pen.log.regression.mgfr12m.variable.importance.tiff", 
     units = "cm", height = 30, width = 12, res = 300)
ggplot(varImp(train.1))
dev.off()


# Apply to test set
predictions = predict(train.1, X_test)
auc(Y_test, predictions)
rocs = performance(prediction(predictions, Y_test), 'tpr', 'fpr')

tiff("figures/pen.log.regression.mgfr12m.roc.tiff", 
     units = "cm", height = 10, width = 12, res = 300)
plot(rocs, curve = TRUE) + abline(a = 0, b = 1)
dev.off()

# above + model hyperparameters
train.2 = train(mgfr_12m ~.,
                data = data.train,
                preProc = c("center", "scale"),
                method = "lasso",
                trControl = trainControl(method = "repeatedcv", 
                                         number = 10, 
                                         repeats = 10, 
                                         search = "random"))
plot(train.2)
ggplot(varImp(train.2))

# Apply to test set
predictions = predict(train.2, X_test)
rocs = performance(prediction(predictions, Y_test), 'tpr', 'fpr')
plot(rocs, curve = TRUE) + abline(a = 0, b = 1)
auc(Y_test, predictions)

testData
trainData
```


Clin + RNA prediction model - to 12m mGFR (slight diff for numeric outcomes)
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
#############################################
# Select reasonable variables for the outcome
#############################################
set.seed(42)
# with only clinical variables
variables.preTx = c("dgf_define", "tx_yr.2016", "tx_age", 
                   "htn_pre", "ihd_pre", "dm_pre", 
                   "renal_dx", "donor_criteria",  "organ_type",
                   "HLA_mm", "dsa1_pre", "dsa2_pre", 
                   "graft_number", "donor_age", "gender",
                   "donor_inotropes", "donor_terminal_urea",
                   "donor_terminal_cr", "donor_intubated_days",
                   "cit_min", "induct_bas")

variables.postTx = c("IFTA_score_12m_meena1", "iIFTA_score_12m_meena1", 
                     "IFTA_score_3m_meena1", "iIFTA_score_3m_meena1", 
                     "ci_0m", "ct_0m",  "i_1m", "t_1m", "ptc_1m", 
                     "cr_d2", "tac_d2", "cr_1m",  
                     "tac_1m", "tac_3m","cr_3m", "mgfr_3m", "mgfr_12m",
                     "bpar_0_1m", "bpar_1m", "bpar_1_3m", "bpar_3m", 
                     "bpar_3_12m", "bpar_any", "subclin_any", 
                     "Subclin_0_1m", "subclin_1m","subclin_1_3m", 
                     "subclin_3m","subclin_3_12m",
                     "transfusion_01m", "bk_viremia", "bkvan")

keep = c(variables.preTx, variables.postTx)
dgf.filter.pred = dgf.filter[,keep]

# Data cleaning for downstream
dgf.filter.pred = dgf.filter.pred[dgf.filter.pred$dgf_define != "DGF.Cr",]
dgf.filter.pred$dgf_define = droplevels(dgf.filter.pred$dgf_define)
dgf.filter.pred$IFTA_score_12m_meena1 = as.numeric(dgf.filter.pred$IFTA_score_12m_meena1)
dgf.filter.pred$iIFTA_score_12m_meena1 = as.numeric(dgf.filter.pred$iIFTA_score_12m_meena1)
dgf.filter.pred$IFTA_score_3m_meena1 = as.numeric(dgf.filter.pred$IFTA_score_3m_meena1)
dgf.filter.pred$iIFTA_score_3m_meena1 = as.numeric(dgf.filter.pred$iIFTA_score_3m_meena1)

dgf.filter.pred$subclin_3_12m = 
  ifelse(is.na(dgf.filter.pred$subclin_3_12m), 
         levels(as.factor(dgf.filter.pred$subclin_3_12m))[1], 
         dgf.filter.pred$subclin_3_12m)


# Prepare the factor/outcome
dgf.filter.pred = dgf.filter.pred[!is.na(dgf.filter.pred$mgfr_12m),]
dgf.filter.pred$mgfr_12m = ifelse(dgf.filter.pred$mgfr_12m< 45, "low", "high")


###########################
# Add Genomics data
##########################


##########################################################
# Split data, p = %, define the outcome factor here in Y
split1 = createDataPartition(dgf.filter.pred$mgfr_12m , p= 0.5, list = FALSE)

train.data = dgf.filter.pred[split1,]
Y_train = train.data$mgfr_12m

test.data = dgf.filter.pred[-split1,]
Y_test = test.data$mgfr_12m

trainData = c()
testData = c()

trainData = train.data[,2:ncol(train.data)]
testData = test.data[,2:ncol(test.data)]

###################################
########### TRAIN SET #############
###################################
# Impute missing
preProcess_missingdata_model = preProcess(trainData, method='knnImpute')
trainData = predict(preProcess_missingdata_model, newdata = trainData)
anyNA(trainData)

trainData$mgfr_12m = Y_train

# ONE-HOT encoding:  Creating dummy variables is converting a categorical variable to as many binary variables as here are categories.
dummies_model = dummyVars(mgfr_12m ~ ., data=trainData)
trainData_mat = predict(dummies_model, newdata = trainData) # Create the dummy variable
trainData = data.frame(trainData_mat) # Convert to dataframe

# Transform/scale
x = trainData
preprocessparam = preProcess(x, method = c("center", "scale"))

X_train = predict(preprocessparam, x)
X_train = as.data.frame(apply(X_train, 2, function(x) as.numeric(x)))
data.train = cbind(mgfr_12m = ifelse(Y_train == levels(as.factor(Y_train))[1], 1, 2), X_train)
data.train = as.data.frame(apply(data.train, 2, function(x) as.numeric(x)))

##################################
########### TEST SET #############
##################################
# Impute missing
preProcess_missingdata_model = preProcess(testData, method='knnImpute')
testData = predict(preProcess_missingdata_model, newdata = testData)
anyNA(testData)

testData$mgfr_12m = as.factor(Y_test)

# ONE-HOT encoding:  Creating dummy variables is converting a categorical variable to as many binary variables as here are categories.
dummies_model = dummyVars(mgfr_12m ~ ., data=testData)
testData_mat = predict(dummies_model, newdata = testData)  # Create the dummy variable
testData = data.frame(testData_mat) # Convert to dataframe

# Transform/scale
x = testData
preprocessparam = preProcess(x, method = c("center", "scale"))

X_test = predict(preprocessparam, x)
X_test = as.data.frame(apply(X_test, 2, function(x) as.numeric(x)))
data.test = cbind(mgfr_12m = ifelse(Y_test == levels(as.factor(Y_test))[1], 1, 2), X_test)
data.test = as.data.frame(apply(data.test, 2, function(x) as.numeric(x)))

####################################################
######### MODEL ####################################
####################################################
# Penalised logistic regression (lasso) with 10-fold CV)
train.1 = train(mgfr_12m ~.,
                data = data.train,
                preProc = c("center", "scale"),
                method = "lasso",
                trControl = trainControl(method = "repeatedcv", 
                                         number = 10, 
                                         repeats = 10))
plot(train.1)
ggplot(varImp(train.1))
ggplot(train.1)

# Apply to test set
predictions = predict(train.1, X_test)
auc(Y_test, predictions)
rocs = performance(prediction(predictions, Y_test), 'tpr', 'fpr')
plot(rocs, curve = TRUE) + abline(a = 0, b = 1)

# above + model hyperparameters
train.2 = train(mgfr_12m ~.,
                data = data.train,
                preProc = c("center", "scale"),
                method = "lasso",
                trControl = trainControl(method = "repeatedcv", 
                                         number = 10, 
                                         repeats = 10, 
                                         search = "random"))
plot(train.2)
ggplot(varImp(train.2))

# Apply to test set
predictions = predict(train.2, X_test)
rocs = performance(prediction(predictions, Y_test), 'tpr', 'fpr')
plot(rocs, curve = TRUE) + abline(a = 0, b = 1)
auc(Y_test, predictions)

```


Simple prediction model - to IFTA >=2 at 12 months
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
#############################################
# Select reasonable variables for the outcome
#############################################
set.seed(123)
# with only clinical variables
variables.preTx = c("dgf_define", "tx_yr.2016", "tx_age", 
                   "htn_pre", "ihd_pre", "dm_pre", 
                   "renal_dx", "donor_criteria",
                   "HLA_mm", "dsa1_pre", "dsa2_pre", 
                   "graft_number", "donor_age", "gender",
                   "donor_inotropes", "donor_terminal_urea",
                   "donor_terminal_cr", "donor_intubated_days",
                   "cit_min", "induct_bas")

variables.postTx = c("IFTA_score_12m_meena1", 
                     "IFTA_score_3m_meena1", "iIFTA_score_3m_meena1", 
                     "ci_0m", "ct_0m",  "i_1m", "t_1m", "ptc_1m", 
                     "cr_d2", "tac_d2", "cr_1m",  
                     "tac_1m", "tac_3m","cr_3m", "mgfr_3m", 
                     "bpar_0_1m", "bpar_1m", "bpar_1_3m", "bpar_3m", 
                     "bpar_3_12m", "bpar_any", "subclin_any", 
                     "Subclin_0_1m", "subclin_1m","subclin_1_3m", 
                     "subclin_3m", "mpred_0_1m", 
                     "mpred_1m", "mpred_1_3m", "mpred_3m", 
                     "transfusion_01m", "bk_viremia", "bkvan")

keep = c(variables.preTx, variables.postTx)
dgf.filter.pred = dgf.filter[,keep]

# Data cleaning for downstream
dgf.filter.pred = dgf.filter.pred[dgf.filter.pred$dgf_define != "DGF.Cr",]
dgf.filter.pred$dgf_define = droplevels(dgf.filter.pred$dgf_define)
dgf.filter.pred$IFTA_score_12m_meena1 = as.numeric(dgf.filter.pred$IFTA_score_12m_meena1)
dgf.filter.pred$IFTA_score_3m_meena1 = as.numeric(dgf.filter.pred$IFTA_score_3m_meena1)
dgf.filter.pred$iIFTA_score_3m_meena1 = as.numeric(dgf.filter.pred$iIFTA_score_3m_meena1)


# Prepare the factor/outcome
dgf.filter.pred = dgf.filter.pred[!is.na(dgf.filter.pred$IFTA_score_12m_meena1),]
dgf.filter.pred$IFTA.cutoff = ifelse(dgf.filter.pred$IFTA_score_12m_meena1 > 1, "high", "low")

dgf.filter.pred = subset(dgf.filter.pred, select = -c(IFTA_score_12m_meena1))
sum(dgf.filter.pred$IFTA.cutoff == "high")
##########################################################
# Split data, p = %, define the outcome factor here in Y
split1 = createDataPartition(dgf.filter.pred$IFTA.cutoff , p= 0.5, list = FALSE)

train.data = dgf.filter.pred[split1,]
Y_train = train.data$IFTA.cutoff

test.data = dgf.filter.pred[-split1,]
Y_test = test.data$IFTA.cutoff

trainData = c()
testData = c()

trainData = train.data[,2:ncol(train.data)]
testData = test.data[,2:ncol(test.data)]

###################################
########### TRAIN SET #############
###################################
# Impute missing
preProcess_missingdata_model = preProcess(trainData, method='knnImpute')
trainData = predict(preProcess_missingdata_model, newdata = trainData)
anyNA(trainData)

trainData$IFTA.cutoff = Y_train

# ONE-HOT encoding:  Creating dummy variables is converting a categorical variable to as many binary variables as here are categories.
dummies_model = dummyVars(IFTA.cutoff ~ ., data=trainData)
trainData_mat = predict(dummies_model, newdata = trainData) # Create the dummy variable
trainData = data.frame(trainData_mat) # Convert to dataframe
anyNA(trainData)

# Transform/scale
x = trainData
preprocessparam = preProcess(x, method = c("center", "scale"))

X_train = predict(preprocessparam, x)
X_train = as.data.frame(apply(X_train, 2, function(x) as.numeric(x)))
data.train = cbind(IFTA.cutoff = ifelse(Y_train == levels(as.factor(Y_train))[1], 1, 2), X_train)
data.train = as.data.frame(apply(data.train, 2, function(x) as.numeric(x)))
# ggpairs(data.train, aes(colour = IFTA.cutoff))

##################################
########### TEST SET #############
##################################
# Impute missing
preProcess_missingdata_model = preProcess(testData, method='knnImpute')
testData = predict(preProcess_missingdata_model, newdata = testData)
anyNA(testData)

testData$IFTA.cutoff = as.factor(Y_test)

# ONE-HOT encoding:  Creating dummy variables is converting a categorical variable to as many binary variables as here are categories.
dummies_model = dummyVars(IFTA.cutoff ~ ., data=testData)
testData_mat = predict(dummies_model, newdata = testData)  # Create the dummy variable
testData = data.frame(testData_mat) # Convert to dataframe

# Transform/scale
x = testData
preprocessparam = preProcess(x, method = c("center", "scale"))

X_test = predict(preprocessparam, x)
X_test = as.data.frame(apply(X_test, 2, function(x) as.numeric(x)))
data.test = cbind(IFTA.cutoff = ifelse(Y_test == levels(as.factor(Y_test))[1], 1, 2), X_test)
data.test = as.data.frame(apply(data.test, 2, function(x) as.numeric(x)))

####################################################
######### MODEL ####################################
####################################################
# Penalised logistic regression (lasso) with 10-fold CV)
train.1 = train(IFTA.cutoff ~.,
                data = data.train,
                preProc = c("center", "scale"),
                method = "lasso",
                trControl = trainControl(method = "repeatedcv", 
                                         number = 10, 
                                         repeats = 10))

assess.lasso = assess.glmnet(train.1, 
                             newx = as.matrix(data.train[,2:ncol(data.train)]), 
                             newy = data.train$IFTA.cutoff)

summary(train.1$results)
plot(train.1$results)

tiff("figures/pen.log.regression.IFTA.variable.importance.tiff", units = "cm", res = 300, height = 30, width = 12)
ggplot(varImp(train.1))
dev.off()

# Apply to test set
predictions = predict(train.1, X_test)
rocs = performance(prediction(predictions, Y_test), 'tpr', 'fpr')
tiff("figures/pen.log.regression.IFTA.ROC.tiff", units = "cm", res = 300, height = 10, width = 12)
plot(rocs, curve = TRUE) + abline(a = 0, b = 1)
dev.off()
auc(Y_test, predictions)


# above + model hyperparameters
train.2 = train(IFTA.cutoff ~.,
                data = data.train,
                preProc = c("center", "scale"),
                method = "lasso",
                trControl = trainControl(method = "repeatedcv", 
                                         number = 10, 
                                         repeats = 10, 
                                         search = "random"))

summary(train.2$results)
plot(train.2$results)
ggplot(varImp(train.2))

# Apply to test set
predictions = predict(train.2, X_test)
rocs = performance(prediction(predictions, Y_test), 'tpr', 'fpr')
plot(rocs, curve = TRUE) + abline(a = 0, b = 1)
auc(Y_test, predictions)

```


### Neutrophils section


```{r,  echo=FALSE, results = FALSE, warning = FALSE, eval = FALSE}
##########################
# NEUTROPHILS 
##########################
neut.dgf = c()
neut.dgf = dgf.filter %>% select(dgf_define,
                                 organ_type,
                                 donor_criteria.dcd,
                                 neut_0m, 
                                 neut_1m, 
                                 neut_3m)
neut.dgf$donor_type = 
  case_when(neut.dgf$donor_criteria.dcd == "any DCD" ~ "DCD",
            neut.dgf$donor_criteria.dcd == "DBD" | 
              neut.dgf$donor_criteria.dcd == "DBD + ECD" ~ "DBD",
            neut.dgf$donor_criteria.dcd == "LKD" ~ "LKD")

neut.table = tbl_summary(neut.dgf,  missing = "no", by = dgf_define)
neut.table = neut.table %>% add_n() %>% 
  add_p() %>% modify_header(label = "Neutrophils") %>% bold_labels()

neut.table = tbl_summary(neut.dgf,  missing = "no", by = organ_type)
neut.table = neut.table %>% add_n() %>% 
  add_p() %>% modify_header(label = "Neutrophils") %>% bold_labels()

neut.table = tbl_summary(neut.dgf,  missing = "no", by = donor_type)
neut.table = neut.table %>% add_n() %>% 
  add_p() %>% modify_header(label = "Neutrophils") %>% bold_labels()


  # plots

p.n0 = ggplot(neut.dgf, 
            aes(x=dgf_define, y=neut_0m, colour = dgf_define))+
    geom_boxplot()+ geom_jitter() + 
  theme_bw()+ ylim(c(0,16))+
  ylab("Neutrophils/hpf")+ ggtitle("0 month")+
  stat_compare_means(comparisons = list(c("Control", "DGF.Cr"), 
                                        c("Control", "DGF.HD"), 
                                        c("DGF.Cr", "DGF.HD")), 
                     label.y = c(11,12,13)) +
  stat_compare_means(label.y = 15)
      
p.n1 = ggplot(neut.dgf, 
            aes(x=dgf_define, y=neut_1m, colour = dgf_define))+
    geom_boxplot()+ geom_jitter() + 
  theme_bw()+ ylim(c(0,16))+
  ylab("Neutrophils/hpf")+ ggtitle("1 month")+
  stat_compare_means(comparisons = list(c("Control", "DGF.Cr"), 
                                        c("Control", "DGF.HD"), 
                                        c("DGF.Cr", "DGF.HD")), 
                     label.y = c(11,12,13)) +
  stat_compare_means(label.y = 15)

p.n3 = ggplot(neut.dgf, 
            aes(x=dgf_define, y=neut_3m, colour = dgf_define))+
    geom_boxplot()+ geom_jitter() + 
  theme_bw()+ ylim(c(0,16))+
  ylab("Neutrophils/hpf")+ ggtitle("3 month")+
  stat_compare_means(comparisons = list(c("Control", "DGF.Cr"), 
                                        c("Control", "DGF.HD"), 
                                        c("DGF.Cr", "DGF.HD")), 
                     label.y = c(11,12,13)) +
  stat_compare_means(label.y = 15)

ggarrange(p.n0, p.n1, p.n3, nrow = 3)

p.dgf = ggarrange(p.n0, p.n1, p.n3, ncol = 3, nrow = 1, common.legend = TRUE, legend = "right")

p.n0a = ggplot(neut.dgf, 
            aes(x=donor_type, y=neut_0m, colour = donor_type))+
    geom_boxplot()+ geom_jitter() + 
  theme_bw()+ ylim(c(0,15))+
  ylab("Neutrophils/hpf")+ ggtitle("0 month")+
  stat_compare_means(comparisons = list(c("Control", "DGF.Cr"), 
                                        c("Control", "DGF.HD"), 
                                        c("DGF.Cr", "DGF.HD")), 
                     label.y = c(11,12,13)) +
  stat_compare_means(label.y = 14)

p.n1a = ggplot(neut.dgf, 
            aes(x=donor_type, y=neut_1m, colour = donor_type))+
    geom_boxplot()+ geom_jitter() + 
  theme_bw()+ ylim(c(0,15))+
  ylab("Neutrophils/hpf")+ ggtitle("1 month")+
  stat_compare_means(comparisons = list(c("Control", "DGF.Cr"), 
                                        c("Control", "DGF.HD"), 
                                        c("DGF.Cr", "DGF.HD")), 
                     label.y = c(11,12,13)) +
  stat_compare_means(label.y = 14)

p.n3a = ggplot(neut.dgf, 
            aes(x=donor_type, y=neut_3m, colour = donor_type))+
    geom_boxplot()+ geom_jitter() + 
  theme_bw()+ ylim(c(0,15))+
  ylab("Neutrophils/hpf")+ ggtitle("3 month")+
  stat_compare_means(comparisons = list(c("Control", "DGF.Cr"), 
                                        c("Control", "DGF.HD"), 
                                        c("DGF.Cr", "DGF.HD")), 
                     label.y = c(11,12,13)) +
  stat_compare_means(label.y = 14)

p.donor = ggarrange(p.n0a, p.n1a, p.n3a, ncol = 3, nrow = 1, common.legend = TRUE, legend = "right")

tiff("figures/neutrophils.tiff", units = "cm", res= 300, height = 18, width = 26)
p.dgf/p.donor
dev.off()



```





### DGE of AUSCAD bulk dataset with main contrast is DGF vs control
Baseline comparisons
0m biopsies (before reperfusion - baseline expression before severe IRI)
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
# contrast DGF, accounting for batch only
dge.all = dge.DGF.HD(auscad.sorted@samples.biopsy.0m, 
                     auscad.sorted@counts.biopsy.0m, 
                     covariates = "nil", 
                     rejection.factor = FALSE)
  saveRDS(dge.all, "rds/dge.allpts.dgf.vs.control.0m.RDS")
  gc()
  
# contrast DGF, accounting for batch and organ type (ie living donors vs deceased donor)
dge.organ = dge.DGF.HD(auscad.sorted@samples.biopsy.0m, 
                     auscad.sorted@counts.biopsy.0m, 
                     covariates = "organ", 
                     rejection.factor = FALSE)
  saveRDS(dge.organ, "rds/dge.covariate.organ.dgf.vs.control.0m.RDS")
  gc()
  
dge.organ = dge.DGF.HD(auscad.sorted@samples.biopsy.0m, 
                     auscad.sorted@counts.biopsy.0m, 
                     covariates = "multi", 
                     rejection.factor = FALSE)
  saveRDS(dge.organ, "rds/dge.covariate.multi.dgf.vs.control.0m.RDS")
  gc()
  
# rejection can be: "nil", "subclin.only", "bpar.only", "any"
dge.organ.reject = dge.DGF.HD(auscad.sorted@samples.biopsy.0m, 
                       auscad.sorted@counts.biopsy.0m, 
                       covariates = "multi",
                       rejection.factor = TRUE)

  saveRDS(dge.organ.reject[1], "rds/dge.covorgan.nilreject.dgf.vs.con.0m.RDS")
  saveRDS(dge.organ.reject[2], "rds/dge.covorgan.subclin.dgf.vs.con.0m.RDS")
  saveRDS(dge.organ.reject[3], "rds/dge.covorgan.bpar.dgf.vs.con.0m.RDS")
  saveRDS(dge.organ.reject[4], "rds/dge.covorgan.anyreject.dgf.vs.con.0ml.RDS")
  gc()
  
dge.organ.reject.in1m = dge.DGF.HD.1m(auscad.sorted@samples.biopsy.0m, 
                       auscad.sorted@counts.biopsy.0m, 
                       covariates = "multi", 
                       rejection.factor = TRUE)
  saveRDS(dge.organ.reject.in1m$subclin.only, "rds/dge.covorgan.subclinEARLY.dgf.vs.con.0m.RDS")
  saveRDS(dge.organ.reject.in1m$no.reject , "rds/dge.covorgan.No.REJECT.EARLY.dgf.vs.con.0m.RDS")
  gc()
  
# short = DGF < 7 days vs control, long = DGF > 7 days vs control, between = DGF >7 vs <7 days
dge.severity = dge.DGF.severity(auscad.sorted@samples.biopsy.0m,
                                 auscad.sorted@counts.biopsy.0m,
                               covariates = "multi",
                                rejection.factor = FALSE)
  saveRDS(dge.severity, "rds/dge.covorgan.dgfseverity.0m.RDS")
  gc()
  
dge.severity.in1m = dge.DGF.severity.1m(auscad.sorted@samples.biopsy.0m,
                                 auscad.sorted@counts.biopsy.0m,
                                covariates = "multi",
                                rejection.factor = TRUE)

  saveRDS(dge.severity.in1m$subclin.only, "rds/dge.covorgan.dgfseverity.subclin.0m.RDS")
  saveRDS(dge.severity.in1m$no.reject, "rds/dge.covorgan.dgfseverity.no.rejection.0m.RDS")
  gc()
  
  

  # Slow graft function to controls
  Cr.cont = DGF.Cr.vs.Cont(auscad.sorted@samples.biopsy.0m, 
                     auscad.sorted@counts.biopsy.0m, 
                     covariates = TRUE)
  saveRDS(Cr.cont, "rds/dge.slowgraftfunction.control.0m.RDS")
  dge.summary(Cr.cont@top.tags.edgeR)
  
  # Compare to HD < 7 days
  samples = auscad.sorted@samples.biopsy.0m
  counts = auscad.sorted@counts.biopsy.0m
  samples = samples %>% filter(dgf.severity != "HD >= 7 days")
  samples = samples %>% filter(bpar_0_1m == "No BPAR")
  hd.cr = DGF.HD.vs.Cr(samples, counts, 
                     covariates = TRUE)
  saveRDS(hd.cr, "rds/dge.slowgraftfunction.dgf.short.0m.RDS")
  dge.summary(hd.cr@top.tags.edgeR)
  hd.cr@summary.numbers
  dge = go.gsea(hd.cr@top.tags.edgeR, 1)
  tiff("figures/gsea.DGF.short.vs.SGF.tiff", units = "cm", res = 300, height =9, width = 18)
  ggplot(dge$gsego.bp1, 
       aes(x = log.p.adj, y = Description)) + 
              geom_point(aes(color= p.adjust, size = Count)) +
              scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
              theme(axis.text.y = element_text(size = 10)) +
              xlab("-log10 of P-value") + ylab("Description") + 
              ggtitle("GO:GSEA") +  theme_bw() + geom_vline(xintercept=0)
  dev.off()
  
  # SGF Compare to HD 
  samples = auscad.sorted@samples.biopsy.0m
  counts = auscad.sorted@counts.biopsy.0m

  hd.cr = DGF.HD.vs.Cr(samples, counts, 
                     covariates = TRUE)
  saveRDS(hd.cr, "rds/dge.dgf.SF.0m.RDS")
  dge.summary(hd.cr@top.tags.edgeR)
    hd.cr@summary.numbers
    
  dge = go.gsea(hd.cr@top.tags.edgeR, 1)
  
  tiff("figures/gsea.DGF.vs.SGF.tiff", units = "cm", res = 300, height =9, width = 18)
  ggplot(dge$gsego.bp1, 
       aes(x = log.p.adj, y = Description)) + 
              geom_point(aes(color= p.adjust, size = Count)) +
              scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
              theme(axis.text.y = element_text(size = 10)) +
              xlab("-log10 of P-value") + ylab("Description") + 
              ggtitle("GO:GSEA") +  theme_bw() + geom_vline(xintercept=0)
  dev.off()
  
# if no BPAR/noBKVAN - DGF and 12m function, mGFR cutoff 60ml/min
  # mGFR = Cont.mGFRhigh - Cont.mGFRlow, mGFR.DGF = DGF.mGFRhigh - DGF.mGFRlow,  
  # mGFR.DGF.high = DGF.mGFRhigh - Cont.mGFRhigh, mGFR.DGF.low = DGF.mGFRlow - Cont.mGFRlow
dge.mgfr.60 = dge.DGF.mgfr(auscad.sorted@samples.biopsy.0m, 
                           auscad.sorted@counts.biopsy.0m,
                           cutoff=60)
  saveRDS(dge.mgfr.60, "rds/dge.covorgan.noreject.nobk.dgfvsMGFR60.0m.RDS")
  gc()
  
dge.mgfr.45 = dge.DGF.mgfr(auscad.sorted@samples.biopsy.0m, 
                           auscad.sorted@counts.biopsy.0m,
                           cutoff=45)
  saveRDS(dge.mgfr.45, "rds/dge.covorgan.noreject.nobk.dgfvsMGFR45.0m.RDS")
  gc()
  

```


1m biopsies 
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
# contrast DGF, accounting for batch only
dge.all = dge.DGF.HD(auscad.sorted@samples.biopsy.1m, 
                     auscad.sorted@counts.biopsy.1m,
                     covariates = "nil",
                     rejection.factor = FALSE)
  saveRDS(dge.all, "rds/dge.allpts.dgf.vs.control.1m.RDS")
  gc()

# contrast DGF, accounting for batch and organ type (ie living donors vs deceased donor)
dge.organ = dge.DGF.HD(auscad.sorted@samples.biopsy.1m, 
                     auscad.sorted@counts.biopsy.1m, 
                     covariates = "multi",
                     rejection.factor = FALSE)
  saveRDS(dge.organ, "rds/dge.covariate.organ.dgf.vs.control.1m.RDS")
  gc()
  
dge.organ.reject = dge.DGF.HD(auscad.sorted@samples.biopsy.1m, 
                     auscad.sorted@counts.biopsy.1m, 
                     covariates = "multi",
                     rejection.factor = TRUE)
  glimpse(dge.organ.reject)
  
  saveRDS(dge.organ.reject[1], "rds/dge.covorgan.nilreject.dgf.vs.con.1m.RDS")
  saveRDS(dge.organ.reject[2], "rds/dge.covorgan.subclin.dgf.vs.con.1m.RDS")
  saveRDS(dge.organ.reject[3], "rds/dge.covorgan.bpar.dgf.vs.con.1m.RDS")
  saveRDS(dge.organ.reject[4], "rds/dge.covorgan.anyreject.dgf.vs.con.1m.RDS")

dge.severity = dge.DGF.severity(auscad.sorted@samples.biopsy.1m, 
                                auscad.sorted@counts.biopsy.1m, 
                                covariates = "multi",
                                rejection.factor = FALSE)

  saveRDS(dge.severity, "rds/dge.covorgan.dgfseverity.1m.RDS")
  gc()
  
dge.severity = dge.DGF.severity(auscad.sorted@samples.biopsy.1m, 
                                auscad.sorted@counts.biopsy.1m, 
                                covariates = "multi",
                                rejection.factor = TRUE)
  saveRDS(dge.severity, "rds/dge.covorgan.dgfseverity.rejection.1m.RDS")
  gc()

```


3m biopsies 
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
# contrast DGF, accounting for batch only
dge.all = dge.DGF.HD(auscad.sorted@samples.biopsy.3m, 
                     auscad.sorted@counts.biopsy.3m, 
                     covariates = "nil",
                     rejection.factor = FALSE)
  saveRDS(dge.all, "rds/dge.allpts.dgf.vs.control.3m.RDS")
  gc()
  
# contrast DGF, accounting for batch and organ type (ie living donors vs deceased donor)
dge.organ = dge.DGF.HD(auscad.sorted@samples.biopsy.3m, 
                     auscad.sorted@counts.biopsy.3m, 
                     covariates = "multi",
                     rejection.factor = FALSE)
  saveRDS(dge.organ, "rds/dge.covariate.organ.dgf.vs.control.3m.RDS")
  gc()

# rejection can be: "nil", "subclin.only", "bpar.only", "any"
dge.organ.reject = dge.DGF.HD(auscad.sorted@samples.biopsy.3m, 
                     auscad.sorted@counts.biopsy.3m, 
                     covariates = "multi",
                     rejection.factor = TRUE)
  glimpse(dge.organ.reject)
  
  saveRDS(dge.organ.reject[1], "rds/dge.covorgan.nilreject.dgf.vs.con.3m.RDS")
  saveRDS(dge.organ.reject[2], "rds/dge.covorgan.subclin.dgf.vs.con.3m.RDS")
  saveRDS(dge.organ.reject[3], "rds/dge.covorgan.bpar.dgf.vs.con.3m.RDS")
  saveRDS(dge.organ.reject[4], "rds/dge.covorgan.anyreject.dgf.vs.con.3m.RDS")
  gc()
  
# short = DGF < 7 days vs control, long = DGF > 7 days vs control, between = DGF >7 vs <7 days
dge.severity = dge.DGF.severity(auscad.sorted@samples.biopsy.3m, 
                                auscad.sorted@counts.biopsy.3m, 
                                covariates = "multi",
                                rejection.factor = TRUE)
  
  saveRDS(dge.severity, "rds/dge.covorgan.dgfseverity.3m.RDS")
  gc()


  # DGF vs SGF
  hd.cr = DGF.HD.vs.Cr(
    auscad.sorted@samples.biopsy.3m[auscad.sorted@samples.biopsy.3m$bparupto03m == "No BPAR 0-3m",],
    auscad.sorted@counts.biopsy.3m,  covariates = TRUE)
  saveRDS(hd.cr, "rds/dge.dgf.SF.3m.RDS")
  
  # dge.summary(hd.cr@top.tags.edgeR)
  #   hd.cr@summary.numbers
    
  # dge = go.gsea(hd.cr@top.tags.edgeR, 1)
  # 
  # tiff("figures/gsea.DGF.vs.SGF.3m.tiff", units = "cm", res = 300, height =9, width = 18)
  # ggplot(dge$gsego.bp1, 
  #      aes(x = log.p.adj, y = Description)) + 
  #             geom_point(aes(color= p.adjust, size = Count)) +
  #             scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
  #             theme(axis.text.y = element_text(size = 10)) +
  #             xlab("-log10 of P-value") + ylab("Description") + 
  #             ggtitle("GO:GSEA") +  theme_bw() + geom_vline(xintercept=0)
  # dev.off()
  
  
  # Slow graft function to controls
  Cr.cont = DGF.Cr.vs.Cont(
    auscad.sorted@samples.biopsy.3m[auscad.sorted@samples.biopsy.3m$bparupto03m == "No BPAR 0-3m",], 
    auscad.sorted@counts.biopsy.3m, covariates = TRUE)
  saveRDS(Cr.cont, "rds/dge.slowgraftfunction.control.3m.RDS")
  # dge.summary(Cr.cont@top.tags.edgeR)
  # Cr.cont@summary.numbers
  # dge = go.gsea(Cr.cont@top.tags.edgeR, 1)
  # 
  # tiff("figures/gsea.SGF.controls.3m.tiff", units = "cm", res = 300, height =9, width = 18)
  # ggplot(dge$gsego.bp1, 
  #      aes(x = log.p.adj, y = Description)) + 
  #             geom_point(aes(color= p.adjust, size = Count)) +
  #             scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
  #             theme(axis.text.y = element_text(size = 10)) +
  #             xlab("-log10 of P-value") + ylab("Description") + 
  #             ggtitle("GO:GSEA") +  theme_bw() + geom_vline(xintercept=0)
  # dev.off()
  
  
```





Files saved from earlier DGE analysis - put into summary form
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
################################################
#Summary: case numbers, number of DGEs, top 10
################################################

# 0m biopsies
cases.0m.split2 = cbind(dge.all.0m@summary.numbers,
                 dge.organ.0m@summary.numbers,
                 dge.organ.no.reject.0m$no.reject@summary.numbers,
                 dge.noreject.early.0m@summary.numbers,
                 dge.organ.sub.reject.0m$subclin.only@summary.numbers,
                 dge.organ.subclinEARLY.0m@summary.numbers,
                 dge.organ.any.reject.0m$any.reject@summary.numbers,
                 dge.organ.DCD.0m@summary.numbers,
                 dge.organ.DBD.0m@summary.numbers)
colnames(cases.0m.split2) = c("all.0m", "all.organ.cov.0m", 
                              "no.rejection.0m", "no.rejection.early",
                              "subclin.0m", "sucblin.early.0m", "any.rej",
                              "DCD.only.0m", "DBD.only.0m" )

dge.0m.split2 = rbind(dge.summary(dge.all.0m@top.tags.edgeR),
                 dge.summary(dge.organ.0m@top.tags.edgeR),
                 dge.summary(dge.organ.no.reject.0m$no.reject@top.tags.edgeR),
                 dge.summary(dge.noreject.early.0m@top.tags.edgeR),
                 dge.summary(dge.organ.sub.reject.0m$subclin.only@top.tags.edgeR),
                 dge.summary(dge.organ.subclinEARLY.0m@top.tags.edgeR),
                 dge.summary(dge.organ.any.reject.0m$any.reject@top.tags.edgeR),
                 dge.summary(dge.organ.DCD.0m@top.tags.edgeR),
                 dge.summary(dge.organ.DBD.0m@top.tags.edgeR))
rownames(dge.0m.split2) = c("all.0m", "all.organ.cov.0m", 
                              "no.rejection.0m", "no.rejection.early",
                              "subclin.0m", "sucblin.early.0m","any.rej",
                              "DCD.only.0m", "DBD.only.0m" )


cases.0m.split3 = cbind(dge.severity.0m@summary.numbers)
colnames(cases.0m.split3) = c("severity")

dge.0m.split3 = rbind(dge.summary(dge.severity.0m@top.tags.short.e),
                      dge.summary(dge.severity.0m@top.tags.long.e),
                      dge.summary(dge.severity.0m@top.tags.between.e))
rownames(dge.0m.split3) = c("short.cont.0m", "long.cont.0m", "long.short.0m")


cases.0m.split4 = dge.mgfr.0m@summary.numbers
dge.0m.split4 = rbind(dge.summary(dge.mgfr.0m@top.tags.fit.mGFR.e),
                 dge.summary(dge.mgfr.0m@top.tags.mGFR.DGF.e),
                 dge.summary(dge.mgfr.0m@top.tags.mGFR.DGF.low.e),
                 dge.summary(dge.mgfr.0m@top.tags.mGFR.DGF.high.e))
rownames(dge.0m.split4) = c("mGFR.high.v.low", "DGF.mGFR.high.v.low", 
                            "low.mGFR.DGF", "high.mGFR.DGF")


cases.0m.split5 = dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@summary.numbers
dge.0m.split5 = rbind(dge.summary(dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@DGF.IFTAhi.vs.noDGF.IFTAlo),
                 dge.summary(dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@DGF.IFTAlo.vs.noDGF.IFTAlo),
                 dge.summary(dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@noDGF.IFTA.hi.vs.noDGF.IFTAlo),
                 dge.summary(dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@DGF.IFTAhi.vs.DGF.IFTA.lo))
rownames(dge.0m.split5) = c("DGF.hiIFTA.vs.cont.loIFTA", 
                            "DGF.loIFTA.v.cont.loIFTA", 
                            "Cont.hiIFTA.v.Cont.loIFTA", 
                            "DGF.hiIFTA.v.DGF.loIFTA")


month0.compare= cbind(t(cases.0m.split2), dge.0m.split2)
month0.compare = cbind(rownames(dge.0m.split2), month0.compare)

month0.severity = cbind(t(cases.0m.split3), dge.0m.split3)
month0.severity= cbind(rownames(dge.0m.split3), month0.severity)

month0.dgf.mGFR = cbind(t(cases.0m.split4), dge.0m.split4)
month0.dgf.mGFR= cbind(rownames(dge.0m.split4), month0.dgf.mGFR)

month0.dgf.IFTA = cbind(t(cases.0m.split5), cases.0m.split5)
month0.dgf.IFTA= cbind(rownames(dge.0m.split5), month0.dgf.IFTA)

View(month0.compare)
View(month0.severity)
View(month0.dgf.mGFR)
View(month0.dgf.IFTA)

  dge.12mIFTA.notDGFsplit
  dge.12mmgfr.notDGFsplit 
  
  
```


Common genes for DGF that's useful
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
#############################################################################
# What genes are seen in DGF that portends poor long term IFTA results?
dge.12mIFTA.cov.ORGAN.DSA.REGRAFT = readRDS("rds/IFTA12m.DGF.COV.organ.DCD.dsa.regraft.0m.RDS")
# glimpse(dge.12mIFTA.cov.ORGAN.DSA.REGRAFT)

DGF.iftaHI.VS.noDGF.iftaLO = dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@DGF.IFTAhi.vs.noDGF.IFTAlo
DGF.iftalo.VS.noDGF.iftaLO = dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@DGF.IFTAlo.vs.noDGF.IFTAlo
noDGF.iftaHI.VS.noDGF.iftaLO = dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@noDGF.IFTA.hi.vs.noDGF.IFTAlo
DGF.iftaHI.VS.DGF.iftalo = dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@DGF.IFTAhi.vs.DGF.IFTA.lo

dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@summary.numbers
rbind(dge.summary(DGF.iftaHI.VS.noDGF.iftaLO), 
      dge.summary(DGF.iftalo.VS.noDGF.iftaLO), 
      dge.summary(noDGF.iftaHI.VS.noDGF.iftaLO),
      dge.summary(DGF.iftaHI.VS.DGF.iftalo))

set1a = DGF.iftaHI.VS.noDGF.iftaLO
set1a = set1a[!(set1a$symbol %in% DGF.iftalo.VS.noDGF.iftaLO$symbol),]

  Res = set1a[abs(set1a$logFC) > log2(1),]$logFC
  names(Res) = set1a[abs(set1a$logFC) > log2(1),]$ensembl
  Res = sort(Res, decreasing = TRUE)
  
  Res1 = set1a[abs(set1a$logFC) > log2(1),]$logFC
  names(Res1) = set1a[abs(set1a$logFC) > log2(1),]$symbol
  Res1 = sort(Res1, decreasing = TRUE)
  
  gse = gseGO(geneList=Res1,
              ont ="ALL",
              keyType = "SYMBOL",
              minGSSize = 3,
              maxGSSize = 8000,
              eps = 0,
              pvalueCutoff = 0.05,
              verbose = TRUE,
              seed = 23,
              OrgDb = org.Hs.eg.db)
  
  ggo.bp = filter(gse, p.adjust < 0.05, qvalues < 0.05)
  ggo.bp1 = as.data.frame(ggo.bp)
  
  ggo.bp1 = ggo.bp1 %>%
    dplyr::mutate(log.p.adj = -log(p.adjust)) %>% arrange(ggo.bp1, NES)
  
  ggo.bp1$Pathway.Direction = cut(-1*ggo.bp1$NES,
                                  c(-Inf,0,Inf), c("Upregulated", "Suppressed"))
  
  ggo.bp1$log.p.adj = ifelse(ggo.bp1$Pathway.Direction == "Suppressed",
                             -1*ggo.bp1$log.p.adj, ggo.bp1$log.p.adj)
  
  colnames(ggo.bp1)[4] = "Count"
  
  ggo.bp1 = removecrap(ggo.bp1)
  ggo.bp1 = pickrelevant(ggo.bp1)
  ggo.bp1 = trimmer(ggo.bp1)
  # 
  tiff("figures/gsea.DGF.IFTA.hi.lo.notincontrol.tiff", units = "cm", res = 300, height = 9, width = 18)
  ggplot(ggo.bp1, 
       aes(x = log.p.adj, y = Description)) + 
              geom_point(aes(color= p.adjust, size = Count)) +
              scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
              theme(axis.text.y = element_text(size = 12)) +
              xlab("-log10 of P-value") + ylab("Description") + 
              ggtitle("GO:GSEA") +  theme_bw() + geom_vline(xintercept=0)
  dev.off()
  
  # CNET plot
  dge.PT.cnet = pairwise_termsim(ggo.bp)
  dge.PT.cnet@result = dge.PT.cnet@result[dge.PT.cnet@result$ONTOLOGY == "BP",]
  dge.PT.cnet@result = dge.PT.cnet@result[dge.PT.cnet@result$Description %in% ggo.bp1$Description,]
                            
  dge.PT.cnet@result = dge.PT.cnet@result[dge.PT.cnet@result$Description %in% 
                                            c("B cell activation",
                                              "adaptive immune response" ,
                                              "antigen receptor-mediated signaling pathway"),]
  dge.PT.cnet@result = dge.PT.cnet@result[dge.PT.cnet@result$setSize>5,]
  
  tiff("figures/DGF.IFTA.cnet.tiff",  units = "cm", res = 300, height = 12, width = 18)
        cnetplot(dge.PT.cnet, 
                 categorySize = "p.adjust", 
                 colorEdge = TRUE, 
                 foldChange = Res1, 
                 node_label = "gene", 
                 showCategory = 10,
                 cex_label_category = 1)
  dev.off()
```



```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
#############################################################################
# What genes are seen in DGF that portends poor long term 12-month mGFR <45?
dge.mgfr.45.0m = readRDS("rds/dge.covorgan.noreject.nobk.dgfvsMGFR45.0m.RDS")
glimpse(dge.mgfr.45.0m)

control.lo.v.hi.mgfr =dge.mgfr.45.0m@control.lo.v.hi.mgfr
dgf.MGFRlo.v.cont.MGFRhi = dge.mgfr.45.0m@dgf.MGFRlo.v.cont.MGFRhi
dgf.MGFRlo.vs.DGF.MGFR.hi = dge.mgfr.45.0m@dgf.MGFRlo.vs.DGF.MGFR.hi
DGF.MGFRhi.vs.cont.MGFRhi = dge.mgfr.45.0m@DGF.MGFRhi.vs.cont.MGFRhi

dge.mgfr.45.0m@summary.numbers

rbind(dge.summary(control.lo.v.hi.mgfr), 
      dge.summary(dgf.MGFRlo.v.cont.MGFRhi), 
      dge.summary(dgf.MGFRlo.vs.DGF.MGFR.hi),
      dge.summary(DGF.MGFRhi.vs.cont.MGFRhi))

set2a = dgf.MGFRlo.v.cont.MGFRhi
set2a = set2a[!(set2a$symbol %in% DGF.MGFRhi.vs.cont.MGFRhi$symbol),]

# GSEA
  Res = set2a[abs(set2a$logFC) > log2(1),]$logFC
  names(Res) = set2a[abs(set2a$logFC) > log2(1),]$ensembl
  Res = sort(Res, decreasing = TRUE)
  
  Res1 = set2a[abs(set2a$logFC) > log2(1),]$logFC
  names(Res1) = set2a[abs(set2a$logFC) > log2(1),]$symbol
  Res1 = sort(Res1, decreasing = TRUE)
  
  gse1 = gseGO(geneList=Res1,
              ont ="ALL",
              keyType = "SYMBOL",
              minGSSize = 3,
              maxGSSize = 8000,
              eps = 0,
              pvalueCutoff = 0.05,
              verbose = TRUE,
              seed = 23,
              OrgDb = org.Hs.eg.db)
  
  ggo.bp = filter(gse1, p.adjust < 0.05, qvalues < 0.05)
  ggo.bp1 = as.data.frame(ggo.bp)
  
  ggo.bp1 = ggo.bp1 %>%
    dplyr::mutate(log.p.adj = -log(p.adjust)) %>% arrange(ggo.bp1, NES)
  
  ggo.bp1$Pathway.Direction = cut(-1*ggo.bp1$NES,
                                  c(-Inf,0,Inf), c("Upregulated", "Suppressed"))
  
  ggo.bp1$log.p.adj = ifelse(ggo.bp1$Pathway.Direction == "Suppressed",
                             -1*ggo.bp1$log.p.adj, ggo.bp1$log.p.adj)
  
  colnames(ggo.bp1)[4] = "Count"
  
  ggo.bp1 = removecrap(ggo.bp1)
  ggo.bp1 = pickrelevant(ggo.bp1)
  ggo.bp1 = trimmer(ggo.bp1)
  
  tiff("figures/gsea.DGF.MGFR.hi.lo.notincontrol.tiff", units = "cm", res = 300, height =9, width = 18)
  ggplot(ggo.bp1, 
       aes(x = log.p.adj, y = Description)) + 
              geom_point(aes(color= p.adjust, size = Count)) +
              scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
              theme(axis.text.y = element_text(size = 10)) +
              xlab("-log10 of P-value") + ylab("Description") + 
              ggtitle("GO:GSEA") +  theme_bw() + geom_vline(xintercept=0)
  dev.off()
  
  # CNET plot
  dge.PT.cnet = pairwise_termsim(ggo.bp)
  dge.PT.cnet@result = dge.PT.cnet@result[dge.PT.cnet@result$ONTOLOGY == "BP",]
  dge.PT.cnet@result = dge.PT.cnet@result[dge.PT.cnet@result$Description %in% 
                                            ggo.bp1$Description,]
  dge.PT.cnet@result = dge.PT.cnet@result[dge.PT.cnet@result$Description != 
                                            "antigen receptor-mediated signaling pathway",]

  tiff("figures/DGF.mGFR.cnet.tiff",  units = "cm", res = 300, height = 14, width = 20)
        cnetplot(dge.PT.cnet, 
                 categorySize = "p.adjust", 
                 colorEdge = TRUE, 
                 foldChange = Res1, 
                 node_label = "gene", 
                 showCategory = 10,
                 cex_label_category = 1.2)
  dev.off()
  
  ##################

  common.up = set1a[set1a$logFC>0,][set1a[set1a$logFC>0,]$symbol %in% set2a[set2a$logFC>0,]$symbol, ]
  common.down = set1a[set1a$logFC<0,][set1a[set1a$logFC<0,]$symbol %in% set2a[set2a$logFC<0,]$symbol, ]
  common = rbind(common.up, common.down)
  View(common)


```


GO:GSEA for 0m (pre implant/reperfusion time point)
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
# # 0m biopsies
# dge.0m.all = readRDS("rds/dge.0m.all.RDS")
# dge.0m.spk.ktx  = readRDS("rds/dge.0m.spk.ktx.norejection.RDS")
# dge.0m.dktx = readRDS("rds/dge.0m.dktx.RDS")
# dge.0m.dktx.noreject = readRDS("rds/dge.0m.dktx.norejection.RDS")
# dge.0m.ktx.dbd = readRDS("rds/dge.0m.dktx.dbd.RDS")
# dge.0m.ktx.dcd = readRDS("rds/dge.0m.dktx.dcd.RDS")
# dge.all.0m = readRDS("rds/dge.allpts.dgf.vs.control.0m.RDS")
# dge.organ.0m = readRDS("rds/dge.covariate.organ.dgf.vs.control.0m.RDS")
# dge.organ.no.reject.0m = readRDS("rds/dge.covorgan.nilreject.dgf.vs.con.0m.RDS")
# dge.organ.sub.reject.0m = readRDS("rds/dge.covorgan.subclin.dgf.vs.con.0m.RDS")
# dge.organ.bpar.reject.0m = readRDS("rds/dge.covorgan.bpar.dgf.vs.con.0m.RDS")
# dge.organ.any.reject.0m = readRDS("rds/dge.covorgan.anyreject.dgf.vs.con.0ml.RDS")
# dge.severity.0m = readRDS("rds/dge.covorgan.dgfseverity.0m.RDS")
# dge.mgfr.0m = readRDS("rds/dge.covorgan.noreject.nobk.dgfvsMGFR.0m.RDS")
# dge.severity.dcdonly.0m = readRDS("rds/dge.severity.DCDonly.0m.RDS")
# dge.severity.dbdonly.0m = readRDS("rds/dge.severity.DBDonly.0m.RDS")
# dge.organ.DCD.0m = readRDS("rds/dge.organ.DCDonly.0m.RDS")
# dge.organ.DBD.0m = readRDS("rds/dge.organ.DBDonly.0m.RDS")
# dge.organ.subclinEARLY.0m = readRDS("rds/dge.covorgan.subclinEARLY.dgf.vs.con.0m.RDS")
# dge.organ.subclinEARLY.severity.0m = readRDS("rds/dge.covorgan.dgfseverity.subclin.0m.RDS")
# dge.organ.severity.noreject.0m = readRDS("rds/dge.covorgan.dgfseverity.no.rejection.0m.RDS")
# dge.12mIFTA = readRDS("rds/IFTA.0m.biopsies.RDS")
# dge.12mIFTA.cov.ORGAN.DSA.REGRAFT = readRDS("rds/IFTA12m.DGF.COV.organ.DCD.dsa.regraft.0m.RDS")
# dge.noreject.early.0m = readRDS("rds/dge.covorgan.No.REJECT.EARLY.dgf.vs.con.0m.RDS")
# dge.mgfr.60.0m = readRDS("rds/dge.covorgan.noreject.nobk.dgfvsMGFR60.0m.RDS")
# dge.mgfr.45.0m = readRDS("rds/dge.covorgan.noreject.nobk.dgfvsMGFR45.0m.RDS")


dge.all = go.gsea(dge.all.0m@top.tags.edgeR,1)
p.summary.all = ggplot(dge.all$gsego.bp1, 
                   aes(x = log.p.adj, y = Description)) + 
              geom_point(aes(color= p.adjust, size = Count)) +
              scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
              theme(axis.text.y = element_text(size = 10)) +
              xlab("-log10 of P-value") + ylab("Description") + 
              ggtitle("GO:GSEA") +  theme_bw() + geom_vline(xintercept=0)
tiff("figures/gsea.genes.DGF.all.tiff", units = "cm", res = 300, height = 10, width = 15)
p.summary.all
dev.off()

##### NO enriched pathways for:
# organ.0m, organ.reject (any), mgfr, HD<7days vs control

##############################################################################
# Organ type as covariate takes out a significant set of DGE between DGE/control
################################################################################
dge.organ.up = dge.all.0m@top.tags.edgeR["logFC" > 0,]
dge.organ.up = dge.organ.up[!(dge.organ.up$symbol %in% 
                                dge.organ.0m@top.tags.edgeR["logFC" > 0,]$symbol),]
dge.organ.down = dge.all.0m@top.tags.edgeR["logFC" < 0,]
dge.organ.down = dge.organ.down[!(dge.organ.down$symbol %in% 
                                    dge.organ.0m@top.tags.edgeR["logFC" < 0,]$symbol),]
dge.organ = rbind(dge.organ.up, dge.organ.down)

View(dge.organ)

dge = go.gsea(dge.organ,1)
p.summary = ggplot(dge$gsego.bp1, 
                   aes(x = log.p.adj, y = Description)) + 
              geom_point(aes(color= p.adjust, size = Count)) +
              scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
              theme(axis.text.y = element_text(size = 10)) +
              xlab("-log10 of P-value") + ylab("Description") + 
              ggtitle("GO:GSEA") +  theme_bw() + geom_vline(xintercept=0)

tiff("figures/gsea.genes.removed.ORGAN.covariate.tiff", units = "cm", res = 300, height = 20, width = 17)
p.summary
dev.off()

gc()

#########################################################
# DBD, DCD and ECD all combined controlling for organ type
########################################################
# No enriched terms found for ... dge.organ.0m@top.tags.edgeR
# dge1 = go.gsea(dge.organ.0m@top.tags.edgeR,1)

#######################################
# DBD only allografts, DGF DGE to GSEA
#######################################
dge1 = go.gsea(dge.organ.DBD.0m@top.tags.edgeR,1)
p.summary1 = ggplot(dge1$gsego.bp1, 
                   aes(x = log.p.adj, y = Description)) + 
              geom_point(aes(color= p.adjust, size = Count)) +
              scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
              theme(axis.text.y = element_text(size = 10)) +
              xlab("-log10 of P-value") + ylab("Description") + 
              ggtitle("GO:GSEA") +  theme_bw() + geom_vline(xintercept=0)

tiff("figures/gsea.DBD.DGF.tiff", units = "cm", res = 300, height = 9, width = 15)
p.summary1
dev.off()
gc()

#######################################
# DCD only allografts, DGF DGE to GSEA
#######################################
dge2 = go.gsea(dge.organ.DCD.0m@top.tags.edgeR,1)
p.summary2 = ggplot(dge2$gsego.bp1, 
                   aes(x = log.p.adj, y = Description)) + 
              geom_point(aes(color= p.adjust, size = Count)) +
              scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
              theme(axis.text.y = element_text(size = 10)) +
              xlab("-log10 of P-value") + ylab("Description") + 
              ggtitle("GO:GSEA") +  theme_bw() + geom_vline(xintercept=0)
p.summary2

tiff("figures/gsea.DCD.DGF.tiff", units = "cm", res = 300, height = 12, width = 18)
p.summary2
dev.off()

gc()

############################
# Stratify for rejection 
############################
# No enriched pathways for 
# 1. go.gsea(dge.organ.no.reject.0m@top.tags.edgeR,1)
# 2. go.gsea(dge.organ.bpar.reject.0m@top.tags.edgeR,1)
# 3. go.gsea(dge.organ.any.reject.0m@top.tags.edgeR,1)

# For those who develop no rejection in first 3m, 0m bx rna
dge3 = go.gsea(dge.noreject.early.0m@top.tags.edgeR,1)
p.summary3 = ggplot(dge3$gsego.bp1, 
                   aes(x = log.p.adj, y = Description)) + 
              geom_point(aes(color= p.adjust, size = Count)) +
              scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
              theme(axis.text.y = element_text(size = 10)) +
              xlab("-log10 of P-value") + ylab("Description") + 
              ggtitle("GO:GSEA") +  theme_bw() + geom_vline(xintercept=0)
tiff("figures/gsea.noreject.0m.DGF.tiff", units = "cm", res = 300, height = 9, width = 18)
p.summary3
dev.off()
gc()


noBPAR.types = readRDS("rds/dge.covorgan.dcd.dsa.noBPAR.DGF.v.control.3m.RDS")
dge3a = go.gsea(noBPAR.types,1)

p.summary3a = ggplot(dge3a$gsego.bp1, 
                   aes(x = log.p.adj, y = Description)) + 
              geom_point(aes(color= p.adjust, size = Count)) +
              scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
              theme(axis.text.y = element_text(size = 10)) +
              xlab("-log10 of P-value") + ylab("Description") + 
              ggtitle("GO:GSEA") +  theme_bw() + geom_vline(xintercept=0)
tiff("figures/gsea.3mbiopsy.noBPAR.DGF.tiff", units = "cm", res = 300, height = 12, width = 18)
p.summary3a
dev.off()


############################
# DGF severity
############################
glimpse(dge.severity.dcdonly.0m)
# No enriched pathways for 
# 1. go.gsea(dge.severity.0m@top.tags.short.e,1) 
# 2. go.gsea(dge.severity.0m@top.tags.between.e,1)
# 3. go.gsea(dge.organ.subclinEARLY.severity.0m@top.tags.short.e,1) 
# 4. go.gsea(dge.organ.subclinEARLY.severity.0m@top.tags.between.e,1) 
# 5. go.gsea(dge.organ.severity.noreject.0m@top.tags.short.e,1) 

# DCD < 7 days vs control
dge4 = go.gsea(dge.severity.dcdonly.0m@top.tags.short.e,1) 
p.summary4 = ggplot(dge4$gsego.bp1, 
                   aes(x = log.p.adj, y = Description)) + 
              geom_point(aes(color= p.adjust, size = Count)) +
              scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
              theme(axis.text.y = element_text(size = 10)) +
              xlab("-log10 of P-value") + ylab("Description") + 
              ggtitle("GO:GSEA") +  theme_bw() + geom_vline(xintercept=0)
p.summary4
gc()

# DGF.HD vs DGF.Cr
dge5 = go.gsea(hd.cr@top.tags.edgeR,1) 
p.summary5 = ggplot(dge5$gsego.bp1, 
                   aes(x = -1*log.p.adj, y = Description)) + 
              geom_point(aes(color= p.adjust, size = Count)) +
              scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
              theme(axis.text.y = element_text(size = 10)) +
              xlab("-log10 of P-value") + ylab("Description") + 
              ggtitle("GO:GSEA DGF.Cr vs DGF.HD") +  theme_bw() + geom_vline(xintercept=0)
tiff("figures/gsea.0m.Cr.vs.HD.tiff", units = "cm", res = 300, height = 12, width = 18)
p.summary5
dev.off()
gc()


Cr.cont = DGF.Cr.vs.Cont(auscad.sorted@samples.biopsy.0m, 
                     auscad.sorted@counts.biopsy.0m, 
                     organ.covariate = TRUE)
  
dge5.a = go.gsea(hd.cr@top.tags.edgeR,1) 
p.summary5.a = ggplot(dge5.a$gsego.bp1, 
                   aes(x = log.p.adj, y = Description)) + 
              geom_point(aes(color= p.adjust, size = Count)) +
              scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
              theme(axis.text.y = element_text(size = 10)) +
              xlab("-log10 of P-value") + ylab("Description") + 
              ggtitle("GO:GSEA") +  theme_bw() + geom_vline(xintercept=0)
tiff("figures/gsea.0m.DGF.lo.CR.tiff", units = "cm", res = 300, height = 10, width = 18)
p.summary5.a
dev.off()
gc()



write_xlsx(list(Organ.type.Covariate = dge.organ,
                Organ.type.Covariate.gsea = dge$gsego.bp1, 
                DBD.DGF.v.control = dge.organ.DBD.0m@top.tags.edgeR,
                DBD.DGF.v.control.gsea = dge1$gsego.bp1,
                DCD.DGF.v.control = dge.organ.DCD.0m@top.tags.edgeR, 
                DCD.DGF.v.control.gsea = dge2$gsego.bp1, 
                Subclin.DGF.v.control = dge.organ.subclinEARLY.0m$subclin.only@top.tags.edgeR, 
                Subclin.DGF.v.control.gsea = dge3$gsego.bp1,
                noBPAR.3m.DGF.v.control = noBPAR.types, 
                noBPAR.3m.DGF.v.control.gsea = dge3a$gsego.bp1,
                Cr.v.control = Cr.cont@top.tags.edgeR, 
                Cr.v.control.gsea = dge5.a$gsego.bp1),
                "csv/dge.0m.summary.xlsx")

############################
# mGFR at 12 months (60 cut off)
############################
glimpse(dge.mgfr.0m)
# No enriched pathways for 
# 1. go.gsea(dge.mgfr.0m@top.tags.fit.mGFR.e,1) = control, high vs low mGFR
# 2. go.gsea(dge.mgfr.0m@top.tags.mGFR.DGF.e,1) = DGF, high vs low mGFR
# 3. go.gsea(dge.mgfr.0m@top.tags.mGFR.DGF.high.e,1) = mGFR > 60, DGF vs control
# 4. go.gsea(dge.mgfr.0m@top.tags.mGFR.DGF.low.e,1) = mGFR < 60, DGF vs control


#################################################################################
# IFTA 12 months with covariates: batch, organ_type, pre-existing DSA, regraft
# excluded patients with BKVAN and BPAR at any stage in 12 months
################################################################################
glimpse(dge.12mIFTA.cov.ORGAN.DSA.REGRAFT)
# No enriched pathways for 
# 1. go.gsea(dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@DGF.IFTAlo.vs.noDGF.IFTAlo,1)
# 2. go.gsea(dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@noDGF.IFTA.hi.vs.noDGF.IFTAlo,1) 
# 3. go.gsea(dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@DGF.IFTAhi.vs.DGF.IFTA.lo,1)

dge6 = go.gsea(dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@DGF.IFTAhi.vs.noDGF.IFTAlo,1) 
p.summary6 = ggplot(dge6$gsego.bp1, 
                   aes(x = log.p.adj, y = Description)) + 
              geom_point(aes(color= p.adjust, size = Count)) +
              scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
              theme(axis.text.y = element_text(size = 10)) +
              xlab("-log10 of P-value") + ylab("Description") + 
              ggtitle("GO:GSEA") +  theme_bw() + geom_vline(xintercept=0)
p.summary6
gc()

dge6a = go.gsea(dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@noDGF.IFTA.hi.vs.noDGF.IFTAlo,1) 
p.summary6a = ggplot(dge6a$gsego.bp1, 
                   aes(x = log.p.adj, y = Description)) + 
              geom_point(aes(color= p.adjust, size = Count)) +
              scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
              theme(axis.text.y = element_text(size = 10)) +
              xlab("-log10 of P-value") + ylab("Description") + 
              ggtitle("GO:GSEA") +  theme_bw() + geom_vline(xintercept=0)
p.summary6a
gc()

# separate.1 = dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@DGF.IFTAhi.vs.noDGF.IFTAlo
# separate.1 = separate.1[!(separate.1$symbol %in% dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@noDGF.IFTA.hi.vs.noDGF.IFTAlo$symbol),]

separate.1 = dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@DGF.IFTAhi.vs.noDGF.IFTAlo
separate.1 = separate.1[!(separate.1$symbol %in% dge.12mIFTA.cov.ORGAN.DSA.REGRAFT@DGF.IFTAlo.vs.noDGF.IFTAlo$symbol),]

dge7 = go.gsea(separate.1,1) 
p.summary7 = ggplot(dge7$gsego.bp1, 
                   aes(x = log.p.adj, y = Description)) + 
              geom_point(aes(color= p.adjust, size = Count)) +
              scale_color_viridis_c(guide=guide_colorbar(reverse=TRUE)) + 
              theme(axis.text.y = element_text(size = 10)) +
              xlab("-log10 of P-value") + ylab("Description") + 
              ggtitle("GO:GSEA") +  theme_bw() + geom_vline(xintercept=0)
p.summary7
gc()





```




Simple prediction DGF with genomics
```{r, echo=TRUE, results = FALSE, warning = FALSE, eval = FALSE}
```











